// @rjsf/utils/dist/utils.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('lodash/isPlainObject'), require('lodash/get'), require('fast-equals'), require('fast-uri'), require('jsonpointer'), require('lodash/isEmpty'), require('lodash/isObject'), require('lodash/omit'), require('lodash/has'), require('lodash/isString'), require('lodash/flattenDeep'), require('lodash/merge'), require('lodash/set'), require('lodash/times'), require('lodash/transform'), require('lodash/uniq'), require('lodash/union'), require('lodash/isNumber'), require('@x0k/json-schema-merge'), require('@x0k/json-schema-merge/lib/array'), require('lodash/reduce'), require('lodash/isNil'), require('lodash/pick'), require('lodash/cloneDeep'), require('lodash/setWith'), require('lodash/difference'), require('lodash/keys'), require('lodash/pickBy'), require('lodash/uniqueId'), require('react'), require('react-is'), require('react/jsx-runtime'), require('lodash/toPath'), require('lodash/forEach')) :
  typeof define === 'function' && define.amd ? define(['exports', 'lodash/isPlainObject', 'lodash/get', 'fast-equals', 'fast-uri', 'jsonpointer', 'lodash/isEmpty', 'lodash/isObject', 'lodash/omit', 'lodash/has', 'lodash/isString', 'lodash/flattenDeep', 'lodash/merge', 'lodash/set', 'lodash/times', 'lodash/transform', 'lodash/uniq', 'lodash/union', 'lodash/isNumber', '@x0k/json-schema-merge', '@x0k/json-schema-merge/lib/array', 'lodash/reduce', 'lodash/isNil', 'lodash/pick', 'lodash/cloneDeep', 'lodash/setWith', 'lodash/difference', 'lodash/keys', 'lodash/pickBy', 'lodash/uniqueId', 'react', 'react-is', 'react/jsx-runtime', 'lodash/toPath', 'lodash/forEach'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global["@rjsf/utils"] = {}, global.isPlainObject2, global.get15, global.fastEquals, global.UriResolver, global.jsonpointer, global.isEmpty4, global.isObject2, global.omit, global.has5, global.isString, global.flattenDeep, global.merge, global.set, global.times, global.transform, global.uniq, global.union, global.isNumber, global.jsonSchemaMerge, global.array, global.reduce, global.isNil, global.pick, global.cloneDeep, global.setWith, global.difference, global.keys, global.pickBy, global.uniqueId, global.react, global.ReactIs, global.jsxRuntime, global.toPath, global.forEach));
})(this, (function (exports, isPlainObject2, get15, fastEquals, UriResolver, jsonpointer, isEmpty4, isObject2, omit, has5, isString, flattenDeep, merge, set, times, transform, uniq, union, isNumber, jsonSchemaMerge, array, reduce, isNil, pick, cloneDeep, setWith, difference, keys, pickBy, uniqueId, react, ReactIs, jsxRuntime, toPath, forEach) { 'use strict';

  // src/isObject.ts
  function isObject(thing) {
    if (typeof thing !== "object" || thing === null) {
      return false;
    }
    if (typeof thing.lastModified === "number" && typeof File !== "undefined" && thing instanceof File) {
      return false;
    }
    if (typeof thing.getMonth === "function" && typeof Date !== "undefined" && thing instanceof Date) {
      return false;
    }
    return !Array.isArray(thing);
  }

  // src/allowAdditionalItems.ts
  function allowAdditionalItems(schema) {
    if (schema.additionalItems === true) {
      console.warn("additionalItems=true is currently not supported");
    }
    return isObject(schema.additionalItems);
  }

  // src/asNumber.ts
  function asNumber(value) {
    if (value === "") {
      return void 0;
    }
    if (value === null) {
      return null;
    }
    if (/\.$/.test(value)) {
      return value;
    }
    if (/\.0$/.test(value)) {
      return value;
    }
    if (/\.\d*0$/.test(value)) {
      return value;
    }
    const n = Number(value);
    const valid = typeof n === "number" && !Number.isNaN(n);
    return valid ? n : value;
  }

  // src/constants.ts
  var ADDITIONAL_PROPERTY_FLAG = "__additional_property";
  var ADDITIONAL_PROPERTIES_KEY = "additionalProperties";
  var ALL_OF_KEY = "allOf";
  var ANY_OF_KEY = "anyOf";
  var CONST_KEY = "const";
  var DEFAULT_KEY = "default";
  var DEFINITIONS_KEY = "definitions";
  var DEPENDENCIES_KEY = "dependencies";
  var ENUM_KEY = "enum";
  var ERRORS_KEY = "__errors";
  var ID_KEY = "$id";
  var IF_KEY = "if";
  var ITEMS_KEY = "items";
  var JUNK_OPTION_ID = "_$junk_option_schema_id$_";
  var NAME_KEY = "$name";
  var ONE_OF_KEY = "oneOf";
  var PATTERN_PROPERTIES_KEY = "patternProperties";
  var PROPERTIES_KEY = "properties";
  var READONLY_KEY = "readonly";
  var REQUIRED_KEY = "required";
  var SUBMIT_BTN_OPTIONS_KEY = "submitButtonOptions";
  var REF_KEY = "$ref";
  var RJSF_REF_KEY = "__rjsf_ref";
  var SCHEMA_KEY = "$schema";
  var DEFAULT_ID_PREFIX = "root";
  var DEFAULT_ID_SEPARATOR = "_";
  var DISCRIMINATOR_PATH = ["discriminator", "propertyName"];
  var FORM_CONTEXT_NAME = "formContext";
  var LOOKUP_MAP_NAME = "layoutGridLookupMap";
  var RJSF_ADDITIONAL_PROPERTIES_FLAG = "__rjsf_additionalProperties";
  var ROOT_SCHEMA_PREFIX = "__rjsf_rootSchema";
  var UI_FIELD_KEY = "ui:field";
  var UI_WIDGET_KEY = "ui:widget";
  var UI_OPTIONS_KEY = "ui:options";
  var UI_GLOBAL_OPTIONS_KEY = "ui:globalOptions";
  var UI_DEFINITIONS_KEY = "ui:definitions";
  var JSON_SCHEMA_DRAFT_2019_09 = "https://json-schema.org/draft/2019-09/schema";
  var JSON_SCHEMA_DRAFT_2020_12 = "https://json-schema.org/draft/2020-12/schema";

  // src/getUiOptions.ts
  function getUiOptions(uiSchema = {}, globalOptions = {}) {
    if (!uiSchema) {
      return { ...globalOptions };
    }
    return Object.keys(uiSchema).filter((key) => key.startsWith("ui:")).reduce(
      (options, key) => {
        const value = uiSchema[key];
        if (key === UI_WIDGET_KEY && isObject(value)) {
          console.error("Setting options via ui:widget object is no longer supported, use ui:options instead");
          return options;
        }
        if (key === UI_OPTIONS_KEY && isObject(value)) {
          return { ...options, ...value };
        }
        return { ...options, [key.substring(3)]: value };
      },
      { ...globalOptions }
    );
  }

  // src/canExpand.ts
  function canExpand(schema, uiSchema = {}, formData) {
    if (!(schema.additionalProperties || schema.patternProperties)) {
      return false;
    }
    const { expandable = true } = getUiOptions(uiSchema);
    if (expandable === false) {
      return expandable;
    }
    if (schema.maxProperties !== void 0 && formData) {
      return Object.keys(formData).length < schema.maxProperties;
    }
    return true;
  }
  function createErrorHandler(formData) {
    const handler = {
      // We store the list of errors for this node in a property named __errors
      // to avoid name collision with a possible sub schema field named
      // 'errors' (see `utils.toErrorSchema`).
      [ERRORS_KEY]: [],
      addError(message) {
        this[ERRORS_KEY].push(message);
      }
    };
    if (Array.isArray(formData)) {
      return formData.reduce((acc, value, key) => ({ ...acc, [key]: createErrorHandler(value) }), handler);
    }
    if (isPlainObject2(formData)) {
      const formObject = formData;
      return Object.keys(formObject).reduce(
        (acc, key) => ({ ...acc, [key]: createErrorHandler(formObject[key]) }),
        handler
      );
    }
    return handler;
  }
  var deepEquals = fastEquals.createCustomEqual({
    circular: true,
    createCustomConfig: () => ({
      areFunctionsEqual(_a, b) {
        return typeof b === "function";
      }
    })
  });
  var deepEquals_default = deepEquals;
  function findEmbeddedSchemaRecursive(schema, ref) {
    if (ID_KEY in schema && UriResolver.equal(schema[ID_KEY], ref)) {
      return schema;
    }
    for (const subSchema of Object.values(schema)) {
      if (Array.isArray(subSchema)) {
        for (const item of subSchema) {
          if (isObject2(item)) {
            const result = findEmbeddedSchemaRecursive(item, ref);
            if (result !== void 0) {
              return result;
            }
          }
        }
      } else if (isObject2(subSchema)) {
        const result = findEmbeddedSchemaRecursive(subSchema, ref);
        if (result !== void 0) {
          return result;
        }
      }
    }
    return void 0;
  }
  function makeAllReferencesAbsolute(schema, baseURI) {
    const currentURI = get15(schema, ID_KEY, baseURI);
    let result = schema;
    if (REF_KEY in result) {
      result = { ...result, [REF_KEY]: UriResolver.resolve(currentURI, result[REF_KEY]) };
    }
    for (const [key, subSchema] of Object.entries(result)) {
      if (Array.isArray(subSchema)) {
        result = {
          ...result,
          [key]: subSchema.map((item) => isObject2(item) ? makeAllReferencesAbsolute(item, currentURI) : item)
        };
      } else if (isObject2(subSchema)) {
        result = { ...result, [key]: makeAllReferencesAbsolute(subSchema, currentURI) };
      }
    }
    return result;
  }
  function splitKeyElementFromObject(key, object) {
    const value = object[key];
    const remaining = omit(object, [key]);
    return [remaining, value];
  }
  function findSchemaDefinitionRecursive($ref, rootSchema = {}, recurseList = [], baseURI = get15(rootSchema, [ID_KEY])) {
    const ref = $ref || "";
    let current = void 0;
    let currentBaseURI = baseURI;
    if (ref.startsWith("#")) {
      const decodedRef = decodeURIComponent(ref.substring(1));
      if (currentBaseURI === void 0 || ID_KEY in rootSchema && rootSchema[ID_KEY] === currentBaseURI) {
        current = jsonpointer.get(rootSchema, decodedRef);
      } else if (rootSchema[SCHEMA_KEY] === JSON_SCHEMA_DRAFT_2020_12) {
        current = findEmbeddedSchemaRecursive(rootSchema, currentBaseURI.replace(/\/$/, ""));
        if (current !== void 0) {
          current = jsonpointer.get(current, decodedRef);
        }
      }
    } else if (rootSchema[SCHEMA_KEY] === JSON_SCHEMA_DRAFT_2020_12) {
      const resolvedRef = currentBaseURI ? UriResolver.resolve(currentBaseURI, ref) : ref;
      const [refId, ...refAnchor] = resolvedRef.replace(/#\/?$/, "").split("#");
      current = findEmbeddedSchemaRecursive(rootSchema, refId.replace(/\/$/, ""));
      if (current !== void 0) {
        currentBaseURI = current[ID_KEY];
        if (!isEmpty4(refAnchor)) {
          current = jsonpointer.get(current, decodeURIComponent(refAnchor.join("#")));
        }
      }
    }
    if (current === void 0) {
      throw new Error(`Could not find a definition for ${$ref}.`);
    }
    const nextRef = current[REF_KEY];
    if (nextRef) {
      if (recurseList.includes(nextRef)) {
        if (recurseList.length === 1) {
          throw new Error(`Definition for ${$ref} is a circular reference`);
        }
        const [firstRef, ...restRefs] = recurseList;
        const circularPath = [...restRefs, ref, firstRef].join(" -> ");
        throw new Error(`Definition for ${firstRef} contains a circular reference through ${circularPath}`);
      }
      const [remaining, theRef] = splitKeyElementFromObject(REF_KEY, current);
      const subSchema = findSchemaDefinitionRecursive(theRef, rootSchema, [...recurseList, ref], currentBaseURI);
      if (Object.keys(remaining).length > 0) {
        if (rootSchema[SCHEMA_KEY] === JSON_SCHEMA_DRAFT_2019_09 || rootSchema[SCHEMA_KEY] === JSON_SCHEMA_DRAFT_2020_12) {
          return { [ALL_OF_KEY]: [remaining, subSchema] };
        }
        return { ...remaining, ...subSchema };
      }
      return subSchema;
    }
    return current;
  }
  function findSchemaDefinition($ref, rootSchema = {}, baseURI = get15(rootSchema, [ID_KEY])) {
    const recurseList = [];
    return findSchemaDefinitionRecursive($ref, rootSchema, recurseList, baseURI);
  }
  function getDiscriminatorFieldFromSchema(schema) {
    let discriminator;
    const maybeString = get15(schema, DISCRIMINATOR_PATH);
    if (isString(maybeString)) {
      discriminator = maybeString;
    } else if (maybeString !== void 0) {
      console.warn(`Expecting discriminator to be a string, got "${typeof maybeString}" instead`);
    }
    return discriminator;
  }

  // src/guessType.ts
  function guessType(value) {
    if (Array.isArray(value)) {
      return "array";
    }
    if (typeof value === "string") {
      return "string";
    }
    if (value == null) {
      return "null";
    }
    if (typeof value === "boolean") {
      return "boolean";
    }
    if (typeof value === "number" && !Number.isNaN(value)) {
      return "number";
    }
    if (typeof value === "object") {
      return "object";
    }
    return "string";
  }

  // src/getSchemaType.ts
  function getSchemaType(schema) {
    let { type } = schema;
    if (!type && schema.const) {
      return guessType(schema.const);
    }
    if (!type && schema.enum) {
      return "string";
    }
    if (!type && (schema.properties || schema.additionalProperties || schema.patternProperties)) {
      return "object";
    }
    if (Array.isArray(type)) {
      if (type.length === 2 && type.includes("null")) {
        type = type.find((t) => t !== "null");
      } else {
        type = type[0];
      }
    }
    return type;
  }

  // src/mergeSchemas.ts
  function mergeSchemas(obj1, obj2) {
    const acc = { ...obj1 };
    return Object.keys(obj2).reduce((accumulator, key) => {
      const left = obj1 ? obj1[key] : {}, right = obj2[key];
      if (obj1 && key in obj1 && isObject(right)) {
        accumulator[key] = mergeSchemas(left, right);
      } else if (obj1 && obj2 && (getSchemaType(obj1) === "object" || getSchemaType(obj2) === "object") && key === REQUIRED_KEY && Array.isArray(left) && Array.isArray(right)) {
        accumulator[key] = union(left, right);
      } else {
        accumulator[key] = right;
      }
      return accumulator;
    }, acc);
  }
  function getOptionMatchingSimpleDiscriminator(formData, options, discriminatorField) {
    if (formData && discriminatorField) {
      const value = get15(formData, discriminatorField);
      if (value === void 0) {
        return void 0;
      }
      for (let i = 0; i < options.length; i += 1) {
        const option = options[i];
        const discriminator = get15(option, [PROPERTIES_KEY, discriminatorField], {});
        if (discriminator.type !== "object" && discriminator.type !== "array") {
          if (discriminator.const === value) {
            return i;
          }
          if (discriminator.enum?.includes(value)) {
            return i;
          }
        }
      }
    }
    return void 0;
  }

  // src/schema/getFirstMatchingOption.ts
  function getFirstMatchingOption(validator, formData, options, rootSchema, discriminatorField) {
    if (formData === void 0) {
      return 0;
    }
    const simpleDiscriminatorMatch = getOptionMatchingSimpleDiscriminator(formData, options, discriminatorField);
    if (isNumber(simpleDiscriminatorMatch)) {
      return simpleDiscriminatorMatch;
    }
    for (let i = 0; i < options.length; i += 1) {
      const option = options[i];
      if (discriminatorField && has5(option, [PROPERTIES_KEY, discriminatorField])) {
        const value = get15(formData, discriminatorField);
        const discriminator = get15(option, [PROPERTIES_KEY, discriminatorField], {});
        if (validator.isValid(discriminator, value, rootSchema)) {
          return i;
        }
      } else if (option[PROPERTIES_KEY]) {
        const requiresAnyOf = {
          anyOf: Object.keys(option[PROPERTIES_KEY]).map((key) => ({
            required: [key]
          }))
        };
        let augmentedSchema;
        if (option.anyOf) {
          const { ...shallowClone } = option;
          if (!shallowClone.allOf) {
            shallowClone.allOf = [];
          } else {
            shallowClone.allOf = shallowClone.allOf.slice();
          }
          shallowClone.allOf.push(requiresAnyOf);
          augmentedSchema = shallowClone;
        } else {
          augmentedSchema = { ...option, ...requiresAnyOf };
        }
        delete augmentedSchema.required;
        if (validator.isValid(augmentedSchema, formData, rootSchema)) {
          return i;
        }
      } else if (validator.isValid(option, formData, rootSchema)) {
        return i;
      }
    }
    return 0;
  }
  var { compareSchemaDefinitions, compareSchemaValues } = jsonSchemaMerge.createComparator();
  var { mergeArrayOfSchemaDefinitions } = jsonSchemaMerge.createMerger({
    intersectJson: array.createIntersector(compareSchemaValues),
    deduplicateJsonSchemaDef: array.createDeduplicator(compareSchemaDefinitions)
  });
  var shallowAllOfMerge_default = jsonSchemaMerge.createShallowAllOfMerge(mergeArrayOfSchemaDefinitions);

  // src/schema/retrieveSchema.ts
  function retrieveSchema(validator, schema, rootSchema = {}, rawFormData, experimental_customMergeAllOf, resolveAnyOfOrOneOfRefs = false) {
    return retrieveSchemaInternal(
      validator,
      schema,
      rootSchema,
      rawFormData,
      void 0,
      void 0,
      experimental_customMergeAllOf,
      resolveAnyOfOrOneOfRefs
    )[0];
  }
  function resolveCondition(validator, schema, rootSchema, expandAllBranches, recurseList, formData, experimental_customMergeAllOf) {
    const { if: expression, then, else: otherwise, ...resolvedSchemaLessConditional } = schema;
    const conditionValue = validator.isValid(expression, formData || {}, rootSchema);
    let resolvedSchemas = [resolvedSchemaLessConditional];
    let schemas = [];
    if (expandAllBranches) {
      if (then && typeof then !== "boolean") {
        schemas = schemas.concat(
          retrieveSchemaInternal(
            validator,
            then,
            rootSchema,
            formData,
            expandAllBranches,
            recurseList,
            experimental_customMergeAllOf
          )
        );
      }
      if (otherwise && typeof otherwise !== "boolean") {
        schemas = schemas.concat(
          retrieveSchemaInternal(
            validator,
            otherwise,
            rootSchema,
            formData,
            expandAllBranches,
            recurseList,
            experimental_customMergeAllOf
          )
        );
      }
    } else {
      const conditionalSchema = conditionValue ? then : otherwise;
      if (conditionalSchema && typeof conditionalSchema !== "boolean") {
        schemas = schemas.concat(
          retrieveSchemaInternal(
            validator,
            conditionalSchema,
            rootSchema,
            formData,
            expandAllBranches,
            recurseList,
            experimental_customMergeAllOf
          )
        );
      }
    }
    if (schemas.length) {
      resolvedSchemas = schemas.map((s) => mergeSchemas(resolvedSchemaLessConditional, s));
    }
    return resolvedSchemas.flatMap(
      (s) => retrieveSchemaInternal(
        validator,
        s,
        rootSchema,
        formData,
        expandAllBranches,
        recurseList,
        experimental_customMergeAllOf
      )
    );
  }
  function getAllPermutationsOfXxxOf(listOfLists) {
    const allPermutations = listOfLists.reduce(
      (permutations, list) => {
        if (list.length > 1) {
          return list.flatMap((element) => times(permutations.length, (i) => [...permutations[i]].concat(element)));
        }
        permutations.forEach((permutation) => permutation.push(list[0]));
        return permutations;
      },
      [[]]
      // Start with an empty list
    );
    return allPermutations;
  }
  function getMatchingPatternProperties(schema, key) {
    return Object.keys(schema.patternProperties).filter((pattern) => RegExp(pattern).test(key)).reduce(
      (obj, pattern) => {
        set(obj, [pattern], schema.patternProperties[pattern]);
        return obj;
      },
      {}
    );
  }
  function resolveSchema(validator, schema, rootSchema, expandAllBranches, recurseList, formData, experimental_customMergeAllOf, resolveAnyOfOrOneOfRefs) {
    const updatedSchemas = resolveReference(
      validator,
      schema,
      rootSchema,
      expandAllBranches,
      recurseList,
      formData,
      experimental_customMergeAllOf,
      resolveAnyOfOrOneOfRefs
    );
    if (updatedSchemas.length > 1 || updatedSchemas[0] !== schema) {
      return updatedSchemas;
    }
    if (DEPENDENCIES_KEY in schema) {
      const resolvedSchemas = resolveDependencies(
        validator,
        schema,
        rootSchema,
        expandAllBranches,
        recurseList,
        formData,
        experimental_customMergeAllOf
      );
      return resolvedSchemas.flatMap(
        (s) => retrieveSchemaInternal(
          validator,
          s,
          rootSchema,
          formData,
          expandAllBranches,
          recurseList,
          experimental_customMergeAllOf
        )
      );
    }
    if (ALL_OF_KEY in schema && Array.isArray(schema[ALL_OF_KEY])) {
      const allOfSchemaElements = schema.allOf.map(
        (allOfSubschema) => retrieveSchemaInternal(
          validator,
          allOfSubschema,
          rootSchema,
          formData,
          expandAllBranches,
          recurseList,
          experimental_customMergeAllOf
        )
      );
      const allPermutations = getAllPermutationsOfXxxOf(allOfSchemaElements);
      return allPermutations.map((permutation) => ({
        ...schema,
        allOf: permutation
      }));
    }
    return [schema];
  }
  function resolveReference(validator, schema, rootSchema, expandAllBranches, recurseList, formData, experimental_customMergeAllOf, resolveAnyOfOrOneOfRefs) {
    const updatedSchema = resolveAllReferences(schema, rootSchema, recurseList, void 0, resolveAnyOfOrOneOfRefs);
    if (updatedSchema !== schema) {
      return retrieveSchemaInternal(
        validator,
        updatedSchema,
        rootSchema,
        formData,
        expandAllBranches,
        recurseList,
        experimental_customMergeAllOf,
        resolveAnyOfOrOneOfRefs
      );
    }
    return [schema];
  }
  function resolveAllReferences(schema, rootSchema, recurseList, baseURI, resolveAnyOfOrOneOfRefs) {
    if (!isObject(schema)) {
      return schema;
    }
    let resolvedSchema = schema;
    let currentBaseURI = baseURI;
    if (REF_KEY in resolvedSchema) {
      const { $ref, ...localSchema } = resolvedSchema;
      if (recurseList.includes($ref)) {
        return resolvedSchema;
      }
      recurseList.push($ref);
      const refSchema = findSchemaDefinition($ref, rootSchema, currentBaseURI);
      resolvedSchema = { ...refSchema, ...localSchema, [RJSF_REF_KEY]: $ref };
      if (ID_KEY in resolvedSchema) {
        currentBaseURI = resolvedSchema[ID_KEY];
      }
    }
    if (PROPERTIES_KEY in resolvedSchema) {
      const childrenLists = [];
      const updatedProps = transform(
        resolvedSchema[PROPERTIES_KEY],
        (acc, value, key) => {
          const childList = [...recurseList];
          acc[key] = resolveAllReferences(value, rootSchema, childList, currentBaseURI, resolveAnyOfOrOneOfRefs);
          childrenLists.push(childList);
        },
        {}
      );
      merge(recurseList, uniq(flattenDeep(childrenLists)));
      resolvedSchema = { ...resolvedSchema, [PROPERTIES_KEY]: updatedProps };
    }
    if (ITEMS_KEY in resolvedSchema && !Array.isArray(resolvedSchema.items) && typeof resolvedSchema.items !== "boolean") {
      resolvedSchema = {
        ...resolvedSchema,
        items: resolveAllReferences(
          resolvedSchema.items,
          rootSchema,
          recurseList,
          currentBaseURI,
          resolveAnyOfOrOneOfRefs
        )
      };
    }
    if (resolveAnyOfOrOneOfRefs) {
      let key;
      let schemas;
      if (ANY_OF_KEY in schema && Array.isArray(schema[ANY_OF_KEY])) {
        key = ANY_OF_KEY;
        schemas = resolvedSchema[ANY_OF_KEY];
      } else if (ONE_OF_KEY in schema && Array.isArray(schema[ONE_OF_KEY])) {
        key = ONE_OF_KEY;
        schemas = resolvedSchema[ONE_OF_KEY];
      }
      if (key && schemas) {
        resolvedSchema = {
          ...resolvedSchema,
          [key]: schemas.map(
            (s) => resolveAllReferences(s, rootSchema, recurseList, currentBaseURI, resolveAnyOfOrOneOfRefs)
          )
        };
      }
    }
    return deepEquals_default(schema, resolvedSchema) ? schema : resolvedSchema;
  }
  function stubExistingAdditionalProperties(validator, theSchema, rootSchema, aFormData, experimental_customMergeAllOf) {
    const schema = {
      ...theSchema,
      properties: { ...theSchema.properties }
    };
    const formData = aFormData && isObject(aFormData) ? aFormData : {};
    Object.keys(formData).forEach((key) => {
      if (key in schema.properties) {
        return;
      }
      if (PATTERN_PROPERTIES_KEY in schema) {
        const matchingProperties = getMatchingPatternProperties(schema, key);
        if (!isEmpty4(matchingProperties)) {
          schema.properties[key] = retrieveSchema(
            validator,
            { [ALL_OF_KEY]: Object.values(matchingProperties) },
            rootSchema,
            get15(formData, [key]),
            experimental_customMergeAllOf
          );
          set(schema.properties, [key, ADDITIONAL_PROPERTY_FLAG], true);
          return;
        }
      }
      if (ADDITIONAL_PROPERTIES_KEY in schema && schema.additionalProperties !== false) {
        let additionalProperties;
        if (typeof schema.additionalProperties !== "boolean") {
          if (REF_KEY in schema.additionalProperties) {
            additionalProperties = retrieveSchema(
              validator,
              { [REF_KEY]: get15(schema.additionalProperties, [REF_KEY]) },
              rootSchema,
              formData,
              experimental_customMergeAllOf
            );
          } else if ("type" in schema.additionalProperties) {
            additionalProperties = { ...schema.additionalProperties };
          } else if (ANY_OF_KEY in schema.additionalProperties || ONE_OF_KEY in schema.additionalProperties) {
            additionalProperties = {
              type: "object",
              ...schema.additionalProperties
            };
          } else {
            additionalProperties = { type: guessType(get15(formData, [key])) };
          }
        } else {
          additionalProperties = { type: guessType(get15(formData, [key])) };
        }
        schema.properties[key] = additionalProperties;
        set(schema.properties, [key, ADDITIONAL_PROPERTY_FLAG], true);
      } else {
        schema.properties[key] = { type: "null" };
        set(schema.properties, [key, ADDITIONAL_PROPERTY_FLAG], true);
      }
    });
    return schema;
  }
  function mergeAllOf(schema) {
    return shallowAllOfMerge_default(schema);
  }
  function retrieveSchemaInternal(validator, schema, rootSchema, rawFormData, expandAllBranches = false, recurseList = [], experimental_customMergeAllOf, resolveAnyOfOrOneOfRefs) {
    if (!isObject(schema)) {
      return [{}];
    }
    const resolvedSchemas = resolveSchema(
      validator,
      schema,
      rootSchema,
      expandAllBranches,
      recurseList,
      rawFormData,
      experimental_customMergeAllOf,
      resolveAnyOfOrOneOfRefs
    );
    return resolvedSchemas.flatMap((s) => {
      let resolvedSchema = s;
      if (IF_KEY in resolvedSchema) {
        return resolveCondition(
          validator,
          resolvedSchema,
          rootSchema,
          expandAllBranches,
          recurseList,
          rawFormData,
          experimental_customMergeAllOf
        );
      }
      if (ALL_OF_KEY in resolvedSchema) {
        if (expandAllBranches) {
          const { allOf, ...restOfSchema } = resolvedSchema;
          return [...allOf, restOfSchema];
        }
        try {
          const withContainsSchemas = [];
          const withoutContainsSchemas = [];
          resolvedSchema.allOf?.forEach((allOfItem) => {
            if (typeof allOfItem === "object" && allOfItem.contains) {
              withContainsSchemas.push(allOfItem);
            } else {
              withoutContainsSchemas.push(allOfItem);
            }
          });
          if (withContainsSchemas.length) {
            resolvedSchema = { ...resolvedSchema, allOf: withoutContainsSchemas };
          }
          resolvedSchema = experimental_customMergeAllOf ? experimental_customMergeAllOf(resolvedSchema) : mergeAllOf(resolvedSchema);
          if (withContainsSchemas.length) {
            resolvedSchema.allOf = withContainsSchemas;
          }
        } catch (e) {
          console.warn("could not merge subschemas in allOf:\n", e);
          const { allOf, ...resolvedSchemaWithoutAllOf } = resolvedSchema;
          return resolvedSchemaWithoutAllOf;
        }
      }
      if (PROPERTIES_KEY in resolvedSchema && PATTERN_PROPERTIES_KEY in resolvedSchema) {
        resolvedSchema = Object.keys(resolvedSchema.properties).reduce(
          (acc, key) => {
            const matchingProperties = getMatchingPatternProperties(acc, key);
            if (!isEmpty4(matchingProperties)) {
              acc.properties[key] = retrieveSchema(
                validator,
                { allOf: [acc.properties[key], ...Object.values(matchingProperties)] },
                rootSchema,
                get15(rawFormData, [key]),
                experimental_customMergeAllOf
              );
            }
            return acc;
          },
          {
            ...resolvedSchema,
            properties: { ...resolvedSchema.properties }
          }
        );
      }
      const hasAdditionalProperties = PATTERN_PROPERTIES_KEY in resolvedSchema || ADDITIONAL_PROPERTIES_KEY in resolvedSchema && resolvedSchema.additionalProperties !== false;
      if (hasAdditionalProperties) {
        return stubExistingAdditionalProperties(
          validator,
          resolvedSchema,
          rootSchema,
          rawFormData,
          experimental_customMergeAllOf
        );
      }
      return resolvedSchema;
    });
  }
  function resolveAnyOrOneOfSchemas(validator, schema, rootSchema, expandAllBranches, rawFormData) {
    let anyOrOneOf;
    const { oneOf, anyOf, ...remaining } = schema;
    if (Array.isArray(oneOf)) {
      anyOrOneOf = oneOf;
    } else if (Array.isArray(anyOf)) {
      anyOrOneOf = anyOf;
    }
    if (anyOrOneOf) {
      const formData = rawFormData === void 0 && expandAllBranches ? {} : rawFormData;
      const discriminator = getDiscriminatorFieldFromSchema(schema);
      anyOrOneOf = anyOrOneOf.map((s) => resolveAllReferences(s, rootSchema, []));
      const option = getFirstMatchingOption(validator, formData, anyOrOneOf, rootSchema, discriminator);
      if (expandAllBranches) {
        const relaxed = relaxOptionsForScoring(anyOrOneOf, false, rootSchema);
        getFirstMatchingOption(validator, formData, relaxed, rootSchema, discriminator);
        return anyOrOneOf.map((item) => mergeSchemas(remaining, item));
      }
      return [mergeSchemas(remaining, anyOrOneOf[option])];
    }
    return [schema];
  }
  function relaxOptionsForScoring(options, resolveRefs = false, rootSchema) {
    return options.map((d) => {
      if (!isObject(d)) {
        return d ? {} : { not: {} };
      }
      const schema = resolveRefs && rootSchema ? resolveAllReferences(d, rootSchema, []) : d;
      return schema.additionalProperties === false ? { ...schema, additionalProperties: true } : schema;
    });
  }
  function resolveDependencies(validator, schema, rootSchema, expandAllBranches, recurseList, formData, experimental_customMergeAllOf) {
    const { dependencies, ...remainingSchema } = schema;
    const resolvedSchemas = resolveAnyOrOneOfSchemas(
      validator,
      remainingSchema,
      rootSchema,
      expandAllBranches,
      formData
    );
    return resolvedSchemas.flatMap(
      (resolvedSchema) => processDependencies(
        validator,
        dependencies,
        resolvedSchema,
        rootSchema,
        expandAllBranches,
        recurseList,
        formData,
        experimental_customMergeAllOf
      )
    );
  }
  function processDependencies(validator, dependencies, resolvedSchema, rootSchema, expandAllBranches, recurseList, formData, experimental_customMergeAllOf) {
    let schemas = [resolvedSchema];
    for (const dependencyKey in dependencies) {
      if ((expandAllBranches || get15(formData, [dependencyKey]) !== void 0) && (!resolvedSchema.properties || dependencyKey in resolvedSchema.properties)) {
        const [remainingDependencies, dependencyValue] = splitKeyElementFromObject(
          dependencyKey,
          dependencies
        );
        if (Array.isArray(dependencyValue)) {
          schemas[0] = withDependentProperties(resolvedSchema, dependencyValue);
        } else if (isObject(dependencyValue)) {
          schemas = withDependentSchema(
            validator,
            resolvedSchema,
            rootSchema,
            dependencyKey,
            dependencyValue,
            expandAllBranches,
            recurseList,
            formData,
            experimental_customMergeAllOf
          );
        }
        return schemas.flatMap(
          (schema) => processDependencies(
            validator,
            remainingDependencies,
            schema,
            rootSchema,
            expandAllBranches,
            recurseList,
            formData,
            experimental_customMergeAllOf
          )
        );
      }
    }
    return schemas;
  }
  function withDependentProperties(schema, additionallyRequired) {
    if (!additionallyRequired) {
      return schema;
    }
    const required = Array.isArray(schema.required) ? Array.from(/* @__PURE__ */ new Set([...schema.required, ...additionallyRequired])) : additionallyRequired;
    return { ...schema, required };
  }
  function withDependentSchema(validator, schema, rootSchema, dependencyKey, dependencyValue, expandAllBranches, recurseList, formData, experimental_customMergeAllOf) {
    const dependentSchemas = retrieveSchemaInternal(
      validator,
      dependencyValue,
      rootSchema,
      formData,
      expandAllBranches,
      recurseList,
      experimental_customMergeAllOf
    );
    return dependentSchemas.flatMap((dependent) => {
      const { oneOf, ...dependentSchema } = dependent;
      const mergedSchema = mergeSchemas(schema, dependentSchema);
      if (oneOf === void 0) {
        return mergedSchema;
      }
      const resolvedOneOfs = oneOf.map((subschema) => {
        if (typeof subschema === "boolean" || !(REF_KEY in subschema)) {
          return [subschema];
        }
        return resolveReference(validator, subschema, rootSchema, expandAllBranches, recurseList, formData);
      });
      const allPermutations = getAllPermutationsOfXxxOf(resolvedOneOfs);
      return allPermutations.flatMap(
        (resolvedOneOf) => withExactlyOneSubschema(
          validator,
          mergedSchema,
          rootSchema,
          dependencyKey,
          resolvedOneOf,
          expandAllBranches,
          recurseList,
          formData,
          experimental_customMergeAllOf
        )
      );
    });
  }
  function withExactlyOneSubschema(validator, schema, rootSchema, dependencyKey, oneOf, expandAllBranches, recurseList, formData, experimental_customMergeAllOf) {
    const validSubschemas = oneOf.filter((subschema) => {
      if (typeof subschema === "boolean" || !subschema || !subschema.properties) {
        return false;
      }
      const { [dependencyKey]: conditionPropertySchema } = subschema.properties;
      if (conditionPropertySchema) {
        const conditionSchema = {
          type: "object",
          properties: {
            [dependencyKey]: conditionPropertySchema
          }
        };
        return validator.isValid(conditionSchema, formData, rootSchema) || expandAllBranches;
      }
      return false;
    });
    if (!expandAllBranches && validSubschemas.length !== 1) {
      console.warn("ignoring oneOf in dependencies because there isn't exactly one subschema that is valid");
      return [schema];
    }
    return validSubschemas.flatMap((s) => {
      const subschema = s;
      const [dependentSubschema] = splitKeyElementFromObject(dependencyKey, subschema.properties);
      const dependentSchema = { ...subschema, properties: dependentSubschema };
      const schemas = retrieveSchemaInternal(
        validator,
        dependentSchema,
        rootSchema,
        formData,
        expandAllBranches,
        recurseList,
        experimental_customMergeAllOf
      );
      return schemas.map((resolvedSubschema) => mergeSchemas(schema, resolvedSubschema));
    });
  }

  // src/schema/findSelectedOptionInXxxOf.ts
  function findSelectedOptionInXxxOf(validator, rootSchema, schema, fallbackField, xxx, formData = {}, experimental_customMergeAllOf) {
    if (Array.isArray(schema[xxx])) {
      const discriminator = getDiscriminatorFieldFromSchema(schema);
      const selectorField = discriminator || fallbackField;
      const xxxOfs = schema[xxx].map(
        (xxxOf) => retrieveSchema(validator, xxxOf, rootSchema, formData, experimental_customMergeAllOf)
      );
      const data = get15(formData, selectorField);
      if (data !== void 0) {
        return xxxOfs.find(
          (xxxOfOption) => deepEquals_default(
            get15(
              xxxOfOption,
              [PROPERTIES_KEY, selectorField, DEFAULT_KEY],
              get15(xxxOfOption, [PROPERTIES_KEY, selectorField, CONST_KEY])
            ),
            data
          )
        );
      }
    }
    return void 0;
  }
  function getFromSchemaInternal(validator, rootSchema, schema, path, experimental_customMergeAllOf) {
    let fieldSchema = schema;
    if (has5(schema, REF_KEY)) {
      fieldSchema = retrieveSchema(validator, schema, rootSchema, void 0, experimental_customMergeAllOf);
    }
    if (isEmpty4(path)) {
      return fieldSchema;
    }
    const pathList = Array.isArray(path) ? [...path] : path.split(".");
    const [part, ...nestedPath] = pathList;
    if (part !== void 0 && part !== "" && has5(fieldSchema, part)) {
      fieldSchema = get15(fieldSchema, part);
      return getFromSchemaInternal(
        validator,
        rootSchema,
        fieldSchema,
        nestedPath,
        experimental_customMergeAllOf
      );
    }
    return void 0;
  }
  function getFromSchema(validator, rootSchema, schema, path, defaultValue, experimental_customMergeAllOf) {
    const result = getFromSchemaInternal(validator, rootSchema, schema, path, experimental_customMergeAllOf);
    if (result === void 0) {
      return defaultValue;
    }
    return result;
  }

  // src/schema/findFieldInSchema.ts
  var NOT_FOUND_SCHEMA = { title: "!@#$_UNKNOWN_$#@!" };
  function findFieldInSchema(validator, rootSchema, schema, path, formData = {}, experimental_customMergeAllOf) {
    const pathList = Array.isArray(path) ? [...path] : path.split(".");
    let parentField = schema;
    const fieldName = pathList.pop();
    const fieldNameKey = String(fieldName);
    if (pathList.length) {
      pathList.forEach((subPath) => {
        parentField = getFromSchema(
          validator,
          rootSchema,
          parentField,
          [PROPERTIES_KEY, subPath],
          {},
          experimental_customMergeAllOf
        );
        if (has5(parentField, ONE_OF_KEY)) {
          parentField = findSelectedOptionInXxxOf(
            validator,
            rootSchema,
            parentField,
            fieldNameKey,
            ONE_OF_KEY,
            get15(formData, subPath),
            experimental_customMergeAllOf
          );
        } else if (has5(parentField, ANY_OF_KEY)) {
          parentField = findSelectedOptionInXxxOf(
            validator,
            rootSchema,
            parentField,
            fieldNameKey,
            ANY_OF_KEY,
            get15(formData, subPath),
            experimental_customMergeAllOf
          );
        }
      });
    }
    if (has5(parentField, ONE_OF_KEY)) {
      parentField = findSelectedOptionInXxxOf(
        validator,
        rootSchema,
        parentField,
        fieldNameKey,
        ONE_OF_KEY,
        formData,
        experimental_customMergeAllOf
      );
    } else if (has5(parentField, ANY_OF_KEY)) {
      parentField = findSelectedOptionInXxxOf(
        validator,
        rootSchema,
        parentField,
        fieldNameKey,
        ANY_OF_KEY,
        formData,
        experimental_customMergeAllOf
      );
    }
    let field = getFromSchema(
      validator,
      rootSchema,
      parentField,
      [PROPERTIES_KEY, fieldName],
      NOT_FOUND_SCHEMA,
      experimental_customMergeAllOf
    );
    if (field === NOT_FOUND_SCHEMA) {
      field = void 0;
    }
    const requiredArray = getFromSchema(
      validator,
      rootSchema,
      parentField,
      REQUIRED_KEY,
      [],
      experimental_customMergeAllOf
    );
    let isRequired;
    if (field && Array.isArray(requiredArray)) {
      isRequired = requiredArray.includes(fieldNameKey);
    }
    return { field, isRequired };
  }
  var JUNK_OPTION = {
    type: "object",
    $id: JUNK_OPTION_ID,
    properties: {
      __not_really_there__: {
        type: "number"
      }
    }
  };
  function calculateIndexScore(validator, rootSchema, schema, formData, experimental_customMergeAllOf) {
    let totalScore = 0;
    if (schema) {
      if (isObject2(schema.properties)) {
        totalScore += reduce(
          schema.properties,
          (score, value, key) => {
            const formValue = get15(formData, key);
            if (typeof value === "boolean") {
              return score;
            }
            if (has5(value, REF_KEY)) {
              const newSchema = retrieveSchema(
                validator,
                value,
                rootSchema,
                formValue,
                experimental_customMergeAllOf
              );
              return score + calculateIndexScore(
                validator,
                rootSchema,
                newSchema,
                formValue || {},
                experimental_customMergeAllOf
              );
            }
            if ((has5(value, ONE_OF_KEY) || has5(value, ANY_OF_KEY)) && formValue) {
              const xxxOfKey = has5(value, ONE_OF_KEY) ? ONE_OF_KEY : ANY_OF_KEY;
              const discriminator = getDiscriminatorFieldFromSchema(value);
              return score + getClosestMatchingOption(
                validator,
                rootSchema,
                formValue,
                get15(value, xxxOfKey),
                -1,
                discriminator,
                experimental_customMergeAllOf
              );
            }
            if (value.type === "object") {
              const structureBoost = isObject2(formValue) ? 1 : 0;
              return score + structureBoost + calculateIndexScore(validator, rootSchema, value, formValue, experimental_customMergeAllOf);
            }
            if (value.type === guessType(formValue)) {
              let newScore = score + 1;
              if (value.default) {
                newScore += formValue === value.default ? 1 : -1;
              } else if (value.const) {
                newScore += formValue === value.const ? 1 : -1;
              }
              return newScore;
            }
            return score;
          },
          0
        );
      } else if (isString(schema.type) && schema.type === guessType(formData)) {
        totalScore += 1;
      }
    }
    return totalScore;
  }
  function getClosestMatchingOption(validator, rootSchema, formData, options, selectedOption = -1, discriminatorField, experimental_customMergeAllOf) {
    const resolvedOptions = options.map((option) => resolveAllReferences(option, rootSchema, []));
    const simpleDiscriminatorMatch = getOptionMatchingSimpleDiscriminator(formData, options, discriminatorField);
    if (isNumber(simpleDiscriminatorMatch)) {
      return simpleDiscriminatorMatch;
    }
    const allValidIndexes = resolvedOptions.reduce((validList, option, index) => {
      const testOptions = [JUNK_OPTION, option];
      const match = getFirstMatchingOption(validator, formData, testOptions, rootSchema, discriminatorField);
      if (match === 1) {
        validList.push(index);
      }
      return validList;
    }, []);
    if (allValidIndexes.length === 1) {
      return allValidIndexes[0];
    }
    if (!allValidIndexes.length) {
      times(resolvedOptions.length, (i) => allValidIndexes.push(i));
    }
    const scoreCount = /* @__PURE__ */ new Set();
    const { bestIndex } = allValidIndexes.reduce(
      (scoreData, index) => {
        const { bestScore } = scoreData;
        const option = resolvedOptions[index];
        const score = calculateIndexScore(validator, rootSchema, option, formData, experimental_customMergeAllOf);
        scoreCount.add(score);
        if (score > bestScore) {
          return { bestIndex: index, bestScore: score };
        }
        return scoreData;
      },
      { bestIndex: selectedOption, bestScore: 0 }
    );
    if (scoreCount.size === 1 && selectedOption >= 0) {
      return selectedOption;
    }
    return bestIndex;
  }
  function constIsAjvDataReference(schema) {
    const schemaConst = schema[CONST_KEY];
    const schemaType = getSchemaType(schema);
    return isObject(schemaConst) && isString(schemaConst?.$data) && schemaType !== "object" && schemaType !== "array";
  }

  // src/isConstant.ts
  function isConstant(schema) {
    return Array.isArray(schema.enum) && schema.enum.length === 1 || CONST_KEY in schema;
  }

  // src/isFixedItems.ts
  function isFixedItems(schema) {
    return Array.isArray(schema.items) && schema.items.length > 0 && schema.items.every((item) => isObject(item));
  }
  function mergeDefaultsWithFormData(defaults, formData, mergeExtraArrayDefaults = false, defaultSupercedesUndefined = false, overrideFormDataWithDefaults = false) {
    if (Array.isArray(formData)) {
      const defaultsArray = Array.isArray(defaults) ? defaults : [];
      const overrideArray = overrideFormDataWithDefaults ? defaultsArray : formData;
      const overrideOppositeArray = overrideFormDataWithDefaults ? formData : defaultsArray;
      const mapped = overrideArray.map((value, idx) => {
        if (overrideOppositeArray[idx] !== void 0) {
          return mergeDefaultsWithFormData(
            defaultsArray[idx],
            formData[idx],
            mergeExtraArrayDefaults,
            defaultSupercedesUndefined,
            overrideFormDataWithDefaults
          );
        }
        return value;
      });
      if ((mergeExtraArrayDefaults || overrideFormDataWithDefaults) && mapped.length < overrideOppositeArray.length) {
        mapped.push(...overrideOppositeArray.slice(mapped.length));
      }
      return mapped;
    }
    if (isObject(formData)) {
      const acc = Object.assign({}, defaults);
      return Object.keys(formData).reduce((accumulator, key) => {
        const keyValue = get15(formData, key);
        const keyExistsInDefaults = isObject(defaults) && key in defaults;
        const keyExistsInFormData = key in formData;
        const keyDefault = get15(defaults, key) ?? {};
        const defaultValueIsNestedObject = keyExistsInDefaults && isObject(keyDefault) && Object.values(keyDefault).some((v) => isObject(v));
        const keyDefaultIsObject = keyExistsInDefaults && isObject(get15(defaults, key));
        const keyHasFormDataObject = keyExistsInFormData && isObject(keyValue);
        if (keyDefaultIsObject && keyHasFormDataObject && !defaultValueIsNestedObject) {
          accumulator[key] = {
            ...get15(defaults, key),
            ...keyValue
          };
          return accumulator;
        }
        accumulator[key] = mergeDefaultsWithFormData(
          get15(defaults, key),
          keyValue,
          mergeExtraArrayDefaults,
          defaultSupercedesUndefined,
          // overrideFormDataWithDefaults can be true only when the key value exists in defaults
          // Or if the key value doesn't exist in formData
          overrideFormDataWithDefaults && (keyExistsInDefaults || !keyExistsInFormData)
        );
        return acc;
      }, acc);
    }
    if (defaultSupercedesUndefined && (!(defaults === void 0) && isNil(formData) || typeof formData === "number" && Number.isNaN(formData)) || overrideFormDataWithDefaults && !isNil(formData)) {
      return defaults;
    }
    return formData;
  }

  // src/mergeObjects.ts
  function mergeObjects(obj1, obj2, concatArrays = false) {
    return Object.keys(obj2).reduce(
      (acc, key) => {
        const left = obj1 ? obj1[key] : {}, right = obj2[key];
        if (obj1 && key in obj1 && isObject(right)) {
          acc[key] = mergeObjects(left, right, concatArrays);
        } else if (concatArrays && Array.isArray(left) && Array.isArray(right)) {
          let toMerge = right;
          if (concatArrays === "preventDuplicates") {
            toMerge = right.reduce((result, value) => {
              if (!left.includes(value)) {
                result.push(value);
              }
              return result;
            }, []);
          }
          acc[key] = left.concat(toMerge);
        } else {
          acc[key] = right;
        }
        return acc;
      },
      { ...obj1 }
    );
  }

  // src/toConstant.ts
  function toConstant(schema) {
    if (ENUM_KEY in schema && Array.isArray(schema.enum) && schema.enum.length === 1) {
      return schema.enum[0];
    }
    if (CONST_KEY in schema) {
      return schema.const;
    }
    throw new Error("schema cannot be inferred as a constant");
  }

  // src/optionsList.ts
  function applyEnumOrder(options, order) {
    const optionsByValue = new Map(options.map((opt) => [String(opt.value), opt]));
    const orderedKeys = new Set(order.filter((v) => v !== "*").map(String));
    const rest = options.filter((opt) => !orderedKeys.has(String(opt.value)));
    return order.flatMap((entry) => {
      if (entry === "*") {
        return rest;
      }
      const opt = optionsByValue.get(String(entry));
      return opt ? [opt] : [];
    });
  }
  function optionsList(schema, uiSchema) {
    if (schema.enum) {
      let enumNames;
      let enumOrder;
      if (uiSchema) {
        const { enumNames: uiEnumNames, enumOrder: uiEnumOrder } = getUiOptions(uiSchema);
        enumNames = uiEnumNames;
        enumOrder = uiEnumOrder;
      }
      let options = schema.enum.map((value, i) => {
        const label = Array.isArray(enumNames) ? enumNames[i] || String(value) : enumNames?.[String(value)] || String(value);
        return { label, value };
      });
      if (enumOrder) {
        options = applyEnumOrder(options, enumOrder);
      }
      return options;
    }
    let altSchemas = void 0;
    let altUiSchemas = void 0;
    if (schema.anyOf) {
      altSchemas = schema.anyOf;
      altUiSchemas = uiSchema?.anyOf;
    } else if (schema.oneOf) {
      altSchemas = schema.oneOf;
      altUiSchemas = uiSchema?.oneOf;
    }
    let selectorField = getDiscriminatorFieldFromSchema(schema);
    if (uiSchema) {
      const { optionsSchemaSelector = selectorField } = getUiOptions(uiSchema);
      selectorField = optionsSchemaSelector;
    }
    return altSchemas && altSchemas.map((aSchemaDef, index) => {
      const { title } = getUiOptions(altUiSchemas?.[index]);
      const aSchema = aSchemaDef;
      let value;
      let label = title;
      if (selectorField) {
        const innerSchema = get15(aSchema, [PROPERTIES_KEY, selectorField], {});
        value = get15(innerSchema, DEFAULT_KEY, get15(innerSchema, CONST_KEY));
        label = label || innerSchema?.title || aSchema.title || String(value);
      } else {
        value = toConstant(aSchema);
        label = label || aSchema.title || String(value);
      }
      return {
        schema: aSchema,
        label,
        value
      };
    });
  }

  // src/schema/isSelect.ts
  function isSelect(validator, theSchema, rootSchema = {}, experimental_customMergeAllOf) {
    const schema = retrieveSchema(validator, theSchema, rootSchema, void 0, experimental_customMergeAllOf);
    const altSchemas = schema.oneOf || schema.anyOf;
    if (Array.isArray(schema.enum)) {
      return true;
    }
    if (Array.isArray(altSchemas)) {
      return altSchemas.every((altSchema) => typeof altSchema !== "boolean" && isConstant(altSchema));
    }
    return false;
  }

  // src/schema/isMultiSelect.ts
  function isMultiSelect(validator, schema, rootSchema, experimental_customMergeAllOf) {
    if (!schema.uniqueItems || !schema.items || typeof schema.items === "boolean") {
      return false;
    }
    return isSelect(validator, schema.items, rootSchema, experimental_customMergeAllOf);
  }

  // src/schema/getDefaultFormState.ts
  var PRIMITIVE_TYPES = ["string", "number", "integer", "boolean", "null"];
  function getInnerSchemaForArrayItem(schema, additionalItems = 0 /* Ignore */, idx = -1) {
    if (idx >= 0) {
      if (Array.isArray(schema.items) && idx < schema.items.length) {
        const item = schema.items[idx];
        if (typeof item !== "boolean") {
          return item;
        }
      }
    } else if (schema.items && !Array.isArray(schema.items) && typeof schema.items !== "boolean") {
      return schema.items;
    }
    if (additionalItems !== 0 /* Ignore */ && isObject(schema.additionalItems)) {
      return schema.additionalItems;
    }
    return {};
  }
  function computeDefaultBasedOnSchemaTypeAndDefaults(schema, computedDefault) {
    const { default: schemaDefault, type } = schema;
    const shouldReturnNullAsDefault = Array.isArray(type) && type.includes("null") && isEmpty4(computedDefault) && schemaDefault === null;
    return shouldReturnNullAsDefault ? null : computedDefault;
  }
  function maybeAddDefaultToObject(acc, key, computedDefault, includeUndefinedValues, isParentRequired, requiredFields = [], experimental_defaultFormStateBehavior = {}, isConst = false, isNullType = false) {
    const { emptyObjectFields = "populateAllDefaults" } = experimental_defaultFormStateBehavior;
    if (includeUndefinedValues === true || isConst) {
      acc[key] = computedDefault;
    } else if (includeUndefinedValues === "excludeObjectChildren") {
      if (isNullType && computedDefault !== void 0 || !isObject(computedDefault) || !isEmpty4(computedDefault)) {
        acc[key] = computedDefault;
      }
    } else if (emptyObjectFields !== "skipDefaults") {
      const isSelfOrParentRequired = isParentRequired === void 0 ? requiredFields.includes(key) : isParentRequired;
      if (isObject(computedDefault)) {
        if (emptyObjectFields === "skipEmptyDefaults") {
          if (!isEmpty4(computedDefault)) {
            acc[key] = computedDefault;
          }
        } else if ((!isEmpty4(computedDefault) || requiredFields.includes(key)) && (isSelfOrParentRequired || emptyObjectFields !== "populateRequiredDefaults")) {
          acc[key] = computedDefault;
        }
      } else if (
        // Store computedDefault if it's a defined primitive (e.g., true) and satisfies certain conditions
        // Condition 1: computedDefault is not undefined
        // Condition 2: If emptyObjectFields is 'populateAllDefaults' or 'skipEmptyDefaults)
        // Or if isSelfOrParentRequired is 'true' and the key is a required field
        computedDefault !== void 0 && (emptyObjectFields === "populateAllDefaults" || emptyObjectFields === "skipEmptyDefaults" || isSelfOrParentRequired && requiredFields.includes(key))
      ) {
        acc[key] = computedDefault;
      }
    }
  }
  function computeDefaults(validator, rawSchema, computeDefaultsProps = {}) {
    const {
      parentDefaults,
      rawFormData,
      rootSchema = {},
      includeUndefinedValues = false,
      _recurseList = [],
      experimental_defaultFormStateBehavior = void 0,
      experimental_customMergeAllOf = void 0,
      required,
      shouldMergeDefaultsIntoFormData = false,
      initialDefaultsGenerated
    } = computeDefaultsProps;
    let formData = isObject(rawFormData) ? rawFormData : {};
    const schema = isObject(rawSchema) ? rawSchema : {};
    let defaults = parentDefaults;
    let schemaToCompute = null;
    let experimental_dfsb_to_compute = experimental_defaultFormStateBehavior;
    let updatedRecurseList = _recurseList;
    if (schema[CONST_KEY] !== void 0 && experimental_defaultFormStateBehavior?.constAsDefaults !== "never" && !constIsAjvDataReference(schema)) {
      defaults = schema[CONST_KEY];
    } else if (isObject(defaults) && isObject(schema.default) && !schema[ANY_OF_KEY] && !schema[ONE_OF_KEY] && !schema[REF_KEY]) {
      defaults = mergeObjects(defaults, schema.default);
    } else if (DEFAULT_KEY in schema && !schema[ANY_OF_KEY] && !schema[ONE_OF_KEY] && !schema[REF_KEY]) {
      defaults = schema.default;
    } else if (REF_KEY in schema) {
      const refName = schema[REF_KEY];
      if (!_recurseList.includes(refName)) {
        updatedRecurseList = _recurseList.concat(refName);
        schemaToCompute = findSchemaDefinition(refName, rootSchema);
      }
      const hasNoExistingData = rawFormData === void 0 || isObject(rawFormData) && isEmpty4(rawFormData);
      if (schemaToCompute && !defaults && hasNoExistingData) {
        defaults = schema.default;
      }
      if (shouldMergeDefaultsIntoFormData && schemaToCompute && !isObject(rawFormData)) {
        formData = rawFormData;
      }
    } else if (DEPENDENCIES_KEY in schema) {
      const defaultFormData = {
        ...getDefaultBasedOnSchemaType(validator, schema, computeDefaultsProps, defaults),
        ...formData
      };
      const resolvedSchema = resolveDependencies(
        validator,
        schema,
        rootSchema,
        false,
        [],
        defaultFormData,
        experimental_customMergeAllOf
      );
      [schemaToCompute] = resolvedSchema;
    } else if (isFixedItems(schema)) {
      defaults = schema.items.map(
        (itemSchema, idx) => computeDefaults(validator, itemSchema, {
          rootSchema,
          includeUndefinedValues,
          _recurseList,
          experimental_defaultFormStateBehavior,
          experimental_customMergeAllOf,
          parentDefaults: Array.isArray(parentDefaults) ? parentDefaults[idx] : void 0,
          rawFormData: formData,
          required,
          shouldMergeDefaultsIntoFormData
        })
      );
    } else if (ONE_OF_KEY in schema) {
      const { oneOf, ...remaining } = schema;
      if (oneOf.length === 0) {
        return void 0;
      }
      const discriminator = getDiscriminatorFieldFromSchema(schema);
      const { type = "null" } = remaining;
      if (!Array.isArray(type) && PRIMITIVE_TYPES.includes(type) && experimental_dfsb_to_compute?.constAsDefaults === "skipOneOf") {
        experimental_dfsb_to_compute = {
          ...experimental_dfsb_to_compute,
          constAsDefaults: "never"
        };
      }
      schemaToCompute = oneOf[getClosestMatchingOption(
        validator,
        rootSchema,
        rawFormData ?? schema.default,
        oneOf,
        0,
        discriminator,
        experimental_customMergeAllOf
      )];
      schemaToCompute = mergeSchemas(remaining, schemaToCompute);
    } else if (ANY_OF_KEY in schema) {
      const { anyOf, ...remaining } = schema;
      if (anyOf.length === 0) {
        return void 0;
      }
      const discriminator = getDiscriminatorFieldFromSchema(schema);
      schemaToCompute = anyOf[getClosestMatchingOption(
        validator,
        rootSchema,
        rawFormData ?? schema.default,
        anyOf,
        0,
        discriminator,
        experimental_customMergeAllOf
      )];
      schemaToCompute = mergeSchemas(remaining, schemaToCompute);
    }
    if (schemaToCompute) {
      return computeDefaults(validator, schemaToCompute, {
        rootSchema,
        includeUndefinedValues,
        _recurseList: updatedRecurseList,
        experimental_defaultFormStateBehavior: experimental_dfsb_to_compute,
        experimental_customMergeAllOf,
        parentDefaults: defaults,
        rawFormData: rawFormData ?? formData,
        required,
        shouldMergeDefaultsIntoFormData,
        initialDefaultsGenerated
      });
    }
    if (defaults === void 0) {
      defaults = schema.default;
    }
    const defaultBasedOnSchemaType = getDefaultBasedOnSchemaType(validator, schema, computeDefaultsProps, defaults);
    let defaultsWithFormData = defaultBasedOnSchemaType ?? defaults;
    if (shouldMergeDefaultsIntoFormData) {
      const { arrayMinItems = {} } = experimental_defaultFormStateBehavior || {};
      const { mergeExtraDefaults } = arrayMinItems;
      const matchingFormData = ensureFormDataMatchingSchema(
        validator,
        schema,
        rootSchema,
        rawFormData,
        experimental_defaultFormStateBehavior,
        experimental_customMergeAllOf
      );
      if (!isObject(rawFormData) || ALL_OF_KEY in schema) {
        defaultsWithFormData = mergeDefaultsWithFormData(
          defaultsWithFormData,
          matchingFormData,
          mergeExtraDefaults,
          true
        );
      }
    }
    return defaultsWithFormData;
  }
  function ensureFormDataMatchingSchema(validator, schema, rootSchema, formData, experimental_defaultFormStateBehavior, experimental_customMergeAllOf) {
    const isSelectField = !isConstant(schema) && isSelect(validator, schema, rootSchema, experimental_customMergeAllOf);
    let validFormData = formData;
    if (isSelectField) {
      const getOptionsList = optionsList(schema);
      const isValid = getOptionsList?.some((option) => deepEquals_default(option.value, formData));
      validFormData = isValid ? formData : void 0;
    }
    const constTakesPrecedence = schema[CONST_KEY] && experimental_defaultFormStateBehavior?.constAsDefaults === "always";
    if (constTakesPrecedence) {
      validFormData = schema.const;
    }
    return validFormData;
  }
  function getObjectDefaults(validator, rawSchema, {
    rawFormData,
    rootSchema = {},
    includeUndefinedValues = false,
    _recurseList = [],
    experimental_defaultFormStateBehavior = void 0,
    experimental_customMergeAllOf = void 0,
    required,
    shouldMergeDefaultsIntoFormData,
    initialDefaultsGenerated
  } = {}, defaults) {
    {
      const formData = isObject(rawFormData) ? rawFormData : {};
      const schema = rawSchema;
      const shouldRetrieveSchema = experimental_defaultFormStateBehavior?.allOf === "populateDefaults" && ALL_OF_KEY in schema || experimental_defaultFormStateBehavior?.emptyObjectFields !== "skipEmptyDefaults" && IF_KEY in schema;
      const retrievedSchema = shouldRetrieveSchema ? retrieveSchema(validator, schema, rootSchema, formData, experimental_customMergeAllOf) : schema;
      const parentConst = retrievedSchema[CONST_KEY];
      const objectDefaults = Object.keys(retrievedSchema.properties || {}).reduce(
        (acc, key) => {
          const propertySchema = get15(retrievedSchema, [PROPERTIES_KEY, key], {});
          const hasParentConst = isObject(parentConst) && parentConst[key] !== void 0;
          const hasConst = (isObject(propertySchema) && CONST_KEY in propertySchema || hasParentConst) && experimental_defaultFormStateBehavior?.constAsDefaults !== "never" && !constIsAjvDataReference(propertySchema);
          const computedDefault = computeDefaults(validator, propertySchema, {
            rootSchema,
            _recurseList,
            experimental_defaultFormStateBehavior,
            experimental_customMergeAllOf,
            includeUndefinedValues: includeUndefinedValues === true,
            parentDefaults: get15(defaults, [key]),
            rawFormData: get15(formData, [key]),
            required: retrievedSchema.required?.includes(key),
            shouldMergeDefaultsIntoFormData,
            initialDefaultsGenerated
          });
          maybeAddDefaultToObject(
            acc,
            key,
            computedDefault,
            includeUndefinedValues,
            required,
            retrievedSchema.required,
            experimental_defaultFormStateBehavior,
            hasConst,
            propertySchema?.type === "null"
          );
          return acc;
        },
        {}
      );
      if (retrievedSchema.additionalProperties && !initialDefaultsGenerated) {
        const additionalPropertiesSchema = isObject(retrievedSchema.additionalProperties) ? retrievedSchema.additionalProperties : {};
        const keys2 = /* @__PURE__ */ new Set();
        const formDataRequired = [];
        Object.keys(formData).filter((key) => !retrievedSchema.properties || !retrievedSchema.properties[key]).forEach((key) => {
          keys2.add(key);
          formDataRequired.push(key);
        });
        if (isObject(defaults) && formDataRequired.length === 0) {
          Object.keys(defaults).filter((key) => !retrievedSchema.properties || !retrievedSchema.properties[key]).forEach((key) => keys2.add(key));
        }
        keys2.forEach((key) => {
          const computedDefault = computeDefaults(validator, additionalPropertiesSchema, {
            rootSchema,
            _recurseList,
            experimental_defaultFormStateBehavior,
            experimental_customMergeAllOf,
            includeUndefinedValues: includeUndefinedValues === true,
            parentDefaults: get15(defaults, [key]),
            rawFormData: get15(formData, [key]),
            required: retrievedSchema.required?.includes(key),
            shouldMergeDefaultsIntoFormData,
            initialDefaultsGenerated
          });
          maybeAddDefaultToObject(
            objectDefaults,
            key,
            computedDefault,
            includeUndefinedValues,
            required,
            formDataRequired
          );
        });
      }
      return computeDefaultBasedOnSchemaTypeAndDefaults(rawSchema, objectDefaults);
    }
  }
  function getArrayDefaults(validator, rawSchema, {
    rawFormData,
    rootSchema = {},
    _recurseList = [],
    experimental_defaultFormStateBehavior = void 0,
    experimental_customMergeAllOf = void 0,
    required,
    requiredAsRoot = false,
    shouldMergeDefaultsIntoFormData,
    initialDefaultsGenerated
  } = {}, initialDefaults) {
    let defaults = initialDefaults;
    const schema = rawSchema;
    const arrayMinItemsStateBehavior = experimental_defaultFormStateBehavior?.arrayMinItems ?? {};
    const { populate: arrayMinItemsPopulate, mergeExtraDefaults: arrayMergeExtraDefaults } = arrayMinItemsStateBehavior;
    const neverPopulate = arrayMinItemsPopulate === "never";
    const ignoreMinItemsFlagSet = arrayMinItemsPopulate === "requiredOnly";
    const isPopulateAll = arrayMinItemsPopulate === "all" || !neverPopulate && !ignoreMinItemsFlagSet;
    const computeSkipPopulate = arrayMinItemsStateBehavior?.computeSkipPopulate ?? (() => false);
    const isSkipEmptyDefaults = experimental_defaultFormStateBehavior?.emptyObjectFields === "skipEmptyDefaults";
    const emptyDefault = isSkipEmptyDefaults ? void 0 : [];
    if (Array.isArray(defaults)) {
      defaults = defaults.map((item, idx) => {
        const schemaItem = getInnerSchemaForArrayItem(schema, 2 /* Fallback */, idx);
        const itemFormData = Array.isArray(rawFormData) ? rawFormData[idx] : void 0;
        return computeDefaults(validator, schemaItem, {
          rootSchema,
          _recurseList,
          experimental_defaultFormStateBehavior,
          experimental_customMergeAllOf,
          parentDefaults: item,
          rawFormData: itemFormData,
          required,
          shouldMergeDefaultsIntoFormData,
          initialDefaultsGenerated
        });
      });
    }
    if (Array.isArray(rawFormData)) {
      const schemaItem = getInnerSchemaForArrayItem(schema);
      if (neverPopulate) {
        defaults = rawFormData;
      } else {
        const itemDefaults = rawFormData.map(
          (item, idx) => computeDefaults(validator, schemaItem, {
            rootSchema,
            _recurseList,
            experimental_defaultFormStateBehavior,
            experimental_customMergeAllOf,
            rawFormData: item,
            parentDefaults: get15(defaults, [idx]),
            required,
            shouldMergeDefaultsIntoFormData,
            initialDefaultsGenerated
          })
        );
        const mergeExtraDefaults = (ignoreMinItemsFlagSet && required || isPopulateAll) && arrayMergeExtraDefaults;
        defaults = mergeDefaultsWithFormData(defaults, itemDefaults, mergeExtraDefaults);
      }
    }
    if (neverPopulate) {
      return defaults ?? emptyDefault;
    }
    if (ignoreMinItemsFlagSet && !required) {
      return defaults || void 0;
    }
    let arrayDefault;
    const defaultsLength = Array.isArray(defaults) ? defaults.length : 0;
    if (!schema.minItems || isMultiSelect(validator, schema, rootSchema, experimental_customMergeAllOf) || computeSkipPopulate(validator, schema, rootSchema) || schema.minItems <= defaultsLength) {
      arrayDefault = defaults || !required && !requiredAsRoot ? defaults : emptyDefault;
    } else {
      const defaultEntries = defaults || [];
      const fillerSchema = getInnerSchemaForArrayItem(schema, 1 /* Invert */);
      const fillerDefault = fillerSchema.default;
      const fillerEntries = Array.from(
        { length: schema.minItems - defaultsLength },
        () => computeDefaults(validator, fillerSchema, {
          parentDefaults: fillerDefault,
          rootSchema,
          _recurseList,
          experimental_defaultFormStateBehavior,
          experimental_customMergeAllOf,
          required,
          shouldMergeDefaultsIntoFormData
        })
      );
      arrayDefault = defaultEntries.concat(fillerEntries);
    }
    return computeDefaultBasedOnSchemaTypeAndDefaults(rawSchema, arrayDefault);
  }
  function getDefaultBasedOnSchemaType(validator, rawSchema, computeDefaultsProps = {}, defaults) {
    switch (getSchemaType(rawSchema)) {
      // We need to recurse for object schema inner default values.
      case "object": {
        return getObjectDefaults(validator, rawSchema, computeDefaultsProps, defaults);
      }
      case "array": {
        return getArrayDefaults(validator, rawSchema, computeDefaultsProps, defaults);
      }
      default:
        return void 0;
    }
  }
  function getDefaultFormState(validator, theSchema, formData, rootSchema, includeUndefinedValues = false, experimental_defaultFormStateBehavior, experimental_customMergeAllOf, initialDefaultsGenerated) {
    if (!isObject(theSchema)) {
      throw new Error(`Invalid schema: ${theSchema}`);
    }
    const schema = retrieveSchema(validator, theSchema, rootSchema, formData, experimental_customMergeAllOf);
    const defaults = computeDefaults(validator, schema, {
      rootSchema,
      includeUndefinedValues,
      experimental_defaultFormStateBehavior,
      experimental_customMergeAllOf,
      rawFormData: formData,
      shouldMergeDefaultsIntoFormData: true,
      initialDefaultsGenerated,
      requiredAsRoot: true
    });
    if (schema.type !== "object" && isObject(schema.default)) {
      return {
        ...defaults,
        ...formData
      };
    }
    if (isObject(formData) || Array.isArray(formData)) {
      const { mergeDefaultsIntoFormData } = experimental_defaultFormStateBehavior || {};
      const defaultSupercedesUndefined = mergeDefaultsIntoFormData === "useDefaultIfFormDataUndefined";
      const result = mergeDefaultsWithFormData(
        defaults,
        formData,
        true,
        // set to true to add any additional default array entries.
        defaultSupercedesUndefined,
        true
        // set to true to override formData with defaults if they exist.
      );
      return result;
    }
    return defaults;
  }

  // src/isCustomWidget.ts
  function isCustomWidget(uiSchema = {}) {
    return (
      // TODO: Remove the `&& uiSchema['ui:widget'] !== 'hidden'` once we support hidden widgets for arrays.
      // https://rjsf-team.github.io/react-jsonschema-form/docs/usage/widgets/#hidden-widgets
      "widget" in getUiOptions(uiSchema) && getUiOptions(uiSchema).widget !== "hidden"
    );
  }

  // src/schema/isFilesArray.ts
  function isFilesArray(validator, schema, uiSchema = {}, rootSchema, experimental_customMergeAllOf) {
    if (uiSchema[UI_WIDGET_KEY] === "files") {
      return true;
    }
    if (schema.items) {
      const itemsSchema = retrieveSchema(
        validator,
        schema.items,
        rootSchema,
        void 0,
        experimental_customMergeAllOf
      );
      return itemsSchema.type === "string" && itemsSchema.format === "data-url";
    }
    return false;
  }

  // src/schema/getDisplayLabel.ts
  function getDisplayLabel(validator, schema, uiSchema = {}, rootSchema, globalOptions, experimental_customMergeAllOf) {
    const uiOptions = getUiOptions(uiSchema, globalOptions);
    const { label = true } = uiOptions;
    let displayLabel = Boolean(label);
    if (displayLabel) {
      const schemaType = getSchemaType(schema);
      const addedByAdditionalProperty = get15(schema, ADDITIONAL_PROPERTY_FLAG, false);
      if (schemaType === "array") {
        displayLabel = addedByAdditionalProperty || isMultiSelect(validator, schema, rootSchema, experimental_customMergeAllOf) || isFilesArray(validator, schema, uiSchema, rootSchema, experimental_customMergeAllOf) || isCustomWidget(uiSchema);
      }
      if (schemaType === "object") {
        displayLabel = addedByAdditionalProperty;
      }
      if (schemaType === "boolean" && uiSchema && !uiSchema[UI_WIDGET_KEY]) {
        displayLabel = false;
      }
      if (uiSchema && uiSchema[UI_FIELD_KEY]) {
        displayLabel = false;
      }
    }
    return displayLabel;
  }
  function getUsedFormData(formData, fields) {
    if (fields.length === 0 && typeof formData !== "object") {
      return formData;
    }
    const data = pick(formData, fields);
    if (Array.isArray(formData)) {
      return Object.keys(data).map((key) => data[key]);
    }
    return data;
  }
  function getFieldNames(pathSchema, formData) {
    const formValueHasData = (value, isLeaf) => typeof value !== "object" || isEmpty4(value) || isLeaf && !isEmpty4(value);
    const getAllPaths = (_obj, acc = [], paths = [[]]) => {
      const objKeys = Object.keys(_obj);
      objKeys.forEach((key) => {
        const data = _obj[key];
        if (typeof data === "object") {
          const newPaths = paths.map((path) => [...path, key]);
          if (data[RJSF_ADDITIONAL_PROPERTIES_FLAG] && data[NAME_KEY] !== "") {
            acc.push(data[NAME_KEY]);
          } else {
            getAllPaths(data, acc, newPaths);
          }
        } else if (key === NAME_KEY && data !== "") {
          paths.forEach((path) => {
            const formValue = get15(formData, path);
            const isLeaf = objKeys.length === 1;
            if (formValueHasData(formValue, isLeaf) || Array.isArray(formValue) && formValue.every((val) => formValueHasData(val, isLeaf))) {
              acc.push(path);
            }
          });
        }
      });
      return acc;
    };
    return getAllPaths(pathSchema);
  }
  function isValueEmpty(value) {
    if (isNil(value) || value === "") {
      return true;
    }
    if (Array.isArray(value)) {
      return value.length === 0;
    }
    if (isObject(value)) {
      return Object.values(value).every(isValueEmpty);
    }
    return false;
  }
  function doMergeAllOf(schema, experimental_customMergeAllOf) {
    return experimental_customMergeAllOf ? experimental_customMergeAllOf(schema) : shallowAllOfMerge_default(schema);
  }
  function omitExtraData(validator, schema, rootSchema = {}, formData, experimental_customMergeAllOf) {
    function isObjectValue(value) {
      return isObject(value);
    }
    function isSchemaObj(schemaDef) {
      return isObject(schemaDef);
    }
    function handleObject(childSchema, source, target) {
      const { properties, additionalProperties, patternProperties, propertyNames } = childSchema;
      const requiredSet = new Set(childSchema.required ?? []);
      function setProperty(key, schemaDef, value, required = false) {
        const v = omit3(schemaDef, value, target[key]);
        if (!required && isObject(v)) {
          let sd = isSchemaObj(schemaDef) ? schemaDef : {};
          if (sd.$ref !== void 0) {
            sd = findSchemaDefinition(sd.$ref, rootSchema);
          }
          const innerRequired = new Set(sd.required ?? []);
          const shouldDrop = Object.entries(v).every(
            ([k, val]) => !innerRequired.has(k) && isValueEmpty(val)
          );
          if (shouldDrop) {
            return;
          }
        }
        if (v !== void 0) {
          target[key] = v;
        }
      }
      if (properties !== void 0) {
        for (const [key, schemaDef] of Object.entries(properties)) {
          setProperty(key, schemaDef, source[key], requiredSet.has(key));
        }
      }
      let patternPropertiesRest;
      if (patternProperties !== void 0) {
        patternPropertiesRest = [];
        const patterns = Object.entries(patternProperties).map(([pattern, schemaDef]) => [
          new RegExp(pattern),
          schemaDef
        ]);
        const knownProperties = new Set(Object.keys(properties ?? {}));
        for (const [key, value] of Object.entries(source)) {
          if (!knownProperties.has(key)) {
            const matched = patterns.find(([re]) => re.test(key));
            if (matched === void 0) {
              patternPropertiesRest.push(key);
            } else {
              setProperty(key, matched[1], value);
            }
          }
        }
      }
      if (additionalProperties !== void 0 && additionalProperties !== false) {
        const addlSchema = additionalProperties;
        if (patternPropertiesRest !== void 0) {
          for (const key of patternPropertiesRest) {
            setProperty(key, addlSchema, source[key]);
          }
        } else {
          const knownProperties = new Set(Object.keys(properties ?? {}));
          for (const [key, value] of Object.entries(source)) {
            if (!knownProperties.has(key)) {
              setProperty(key, addlSchema, value);
            }
          }
        }
      }
      if (propertyNames !== void 0) {
        for (const [key, value] of Object.entries(source)) {
          target[key] = value;
        }
      }
      return target;
    }
    function handleArray(childSchema, source, target) {
      const { items, additionalItems } = childSchema;
      if (items !== void 0) {
        if (Array.isArray(items)) {
          for (let i = 0; i < items.length; i += 1) {
            target.push(omit3(items[i], source[i]));
          }
        } else {
          for (let i = 0; i < source.length; i += 1) {
            target.push(omit3(items, source[i]));
          }
        }
      }
      if (additionalItems) {
        for (let i = target.length; i < source.length; i += 1) {
          target.push(omit3(additionalItems, source[i]));
        }
      }
      return target;
    }
    function handleConditions(childSchema, source, target) {
      const { if: condition, then, else: otherwise } = childSchema;
      if (condition === void 0) {
        return target;
      }
      const isThenBranch = isSchemaObj(condition) ? validator.isValid(condition, source, rootSchema) : condition;
      const branch = isThenBranch ? then : otherwise;
      return branch === void 0 ? target : omit3(branch, source, target);
    }
    function handleOneOf(oneOf, childSchema, source, target) {
      if (!Array.isArray(oneOf) || isSelect(validator, childSchema, rootSchema, experimental_customMergeAllOf)) {
        return target;
      }
      const scoringOptions = relaxOptionsForScoring(oneOf, true, rootSchema);
      const bestIndex = getClosestMatchingOption(
        validator,
        rootSchema,
        source,
        scoringOptions,
        0,
        getDiscriminatorFieldFromSchema(childSchema),
        experimental_customMergeAllOf
      );
      const winning = oneOf[bestIndex];
      const resolved = isObject(winning) ? resolveAllReferences(winning, rootSchema, []) : scoringOptions[bestIndex];
      return omit3(resolved, source, target);
    }
    function handleAnyOf(childSchema, source, target) {
      const { anyOf } = childSchema;
      if (!Array.isArray(anyOf)) {
        return target;
      }
      if (source === void 0 || Array.isArray(source) && source.length === 0 || isObject(source) && Object.keys(source).length === 0) {
        let result = target;
        for (const branch of anyOf) {
          result = omit3(branch, source, result);
        }
        return result;
      }
      return handleOneOf(anyOf, childSchema, source, target);
    }
    function handleDependencies(childSchema, source, target) {
      const { dependencies } = childSchema;
      if (dependencies === void 0 || !isObjectValue(source)) {
        return target;
      }
      let result = target;
      for (const [key, deps] of Object.entries(dependencies)) {
        if (key in source && !Array.isArray(deps)) {
          result = omit3(deps, source, result);
        }
      }
      return result;
    }
    function omit3(schemaDef, source, target) {
      if (source === void 0 || schemaDef === false) {
        return void 0;
      }
      if (schemaDef === true || isEmpty4(schemaDef)) {
        return source;
      }
      let localSchema = schemaDef;
      const { $ref: ref, allOf } = localSchema;
      if (ref !== void 0) {
        return omit3(findSchemaDefinition(ref, rootSchema), source, target);
      }
      if (allOf) {
        localSchema = doMergeAllOf(localSchema, experimental_customMergeAllOf);
      }
      let filtered = handleAnyOf(localSchema, source, handleOneOf(localSchema.oneOf, localSchema, source, target));
      const type = getSchemaType(localSchema);
      if (type === "object") {
        if (!isObjectValue(source)) {
          return void 0;
        }
        filtered = handleObject(localSchema, source, isObjectValue(filtered) ? filtered : {});
      } else if (type === "array") {
        if (!Array.isArray(source)) {
          return void 0;
        }
        filtered = handleArray(localSchema, source, Array.isArray(filtered) ? filtered : []);
      } else if (filtered === void 0) {
        filtered = source;
      }
      return handleDependencies(localSchema, source, handleConditions(localSchema, source, filtered));
    }
    return omit3(schema, formData);
  }
  var NO_VALUE = /* @__PURE__ */ Symbol("no Value");
  function enumValuesForSchema(schema) {
    if (Array.isArray(schema.enum)) {
      return schema.enum;
    }
    const options = schema.oneOf || schema.anyOf;
    if (!Array.isArray(options)) {
      return void 0;
    }
    const values = options.map((option) => {
      if (CONST_KEY in option) {
        return option[CONST_KEY];
      }
      return Array.isArray(option.enum) && option.enum.length === 1 ? option.enum[0] : NO_VALUE;
    }).filter((value) => value !== NO_VALUE);
    return values.length > 0 ? values : void 0;
  }
  function replacementForInvalidEnumValue(schema, formValue) {
    const enumValues = enumValuesForSchema(schema);
    if (!enumValues || enumValues.some((value) => deepEquals_default(value, formValue))) {
      return NO_VALUE;
    }
    const defaultValue = get15(schema, DEFAULT_KEY, NO_VALUE);
    if (defaultValue !== NO_VALUE && enumValues.some((value) => deepEquals_default(value, defaultValue))) {
      return defaultValue;
    }
    return enumValues.length === 1 ? enumValues[0] : void 0;
  }
  function sanitizeDataForNewSchema(validator, rootSchema, newSchema, oldSchema, data = {}, experimental_customMergeAllOf) {
    let newFormData;
    if (has5(newSchema, PROPERTIES_KEY)) {
      const removeOldSchemaData = {};
      if (has5(oldSchema, PROPERTIES_KEY)) {
        const properties = get15(oldSchema, PROPERTIES_KEY, {});
        Object.keys(properties).forEach((key) => {
          if (has5(data, key)) {
            removeOldSchemaData[key] = void 0;
          }
        });
      }
      const keys2 = Object.keys(get15(newSchema, PROPERTIES_KEY, {}));
      const nestedData = {};
      keys2.forEach((key) => {
        const formValue = get15(data, key);
        let oldKeyedSchema = get15(oldSchema, [PROPERTIES_KEY, key], {});
        let newKeyedSchema = get15(newSchema, [PROPERTIES_KEY, key], {});
        if (has5(oldKeyedSchema, REF_KEY)) {
          oldKeyedSchema = retrieveSchema(
            validator,
            oldKeyedSchema,
            rootSchema,
            formValue,
            experimental_customMergeAllOf
          );
        }
        if (has5(newKeyedSchema, REF_KEY)) {
          newKeyedSchema = retrieveSchema(
            validator,
            newKeyedSchema,
            rootSchema,
            formValue,
            experimental_customMergeAllOf
          );
        }
        const oldSchemaTypeForKey = get15(oldKeyedSchema, "type");
        const newSchemaTypeForKey = get15(newKeyedSchema, "type");
        if (!oldSchemaTypeForKey || oldSchemaTypeForKey === newSchemaTypeForKey) {
          if (has5(removeOldSchemaData, key)) {
            delete removeOldSchemaData[key];
          }
          if (newSchemaTypeForKey === "object" || newSchemaTypeForKey === "array" && Array.isArray(formValue)) {
            const itemData = sanitizeDataForNewSchema(
              validator,
              rootSchema,
              newKeyedSchema,
              oldKeyedSchema,
              formValue,
              experimental_customMergeAllOf
            );
            if (itemData !== void 0 || newSchemaTypeForKey === "array") {
              nestedData[key] = itemData;
            }
          } else {
            const newOptionDefault = get15(newKeyedSchema, DEFAULT_KEY, NO_VALUE);
            const oldOptionDefault = get15(oldKeyedSchema, DEFAULT_KEY, NO_VALUE);
            if (newOptionDefault !== NO_VALUE && newOptionDefault !== formValue) {
              if (oldOptionDefault === formValue) {
                removeOldSchemaData[key] = newOptionDefault;
              } else if (get15(newKeyedSchema, "readOnly") === true) {
                removeOldSchemaData[key] = void 0;
              }
            }
            const newOptionConst = get15(newKeyedSchema, CONST_KEY, NO_VALUE);
            const oldOptionConst = get15(oldKeyedSchema, CONST_KEY, NO_VALUE);
            if (newOptionConst !== NO_VALUE && newOptionConst !== formValue) {
              removeOldSchemaData[key] = oldOptionConst === formValue ? newOptionConst : void 0;
            }
            if (has5(data, key)) {
              const enumReplacement = replacementForInvalidEnumValue(newKeyedSchema, formValue);
              if (enumReplacement !== NO_VALUE) {
                removeOldSchemaData[key] = enumReplacement;
              }
            }
          }
        }
      });
      newFormData = {
        ...typeof data === "string" || Array.isArray(data) ? void 0 : data,
        ...removeOldSchemaData,
        ...nestedData
      };
    } else if (get15(oldSchema, "type") === "array" && get15(newSchema, "type") === "array" && Array.isArray(data)) {
      let oldSchemaItems = get15(oldSchema, "items");
      let newSchemaItems = get15(newSchema, "items");
      if (typeof oldSchemaItems === "object" && typeof newSchemaItems === "object" && !Array.isArray(oldSchemaItems) && !Array.isArray(newSchemaItems)) {
        if (has5(oldSchemaItems, REF_KEY)) {
          oldSchemaItems = retrieveSchema(
            validator,
            oldSchemaItems,
            rootSchema,
            data,
            experimental_customMergeAllOf
          );
        }
        if (has5(newSchemaItems, REF_KEY)) {
          newSchemaItems = retrieveSchema(
            validator,
            newSchemaItems,
            rootSchema,
            data,
            experimental_customMergeAllOf
          );
        }
        const oldSchemaType = get15(oldSchemaItems, "type");
        const newSchemaType = get15(newSchemaItems, "type");
        if (!oldSchemaType || oldSchemaType === newSchemaType) {
          const maxItems = get15(newSchema, "maxItems", -1);
          if (newSchemaType === "object") {
            newFormData = data.reduce((newValue, aValue) => {
              const itemValue = sanitizeDataForNewSchema(
                validator,
                rootSchema,
                newSchemaItems,
                oldSchemaItems,
                aValue,
                experimental_customMergeAllOf
              );
              if (itemValue !== void 0 && (maxItems < 0 || newValue.length < maxItems)) {
                newValue.push(itemValue);
              }
              return newValue;
            }, []);
          } else {
            const newItemEnumValues = enumValuesForSchema(newSchemaItems);
            const filteredData = newItemEnumValues ? data.filter((item) => newItemEnumValues.some((v) => deepEquals_default(v, item))) : data;
            newFormData = maxItems > 0 && filteredData.length > maxItems ? filteredData.slice(0, maxItems) : filteredData;
          }
        }
      } else if (typeof oldSchemaItems === "boolean" && typeof newSchemaItems === "boolean" && oldSchemaItems === newSchemaItems) {
        newFormData = data;
      }
    }
    return newFormData;
  }
  function toPathSchemaInternal(validator, schema, name, rootSchema, formData, _recurseList = [], experimental_customMergeAllOf) {
    if (REF_KEY in schema || DEPENDENCIES_KEY in schema || ALL_OF_KEY in schema || IF_KEY in schema) {
      const innerSchema = retrieveSchema(validator, schema, rootSchema, formData, experimental_customMergeAllOf);
      const sameSchemaIndex = _recurseList.findIndex((item) => deepEquals_default(item, innerSchema));
      if (sameSchemaIndex === -1) {
        return toPathSchemaInternal(
          validator,
          innerSchema,
          name,
          rootSchema,
          formData,
          _recurseList.concat(innerSchema),
          experimental_customMergeAllOf
        );
      }
    }
    let pathSchema = {
      [NAME_KEY]: name.replace(/^\./, "")
    };
    if (ONE_OF_KEY in schema || ANY_OF_KEY in schema) {
      const xxxOf = ONE_OF_KEY in schema ? schema.oneOf : schema.anyOf;
      const discriminator = getDiscriminatorFieldFromSchema(schema);
      const index = getClosestMatchingOption(
        validator,
        rootSchema,
        formData,
        xxxOf,
        0,
        discriminator,
        experimental_customMergeAllOf
      );
      const innerSchema = xxxOf[index];
      pathSchema = {
        ...pathSchema,
        ...toPathSchemaInternal(
          validator,
          innerSchema,
          name,
          rootSchema,
          formData,
          _recurseList,
          experimental_customMergeAllOf
        )
      };
    }
    if (ADDITIONAL_PROPERTIES_KEY in schema && schema[ADDITIONAL_PROPERTIES_KEY] !== false) {
      set(pathSchema, RJSF_ADDITIONAL_PROPERTIES_FLAG, true);
      const additionalSchema = isObject2(schema[ADDITIONAL_PROPERTIES_KEY]) ? schema[ADDITIONAL_PROPERTIES_KEY] : {};
      const definedProperties = get15(schema, PROPERTIES_KEY, {});
      for (const key of Object.keys(formData ?? {})) {
        if (!(key in definedProperties)) {
          pathSchema[key] = toPathSchemaInternal(
            validator,
            additionalSchema,
            `${name}.${key}`,
            rootSchema,
            get15(formData, [key]),
            _recurseList,
            experimental_customMergeAllOf
          );
        }
      }
    }
    if (ITEMS_KEY in schema && Array.isArray(formData)) {
      const { items: schemaItems, additionalItems: schemaAdditionalItems } = schema;
      if (Array.isArray(schemaItems)) {
        formData.forEach((element, i) => {
          if (schemaItems[i]) {
            pathSchema[i] = toPathSchemaInternal(
              validator,
              schemaItems[i],
              `${name}.${i}`,
              rootSchema,
              element,
              _recurseList,
              experimental_customMergeAllOf
            );
          } else if (schemaAdditionalItems) {
            pathSchema[i] = toPathSchemaInternal(
              validator,
              schemaAdditionalItems,
              `${name}.${i}`,
              rootSchema,
              element,
              _recurseList,
              experimental_customMergeAllOf
            );
          } else {
            console.warn(`Unable to generate path schema for "${name}.${i}". No schema defined for it`);
          }
        });
      } else {
        formData.forEach((element, i) => {
          pathSchema[i] = toPathSchemaInternal(
            validator,
            schemaItems,
            `${name}.${i}`,
            rootSchema,
            element,
            _recurseList,
            experimental_customMergeAllOf
          );
        });
      }
    } else if (PROPERTIES_KEY in schema) {
      for (const property in schema.properties) {
        const field = get15(schema, [PROPERTIES_KEY, property], {});
        pathSchema[property] = toPathSchemaInternal(
          validator,
          field,
          `${name}.${property}`,
          rootSchema,
          // It's possible that formData is not an object -- this can happen if an
          // array item has just been added, but not populated with data yet
          get15(formData, [property]),
          _recurseList,
          experimental_customMergeAllOf
        );
      }
    }
    return pathSchema;
  }
  function toPathSchema(validator, schema, name = "", rootSchema, formData, experimental_customMergeAllOf) {
    return toPathSchemaInternal(validator, schema, name, rootSchema, formData, void 0, experimental_customMergeAllOf);
  }

  // src/createSchemaUtils.ts
  var SchemaUtils = class {
    /** Constructs the `SchemaUtils` instance with the given `validator` and `rootSchema` stored as instance variables
     *
     * @param validator - An implementation of the `ValidatorType` interface that will be forwarded to all the APIs
     * @param rootSchema - The root schema that will be forwarded to all the APIs
     * @param experimental_defaultFormStateBehavior - Configuration flags to allow users to override default form state behavior
     * @param [experimental_customMergeAllOf] - Optional function that allows for custom merging of `allOf` schemas
     */
    constructor(validator, rootSchema, experimental_defaultFormStateBehavior, experimental_customMergeAllOf) {
      if (rootSchema && rootSchema[SCHEMA_KEY] === JSON_SCHEMA_DRAFT_2020_12) {
        this.rootSchema = makeAllReferencesAbsolute(rootSchema, get15(rootSchema, ID_KEY, "#"));
      } else {
        this.rootSchema = rootSchema;
      }
      this.validator = validator;
      this.experimental_defaultFormStateBehavior = experimental_defaultFormStateBehavior;
      this.experimental_customMergeAllOf = experimental_customMergeAllOf;
    }
    /** Returns the `rootSchema` in the `SchemaUtilsType`
     *
     * @returns - The `rootSchema`
     */
    getRootSchema() {
      return this.rootSchema;
    }
    /** Returns the `ValidatorType` in the `SchemaUtilsType`
     *
     * @returns - The `ValidatorType`
     */
    getValidator() {
      return this.validator;
    }
    /** Determines whether either the `validator` and `rootSchema` differ from the ones associated with this instance of
     * the `SchemaUtilsType`. If either `validator` or `rootSchema` are falsy, then return false to prevent the creation
     * of a new `SchemaUtilsType` with incomplete properties.
     *
     * @param validator - An implementation of the `ValidatorType` interface that will be compared against the current one
     * @param rootSchema - The root schema that will be compared against the current one
     * @param [experimental_defaultFormStateBehavior] Optional configuration object, if provided, allows users to override default form state behavior
     * @param [experimental_customMergeAllOf] - Optional function that allows for custom merging of `allOf` schemas
     * @returns - True if the `SchemaUtilsType` differs from the given `validator` or `rootSchema`
     */
    doesSchemaUtilsDiffer(validator, rootSchema, experimental_defaultFormStateBehavior = {}, experimental_customMergeAllOf) {
      if (!validator || !rootSchema) {
        return false;
      }
      return this.validator !== validator || !deepEquals_default(this.rootSchema, rootSchema) || !deepEquals_default(this.experimental_defaultFormStateBehavior, experimental_defaultFormStateBehavior) || this.experimental_customMergeAllOf !== experimental_customMergeAllOf;
    }
    /** Finds the field specified by the `path` within the root or recursed `schema`. If there is no field for the specified
     * `path`, then the default `{ field: undefined, isRequired: undefined }` is returned. It determines whether a leaf
     * field is in the `required` list for its parent and if so, it is marked as required on return.
     *
     * @param schema - The current node within the JSON schema
     * @param path - The remaining keys in the path to the desired field
     * @param [formData] - The form data that is used to determine which oneOf option
     * @returns - An object that contains the field and its required state. If no field can be found then
     *            `{ field: undefined, isRequired: undefined }` is returned.
     */
    findFieldInSchema(schema, path, formData) {
      return findFieldInSchema(
        this.validator,
        this.rootSchema,
        schema,
        path,
        formData,
        this.experimental_customMergeAllOf
      );
    }
    /** Finds the oneOf option inside the `schema['any/oneOf']` list which has the `properties[selectorField].default` that
     * matches the `formData[selectorField]` value. For the purposes of this function, `selectorField` is either
     * `schema.discriminator.propertyName` or `fallbackField`.
     *
     * @param schema - The schema element in which to search for the selected oneOf option
     * @param fallbackField - The field to use as a backup selector field if the schema does not have a required field
     * @param xxx - Either `oneOf` or `anyOf`, defines which value is being sought
     * @param [formData={}] - The form data that is used to determine which oneOf option
     * @returns - The anyOf/oneOf option that matches the selector field in the schema or undefined if nothing is selected
     */
    findSelectedOptionInXxxOf(schema, fallbackField, xxx, formData) {
      return findSelectedOptionInXxxOf(
        this.validator,
        this.rootSchema,
        schema,
        fallbackField,
        xxx,
        formData,
        this.experimental_customMergeAllOf
      );
    }
    /** Returns the superset of `formData` that includes the given set updated to include any missing fields that have
     * computed to have defaults provided in the `schema`.
     *
     * @param schema - The schema for which the default state is desired
     * @param [formData] - The current formData, if any, onto which to provide any missing defaults
     * @param [includeUndefinedValues=false] - Optional flag, if true, cause undefined values to be added as defaults.
     *          If "excludeObjectChildren", pass `includeUndefinedValues` as false when computing defaults for any nested
     *          object properties.
     * @param initialDefaultsGenerated - Indicates whether or not initial defaults have been generated
     * @returns - The resulting `formData` with all the defaults provided
     */
    getDefaultFormState(schema, formData, includeUndefinedValues = false, initialDefaultsGenerated) {
      return getDefaultFormState(
        this.validator,
        schema,
        formData,
        this.rootSchema,
        includeUndefinedValues,
        this.experimental_defaultFormStateBehavior,
        this.experimental_customMergeAllOf,
        initialDefaultsGenerated
      );
    }
    /** Determines whether the combination of `schema` and `uiSchema` properties indicates that the label for the `schema`
     * should be displayed in a UI.
     *
     * @param schema - The schema for which the display label flag is desired
     * @param [uiSchema] - The UI schema from which to derive potentially displayable information
     * @param [globalOptions={}] - The optional Global UI Schema from which to get any fallback `xxx` options
     * @returns - True if the label should be displayed or false if it should not
     */
    getDisplayLabel(schema, uiSchema, globalOptions) {
      return getDisplayLabel(
        this.validator,
        schema,
        uiSchema,
        this.rootSchema,
        globalOptions,
        this.experimental_customMergeAllOf
      );
    }
    /** Determines which of the given `options` provided most closely matches the `formData`.
     * Returns the index of the option that is valid and is the closest match, or 0 if there is no match.
     *
     * The closest match is determined using the number of matching properties, and more heavily favors options with
     * matching readOnly, default, or const values.
     *
     * @param formData - The form data associated with the schema
     * @param options - The list of options that can be selected from
     * @param [selectedOption] - The index of the currently selected option, defaulted to -1 if not specified
     * @param [discriminatorField] - The optional name of the field within the options object whose value is used to
     *          determine which option is selected
     * @returns - The index of the option that is the closest match to the `formData` or the `selectedOption` if no match
     */
    getClosestMatchingOption(formData, options, selectedOption, discriminatorField) {
      return getClosestMatchingOption(
        this.validator,
        this.rootSchema,
        formData,
        options,
        selectedOption,
        discriminatorField,
        this.experimental_customMergeAllOf
      );
    }
    /** Given the `formData` and list of `options`, attempts to find the index of the first option that matches the data.
     * Always returns the first option if there is nothing that matches.
     *
     * @param formData - The current formData, if any, used to figure out a match
     * @param options - The list of options to find a matching options from
     * @param [discriminatorField] - The optional name of the field within the options object whose value is used to
     *          determine which option is selected
     * @returns - The firstindex of the matched option or 0 if none is available
     */
    getFirstMatchingOption(formData, options, discriminatorField) {
      return getFirstMatchingOption(this.validator, formData, options, this.rootSchema, discriminatorField);
    }
    getFromSchema(schema, path, defaultValue) {
      return getFromSchema(
        this.validator,
        this.rootSchema,
        schema,
        path,
        // @ts-expect-error TS2769: No overload matches this call
        defaultValue,
        this.experimental_customMergeAllOf
      );
    }
    /** Checks to see if the `schema` and `uiSchema` combination represents an array of files
     *
     * @param schema - The schema for which check for array of files flag is desired
     * @param [uiSchema] - The UI schema from which to check the widget
     * @returns - True if schema/uiSchema contains an array of files, otherwise false
     */
    isFilesArray(schema, uiSchema) {
      return isFilesArray(this.validator, schema, uiSchema, this.rootSchema, this.experimental_customMergeAllOf);
    }
    /** Checks to see if the `schema` combination represents a multi-select
     *
     * @param schema - The schema for which check for a multi-select flag is desired
     * @returns - True if schema contains a multi-select, otherwise false
     */
    isMultiSelect(schema) {
      return isMultiSelect(this.validator, schema, this.rootSchema, this.experimental_customMergeAllOf);
    }
    /** Checks to see if the `schema` combination represents a select
     *
     * @param schema - The schema for which check for a select flag is desired
     * @returns - True if schema contains a select, otherwise false
     */
    isSelect(schema) {
      return isSelect(this.validator, schema, this.rootSchema, this.experimental_customMergeAllOf);
    }
    /**
     * The function takes a `schema` and `formData` and returns a copy of the formData with any fields not defined in the schema removed.
     * This is useful for ensuring that only data that is relevant to the schema is preserved. Objects with `additionalProperties`
     * keyword set to `true` will not have their extra fields removed.
     *
     * @param schema - The schema to use for filtering the `formData`
     * @param [formData] - The formData to filter
     * @returns The new form data, with any fields not defined in the schema removed
     */
    omitExtraData(schema, formData) {
      return omitExtraData(this.validator, schema, this.rootSchema, formData);
    }
    /** Retrieves an expanded schema that has had all of its conditions, additional properties, references and
     * dependencies resolved and merged into the `schema` given a `rawFormData` that is used to do the potentially
     * recursive resolution.
     *
     * @param schema - The schema for which retrieving a schema is desired
     * @param [rawFormData] - The current formData, if any, to assist retrieving a schema
     * @param [resolveAnyOfOrOneOfRefs] - Optional flag indicating whether to resolved refs in anyOf/oneOf lists
     * @returns - The schema having its conditions, additional properties, references and dependencies resolved
     */
    retrieveSchema(schema, rawFormData, resolveAnyOfOrOneOfRefs) {
      return retrieveSchema(
        this.validator,
        schema,
        this.rootSchema,
        rawFormData,
        this.experimental_customMergeAllOf,
        resolveAnyOfOrOneOfRefs
      );
    }
    /** Sanitize the `data` associated with the `oldSchema` so it is considered appropriate for the `newSchema`. If the
     * new schema does not contain any properties, then `undefined` is returned to clear all the form data. Due to the
     * nature of schemas, this sanitization happens recursively for nested objects of data. Also, any properties in the
     * old schemas that are non-existent in the new schema are set to `undefined`.
     *
     * @param [newSchema] - The new schema for which the data is being sanitized
     * @param [oldSchema] - The old schema from which the data originated
     * @param [data={}] - The form data associated with the schema, defaulting to an empty object when undefined
     * @returns - The new form data, with all the fields uniquely associated with the old schema set
     *      to `undefined`. Will return `undefined` if the new schema is not an object containing properties.
     */
    sanitizeDataForNewSchema(newSchema, oldSchema, data) {
      return sanitizeDataForNewSchema(
        this.validator,
        this.rootSchema,
        newSchema,
        oldSchema,
        data,
        this.experimental_customMergeAllOf
      );
    }
    /** Generates an `PathSchema` object for the `schema`, recursively
     *
     * @param schema - The schema for which the display label flag is desired
     * @param [name] - The base name for the schema
     * @param [formData] - The current formData, if any, onto which to provide any missing defaults
     * @returns - The `PathSchema` object for the `schema`
     */
    // oxlint-disable-next-line typescript/no-deprecated
    toPathSchema(schema, name, formData) {
      return toPathSchema(
        this.validator,
        schema,
        name,
        this.rootSchema,
        formData,
        this.experimental_customMergeAllOf
      );
    }
  };
  function createSchemaUtils(validator, rootSchema, experimental_defaultFormStateBehavior = {}, experimental_customMergeAllOf) {
    return new SchemaUtils(
      validator,
      rootSchema,
      experimental_defaultFormStateBehavior,
      experimental_customMergeAllOf
    );
  }

  // src/dataURItoBlob.ts
  function dataURItoBlob(dataURILike) {
    if (!dataURILike.includes("data:")) {
      throw new Error("File is invalid: URI must be a dataURI");
    }
    const dataURI = dataURILike.slice(5);
    const splitted = dataURI.split(";base64,");
    if (splitted.length !== 2) {
      throw new Error("File is invalid: dataURI must be base64");
    }
    const [media, base64] = splitted;
    const [mime, ...mediaparams] = media.split(";");
    const type = mime || "";
    const name = decodeURI(
      // parse the parameters into key-value pairs, find a key, and extract a value
      // if no key is found, then the name is unknown
      mediaparams.map((param) => param.split("=")).find(([key]) => key === "name")?.[1] || "unknown"
    );
    try {
      const binary = atob(base64);
      const array = new Array(binary.length);
      for (let i = 0; i < binary.length; i += 1) {
        array[i] = binary.charCodeAt(i);
      }
      const blob = new window.Blob([new Uint8Array(array)], { type });
      return { blob, name };
    } catch (error) {
      throw new Error(`File is invalid: ${error.message}`);
    }
  }

  // src/pad.ts
  function pad(num, width) {
    let s = String(num);
    while (s.length < width) {
      s = `0${s}`;
    }
    return s;
  }

  // src/dateRangeOptions.ts
  function dateRangeOptions(start, stop) {
    const bothRelative = start <= 0 && stop <= 0;
    const eitherNegative = start < 0 || stop < 0;
    if (!bothRelative && eitherNegative) {
      throw new Error(`Both start (${start}) and stop (${stop}) must both be <= 0 or > 0, got one of each`);
    }
    const currentYear = (/* @__PURE__ */ new Date()).getFullYear();
    const resolvedStart = bothRelative ? currentYear + start : start;
    const resolvedStop = bothRelative ? currentYear + stop : stop;
    if (resolvedStart > resolvedStop) {
      return dateRangeOptions(stop, start).reverse();
    }
    const options = [];
    for (let i = resolvedStart; i <= resolvedStop; i += 1) {
      options.push({ value: i, label: pad(i, 2) });
    }
    return options;
  }

  // src/replaceStringParameters.ts
  function replaceStringParameters(inputString, params) {
    let output = inputString;
    if (Array.isArray(params)) {
      const parts = output.split(/(%\d)/);
      params.forEach((param, index) => {
        const partIndex = parts.findIndex((part) => part === `%${index + 1}`);
        if (partIndex >= 0) {
          parts[partIndex] = param;
        }
      });
      output = parts.join("");
    }
    return output;
  }

  // src/englishStringTranslator.ts
  function englishStringTranslator(stringToTranslate, params) {
    return replaceStringParameters(stringToTranslate, params);
  }

  // src/enumOptionsValueForIndex.ts
  function enumOptionsValueForIndex(valueIndex, allEnumOptions = [], emptyValue) {
    if (Array.isArray(valueIndex)) {
      return valueIndex.map((index2) => enumOptionsValueForIndex(index2, allEnumOptions)).filter((val) => val !== void 0);
    }
    const index = valueIndex === "" || valueIndex === null ? -1 : Number(valueIndex);
    const option = allEnumOptions[index];
    return option ? option.value : emptyValue;
  }

  // src/enumOptionsDeselectValue.ts
  function enumOptionsDeselectValue(valueIndex, selected, allEnumOptions = []) {
    const value = enumOptionsValueForIndex(valueIndex, allEnumOptions);
    if (Array.isArray(selected)) {
      return selected.filter((v) => !deepEquals_default(v, value));
    }
    return deepEquals_default(value, selected) ? void 0 : selected;
  }

  // src/enumOptionsIsSelected.ts
  function enumOptionsIsSelected(value, selected) {
    if (Array.isArray(selected)) {
      return selected.some((sel) => deepEquals_default(sel, value));
    }
    return deepEquals_default(selected, value);
  }

  // src/enumOptionsIndexForValue.ts
  function enumOptionsIndexForValue(value, allEnumOptions = [], multiple = false) {
    const selectedIndexes = allEnumOptions.map((opt, index) => enumOptionsIsSelected(opt.value, value) ? String(index) : void 0).filter((opt) => typeof opt !== "undefined");
    if (!multiple) {
      return selectedIndexes[0];
    }
    return selectedIndexes;
  }

  // src/enumOptionSelectedValue.ts
  function enumOptionSelectedValue(value, enumOptions, multiple, format = "indexed", emptyValue) {
    const isEmpty9 = typeof value === "undefined" || multiple && Array.isArray(value) && value.length < 1 || !multiple && value === emptyValue;
    if (isEmpty9) {
      return emptyValue;
    }
    if (format === "realValue") {
      return multiple ? value.map(String) : String(value);
    }
    const indexes = enumOptionsIndexForValue(value, enumOptions, multiple);
    return typeof indexes === "undefined" ? emptyValue : indexes;
  }
  function enumOptionsSelectValue(valueIndex, selected, allEnumOptions = []) {
    const value = enumOptionsValueForIndex(valueIndex, allEnumOptions);
    if (!isNil(value)) {
      const index = allEnumOptions.findIndex((opt) => value === opt.value);
      const all = allEnumOptions.map(({ value: val }) => val);
      const updated = selected.slice(0, index).concat(value, selected.slice(index));
      return updated.sort((a, b) => Number(all.indexOf(a) > all.indexOf(b)));
    }
    return selected;
  }

  // src/enumOptionValueDecoder.ts
  function decodeSingle(value, enumOptions, emptyValue) {
    if (value === "" || !Array.isArray(enumOptions)) {
      return emptyValue;
    }
    const match = enumOptions.find((opt) => String(opt.value) === value);
    if (match) {
      return match.value;
    }
    const index = Number(value);
    if (!Number.isNaN(index) && index >= 0 && index < enumOptions.length) {
      return enumOptions[index].value;
    }
    return emptyValue;
  }
  function enumOptionValueDecoder(value, enumOptions, format = "indexed", emptyValue) {
    if (format !== "realValue") {
      return enumOptionsValueForIndex(value, enumOptions, emptyValue);
    }
    if (Array.isArray(value)) {
      return value.map((v) => decodeSingle(v, enumOptions, emptyValue));
    }
    return decodeSingle(value, enumOptions, emptyValue);
  }
  function enumOptionValueEncoder(value, index, format = "indexed") {
    if (format !== "realValue") {
      return String(index);
    }
    if (isNil(value)) {
      return "";
    }
    if (typeof value === "object") {
      return String(index);
    }
    return String(value);
  }
  var ErrorSchemaBuilder = class {
    /** Construct an `ErrorSchemaBuilder` with an optional initial set of errors in an `ErrorSchema`.
     *
     * @param [initialSchema] - The optional set of initial errors, that will be cloned into the class
     */
    constructor(initialSchema) {
      /** The error schema being built
       *
       * @private
       */
      this.errorSchema = {};
      this.resetAllErrors(initialSchema);
    }
    /** Returns the `ErrorSchema` that has been updated by the methods of the `ErrorSchemaBuilder`
     */
    get ErrorSchema() {
      return this.errorSchema;
    }
    /** Will get an existing `ErrorSchema` at the specified `pathOfError` or create and return one.
     *
     * @param [pathOfError] - The optional path into the `ErrorSchema` at which to add the error(s)
     * @returns - The error block for the given `pathOfError` or the root if not provided
     * @private
     */
    getOrCreateErrorBlock(pathOfError) {
      const hasPath = Array.isArray(pathOfError) && pathOfError.length > 0 || typeof pathOfError === "string";
      let errorBlock = hasPath ? get15(this.errorSchema, pathOfError) : this.errorSchema;
      if (!errorBlock && pathOfError) {
        errorBlock = {};
        setWith(this.errorSchema, pathOfError, errorBlock, Object);
      }
      return errorBlock;
    }
    /** Resets all errors in the `ErrorSchemaBuilder` back to the `initialSchema` if provided, otherwise an empty set.
     *
     * @param [initialSchema] - The optional set of initial errors, that will be cloned into the class
     * @returns - The `ErrorSchemaBuilder` object for chaining purposes
     */
    resetAllErrors(initialSchema) {
      this.errorSchema = initialSchema ? cloneDeep(initialSchema) : {};
      return this;
    }
    /** Adds the `errorOrList` to the list of errors in the `ErrorSchema` at either the root level or the location within
     * the schema described by the `pathOfError`. For more information about how to specify the path see the
     * [eslint lodash plugin docs](https://github.com/wix/eslint-plugin-lodash/blob/master/docs/rules/path-style.md).
     *
     * @param errorOrList - The error or list of errors to add into the `ErrorSchema`
     * @param [pathOfError] - The optional path into the `ErrorSchema` at which to add the error(s)
     * @returns - The `ErrorSchemaBuilder` object for chaining purposes
     */
    addErrors(errorOrList, pathOfError) {
      const errorBlock = this.getOrCreateErrorBlock(pathOfError);
      let errorsList = get15(errorBlock, ERRORS_KEY);
      if (!Array.isArray(errorsList)) {
        errorsList = [];
        errorBlock[ERRORS_KEY] = errorsList;
      }
      if (Array.isArray(errorOrList)) {
        set(errorBlock, ERRORS_KEY, [.../* @__PURE__ */ new Set([...errorsList, ...errorOrList])]);
      } else {
        set(errorBlock, ERRORS_KEY, [.../* @__PURE__ */ new Set([...errorsList, errorOrList])]);
      }
      return this;
    }
    /** Sets/replaces the `errorOrList` as the error(s) in the `ErrorSchema` at either the root level or the location
     * within the schema described by the `pathOfError`. For more information about how to specify the path see the
     * [eslint lodash plugin docs](https://github.com/wix/eslint-plugin-lodash/blob/master/docs/rules/path-style.md).
     *
     * @param errorOrList - The error or list of errors to set into the `ErrorSchema`
     * @param [pathOfError] - The optional path into the `ErrorSchema` at which to set the error(s)
     * @returns - The `ErrorSchemaBuilder` object for chaining purposes
     */
    setErrors(errorOrList, pathOfError) {
      const errorBlock = this.getOrCreateErrorBlock(pathOfError);
      const listToAdd = Array.isArray(errorOrList) ? [.../* @__PURE__ */ new Set([...errorOrList])] : [errorOrList];
      set(errorBlock, ERRORS_KEY, listToAdd);
      return this;
    }
    /** Clears the error(s) in the `ErrorSchema` at either the root level or the location within the schema described by
     * the `pathOfError`. For more information about how to specify the path see the
     * [eslint lodash plugin docs](https://github.com/wix/eslint-plugin-lodash/blob/master/docs/rules/path-style.md).
     *
     * @param [pathOfError] - The optional path into the `ErrorSchema` at which to clear the error(s)
     * @returns - The `ErrorSchemaBuilder` object for chaining purposes
     */
    clearErrors(pathOfError) {
      const errorBlock = this.getOrCreateErrorBlock(pathOfError);
      set(errorBlock, ERRORS_KEY, []);
      return this;
    }
  };
  function getChangedFields(a, b) {
    const aIsPlainObject = isPlainObject2(a);
    const bIsPlainObject = isPlainObject2(b);
    if (a === b || !aIsPlainObject && !bIsPlainObject) {
      return [];
    }
    if (aIsPlainObject && !bIsPlainObject) {
      return keys(a);
    }
    if (!aIsPlainObject && bIsPlainObject) {
      return keys(b);
    }
    const unequalFields = keys(pickBy(a, (value, key) => !deepEquals_default(value, get15(b, key))));
    const diffFields = difference(keys(b), keys(a));
    return [...unequalFields, ...diffFields];
  }

  // src/getDateElementProps.ts
  function getDateElementProps(date, time, yearRange = [1900, (/* @__PURE__ */ new Date()).getFullYear() + 2], format = "YMD") {
    const { day, month, year, hour, minute, second } = date;
    const dayObj = { type: "day", range: [1, 31], value: day };
    const monthObj = { type: "month", range: [1, 12], value: month };
    const yearObj = { type: "year", range: yearRange, value: year };
    const dateElementProp = [];
    switch (format) {
      case "MDY":
        dateElementProp.push(monthObj, dayObj, yearObj);
        break;
      case "DMY":
        dateElementProp.push(dayObj, monthObj, yearObj);
        break;
      case "YMD":
      default:
        dateElementProp.push(yearObj, monthObj, dayObj);
    }
    if (time) {
      dateElementProp.push(
        { type: "hour", range: [0, 23], value: hour },
        { type: "minute", range: [0, 59], value: minute },
        { type: "second", range: [0, 59], value: second }
      );
    }
    return dateElementProp;
  }

  // src/rangeSpec.ts
  function rangeSpec(schema) {
    const spec = {};
    if (schema.multipleOf) {
      spec.step = schema.multipleOf;
    }
    if (schema.minimum || schema.minimum === 0) {
      spec.min = schema.minimum;
    }
    if (schema.maximum || schema.maximum === 0) {
      spec.max = schema.maximum;
    }
    return spec;
  }

  // src/getInputProps.ts
  function getInputProps(schema, defaultType, options = {}, autoDefaultStepAny = true) {
    const inputProps = {
      type: defaultType || "text",
      ...rangeSpec(schema)
    };
    if (options.inputType) {
      inputProps.type = options.inputType;
    } else if (!defaultType) {
      if (schema.type === "number") {
        inputProps.type = "number";
        if (autoDefaultStepAny && inputProps.step === void 0) {
          inputProps.step = "any";
        }
      } else if (schema.type === "integer") {
        inputProps.type = "number";
        if (inputProps.step === void 0) {
          inputProps.step = 1;
        }
      }
    }
    if (["date", "datetime-local", "time", "week", "month"].includes(inputProps.type)) {
      if (schema.formatMinimum !== void 0) {
        inputProps.min = schema.formatMinimum;
      }
      if (schema.formatMaximum !== void 0) {
        inputProps.max = schema.formatMaximum;
      }
    }
    if (options.autocomplete) {
      inputProps.autoComplete = options.autocomplete;
    }
    if (options.accept) {
      inputProps.accept = options.accept;
    }
    return inputProps;
  }

  // src/getOptionValueFormat.ts
  function getOptionValueFormat(options) {
    return options?.optionValueFormat ?? "indexed";
  }

  // src/getSubmitButtonOptions.ts
  var DEFAULT_OPTIONS = {
    props: {
      disabled: false
    },
    submitText: "Submit",
    norender: false
  };
  function getSubmitButtonOptions(uiSchema = {}) {
    const uiOptions = getUiOptions(uiSchema);
    if (uiOptions && uiOptions[SUBMIT_BTN_OPTIONS_KEY]) {
      const options = uiOptions[SUBMIT_BTN_OPTIONS_KEY];
      return { ...DEFAULT_OPTIONS, ...options };
    }
    return DEFAULT_OPTIONS;
  }

  // src/getTemplate.ts
  function getTemplate(name, registry, uiOptions = {}) {
    const { templates } = registry;
    if (name === "ButtonTemplates") {
      return templates[name];
    }
    if (Object.hasOwn(uiOptions, name) && typeof uiOptions[name] === "string" && Object.hasOwn(templates, uiOptions[name])) {
      const key = uiOptions[name];
      return templates[key];
    }
    return (
      // Evaluating uiOptions[name] results in TS2590: Expression produces a union type that is too complex to represent
      // To avoid that, we cast uiOptions to `any` before accessing the name field
      uiOptions[name] || templates[name]
    );
  }
  function getTestIds() {
    if (get15(globalThis, "process.env.NODE_ENV") !== "test") {
      return {};
    }
    const ids = /* @__PURE__ */ new Map();
    return new Proxy(
      {},
      {
        get(_obj, prop) {
          if (!ids.has(prop)) {
            ids.set(prop, uniqueId("test-id-"));
          }
          return ids.get(prop);
        }
      }
    );
  }
  var widgetMap = {
    boolean: {
      checkbox: "CheckboxWidget",
      radio: "RadioWidget",
      select: "SelectWidget",
      hidden: "HiddenWidget"
    },
    string: {
      text: "TextWidget",
      password: "PasswordWidget",
      email: "EmailWidget",
      hostname: "TextWidget",
      ipv4: "TextWidget",
      ipv6: "TextWidget",
      uri: "URLWidget",
      "data-url": "FileWidget",
      radio: "RadioWidget",
      select: "SelectWidget",
      textarea: "TextareaWidget",
      hidden: "HiddenWidget",
      date: "DateWidget",
      datetime: "DateTimeWidget",
      "date-time": "DateTimeWidget",
      "alt-date": "AltDateWidget",
      "alt-datetime": "AltDateTimeWidget",
      time: "TimeWidget",
      color: "ColorWidget",
      file: "FileWidget"
    },
    number: {
      text: "TextWidget",
      select: "SelectWidget",
      updown: "UpDownWidget",
      range: "RangeWidget",
      radio: "RadioWidget",
      hidden: "HiddenWidget"
    },
    integer: {
      text: "TextWidget",
      select: "SelectWidget",
      updown: "UpDownWidget",
      range: "RangeWidget",
      radio: "RadioWidget",
      hidden: "HiddenWidget"
    },
    array: {
      select: "SelectWidget",
      checkboxes: "CheckboxesWidget",
      files: "FileWidget",
      hidden: "HiddenWidget"
    }
  };
  function mergeWidgetOptions(AWidget) {
    let MergedWidget = get15(AWidget, "MergedWidget");
    if (!MergedWidget) {
      const defaultOptions = AWidget.defaultProps && AWidget.defaultProps.options || {};
      MergedWidget = ({ options, ...props }) => /* @__PURE__ */ jsxRuntime.jsx(AWidget, { options: { ...defaultOptions, ...options }, ...props });
      set(AWidget, "MergedWidget", MergedWidget);
    }
    return MergedWidget;
  }
  function getWidget(schema, widget, registeredWidgets = {}) {
    const type = getSchemaType(schema);
    if (typeof widget === "function" || widget && ReactIs.isForwardRef(react.createElement(widget)) || ReactIs.isMemo(widget)) {
      return mergeWidgetOptions(widget);
    }
    if (typeof widget !== "string") {
      throw new Error(`Unsupported widget definition: ${typeof widget} in schema: ${JSON.stringify(schema)}`);
    }
    if (widget in registeredWidgets) {
      const registeredWidget = registeredWidgets[widget];
      return getWidget(schema, registeredWidget, registeredWidgets);
    }
    if (typeof type === "string") {
      if (!(type in widgetMap)) {
        throw new Error(`No widget for type '${type}' in schema: ${JSON.stringify(schema)}`);
      }
      if (widget in widgetMap[type]) {
        const registeredWidget = registeredWidgets[widgetMap[type][widget]];
        return getWidget(schema, registeredWidget, registeredWidgets);
      }
    }
    throw new Error(`No widget '${widget}' for type '${type}' in schema: ${JSON.stringify(schema)}`);
  }

  // src/hashForSchema.ts
  function hashString(string) {
    let hash = 0;
    for (let i = 0; i < string.length; i++) {
      const chr = string.charCodeAt(i);
      hash = (hash << 5) - hash + chr;
      hash &= hash;
    }
    return hash.toString(16);
  }
  function sortedJSONStringify(object) {
    const allKeys = /* @__PURE__ */ new Set();
    JSON.stringify(object, (key, value) => {
      allKeys.add(key);
      return value;
    });
    return JSON.stringify(object, Array.from(allKeys).sort());
  }
  function hashObject(object) {
    return hashString(sortedJSONStringify(object));
  }
  function hashForSchema(schema) {
    return hashObject(schema);
  }

  // src/hasWidget.ts
  function hasWidget(schema, widget, registeredWidgets = {}) {
    try {
      getWidget(schema, widget, registeredWidgets);
      return true;
    } catch (e) {
      const err = e;
      if (err.message && (err.message.startsWith("No widget") || err.message.startsWith("Unsupported widget"))) {
        return false;
      }
      throw e;
    }
  }
  function idGenerator(id, suffix) {
    const theId = isString(id) ? id : id[ID_KEY];
    return `${theId}__${suffix}`;
  }
  function descriptionId(id) {
    return idGenerator(id, "description");
  }
  function errorId(id) {
    return idGenerator(id, "error");
  }
  function examplesId(id) {
    return idGenerator(id, "examples");
  }
  function helpId(id) {
    return idGenerator(id, "help");
  }
  function titleId(id) {
    return idGenerator(id, "title");
  }
  function ariaDescribedByIds(id, includeExamples = false) {
    const examples = includeExamples ? ` ${examplesId(id)}` : "";
    return `${errorId(id)} ${descriptionId(id)} ${helpId(id)}${examples}`;
  }
  function optionId(id, optionIndex) {
    return `${id}-${optionIndex}`;
  }
  function buttonId(id, btn) {
    return idGenerator(id, btn);
  }
  function optionalControlsId(id, element) {
    return idGenerator(id, `optional${element}`);
  }
  function isFormDataAvailable(formData) {
    return !isNil(formData) && (!isObject2(formData) || Array.isArray(formData) || !isEmpty4(formData));
  }
  function isRootSchema(registry, schemaToCompare) {
    const { rootSchema, schemaUtils } = registry;
    if (deepEquals_default(schemaToCompare, rootSchema)) {
      return true;
    }
    if (REF_KEY in rootSchema) {
      const resolvedSchema = schemaUtils.retrieveSchema(rootSchema);
      return deepEquals_default(schemaToCompare, omit(resolvedSchema, RJSF_REF_KEY));
    }
    return false;
  }

  // src/labelValue.ts
  function labelValue(label, hideLabel, fallback) {
    return hideLabel ? fallback : label;
  }

  // src/localToUTC.ts
  function localToUTC(dateString) {
    return dateString ? new Date(dateString).toJSON() : void 0;
  }
  function lookupFromFormContext(regOrFc, toLookup, fallback) {
    const lookupPath = [LOOKUP_MAP_NAME];
    if (has5(regOrFc, FORM_CONTEXT_NAME)) {
      lookupPath.unshift(FORM_CONTEXT_NAME);
    }
    return get15(regOrFc, [...lookupPath, toLookup], fallback);
  }

  // src/nameGenerators.ts
  var bracketNameGenerator = (path, idPrefix, isMultiValue) => {
    if (!path || path.length === 0) {
      return idPrefix;
    }
    const baseName = path.reduce((acc, pathUnit, index) => {
      if (index === 0) {
        return `${idPrefix}[${String(pathUnit)}]`;
      }
      return `${acc}[${String(pathUnit)}]`;
    }, "");
    return isMultiValue ? `${baseName}[]` : baseName;
  };
  var dotNotationNameGenerator = (path, idPrefix, _isMultiValue) => {
    if (!path || path.length === 0) {
      return idPrefix;
    }
    return `${idPrefix}.${path.map(String).join(".")}`;
  };

  // src/orderProperties.ts
  function orderProperties(properties, order) {
    if (!Array.isArray(order)) {
      return properties;
    }
    const arrayToHash = (arr) => arr.reduce((acc, curr) => {
      acc[curr] = true;
      return acc;
    }, {});
    const errorPropList = (arr) => arr.length > 1 ? `properties '${arr.join("', '")}'` : `property '${arr[0]}'`;
    const propertyHash = arrayToHash(properties);
    const orderFiltered = order.filter((prop) => prop === "*" || propertyHash[prop]);
    const orderHash = arrayToHash(orderFiltered);
    const rest = properties.filter((prop) => !orderHash[prop]);
    const restIndex = orderFiltered.indexOf("*");
    if (restIndex === -1) {
      if (rest.length) {
        throw new Error(`uiSchema order list does not contain ${errorPropList(rest)}`);
      }
      return orderFiltered;
    }
    if (restIndex !== orderFiltered.lastIndexOf("*")) {
      throw new Error("uiSchema order list contains more than one wildcard item");
    }
    const complete = [...orderFiltered];
    complete.splice(restIndex, 1, ...rest);
    return complete;
  }

  // src/parseDateString.ts
  function parseDateString(dateString, includeTime = true) {
    if (!dateString) {
      return {
        year: -1,
        month: -1,
        day: -1,
        hour: includeTime ? -1 : 0,
        minute: includeTime ? -1 : 0,
        second: includeTime ? -1 : 0
      };
    }
    const date = new Date(dateString);
    if (Number.isNaN(date.getTime())) {
      throw new Error(`Unable to parse date ${dateString}`);
    }
    return {
      year: date.getUTCFullYear(),
      month: date.getUTCMonth() + 1,
      // oh you, javascript.
      day: date.getUTCDate(),
      hour: includeTime ? date.getUTCHours() : 0,
      minute: includeTime ? date.getUTCMinutes() : 0,
      second: includeTime ? date.getUTCSeconds() : 0
    };
  }

  // src/removeOptionalEmptyObjects.ts
  function removeOptionalEmptyObjects(validator, schema, rootSchema, formData) {
    if (!isObject(schema)) {
      return formData;
    }
    const resolvedSchema = retrieveSchema(validator, schema, rootSchema, formData);
    if (Array.isArray(formData)) {
      const itemsSchema = resolvedSchema.items;
      if (!itemsSchema) {
        return formData;
      }
      let hasChanges = false;
      const mapped = formData.map((item, index) => {
        let itemSchema = itemsSchema;
        if (Array.isArray(itemsSchema)) {
          itemSchema = itemsSchema[index] || resolvedSchema.additionalItems || {};
        }
        const cleaned = removeOptionalEmptyObjects(validator, itemSchema, rootSchema, item);
        if (cleaned !== item) {
          hasChanges = true;
        }
        return cleaned === void 0 ? {} : cleaned;
      });
      return hasChanges ? mapped : formData;
    }
    const { properties, required: requiredFields = [] } = resolvedSchema;
    if (!isObject(formData) || !properties) {
      return formData;
    }
    const result = {};
    const data = formData;
    let hasAnyValue = false;
    for (const key of Object.keys(data)) {
      const value = data[key];
      const propertySchema = properties[key] || {};
      const isRequired = requiredFields.includes(key);
      const isObj = isObject(value);
      const isArr = Array.isArray(value);
      if ((isObj || isArr) && properties[key]) {
        const cleaned = removeOptionalEmptyObjects(validator, propertySchema, rootSchema, value);
        if (isRequired || !isValueEmpty(cleaned)) {
          result[key] = cleaned;
          hasAnyValue = true;
        }
      } else if (isRequired || !isValueEmpty(value) || !properties[key]) {
        result[key] = value;
        if (!isValueEmpty(value)) {
          hasAnyValue = true;
        }
      }
    }
    if (!hasAnyValue && Object.keys(result).length === 0) {
      return void 0;
    }
    return result;
  }
  function resolveUiSchema(schema, localUiSchema, registry) {
    const ref = schema[RJSF_REF_KEY] ?? schema[REF_KEY];
    const definitions = registry.uiSchemaDefinitions;
    const definitionUiSchema = ref && definitions ? definitions[ref] : void 0;
    let result;
    if (!definitionUiSchema) {
      result = localUiSchema || {};
    } else if (!localUiSchema || isEmpty4(localUiSchema)) {
      result = { ...definitionUiSchema };
    } else {
      result = mergeObjects(definitionUiSchema, localUiSchema);
    }
    if (definitions) {
      let resolvedSchema = schema;
      if (ref && schema[REF_KEY] && !schema[RJSF_REF_KEY]) {
        try {
          resolvedSchema = findSchemaDefinition(ref, registry.rootSchema);
        } catch (e) {
          console.warn("could not resolve $ref in resolveUiSchema:\n", e);
          return result;
        }
      }
      for (const keyword of [ONE_OF_KEY, ANY_OF_KEY]) {
        const schemaOptions = resolvedSchema[keyword];
        if (Array.isArray(schemaOptions) && schemaOptions.length > 0) {
          const currentUiSchemaArray = result[keyword];
          const uiSchemaArray = Array.isArray(currentUiSchemaArray) ? [...currentUiSchemaArray] : [];
          let hasExpanded = false;
          for (let i = 0; i < schemaOptions.length; i++) {
            const option = schemaOptions[i];
            const optionRef = option?.[RJSF_REF_KEY] ?? option?.[REF_KEY];
            if (optionRef && optionRef in definitions) {
              const optionUiSchema = uiSchemaArray[i] || {};
              uiSchemaArray[i] = mergeObjects(definitions[optionRef], optionUiSchema);
              hasExpanded = true;
            }
          }
          if (hasExpanded) {
            result[keyword] = uiSchemaArray;
          }
        }
      }
    }
    return result;
  }

  // src/schemaRequiresTrueValue.ts
  function schemaRequiresTrueValue(schema) {
    if (schema.const) {
      return true;
    }
    if (schema.enum && schema.enum.length === 1 && schema.enum[0] === true) {
      return true;
    }
    if (schema.anyOf && schema.anyOf.length === 1) {
      return schemaRequiresTrueValue(schema.anyOf[0]);
    }
    if (schema.oneOf && schema.oneOf.length === 1) {
      return schemaRequiresTrueValue(schema.oneOf[0]);
    }
    if (schema.allOf) {
      const schemaSome = (subSchema) => schemaRequiresTrueValue(subSchema);
      return schema.allOf.some(schemaSome);
    }
    return false;
  }

  // src/shallowEquals.ts
  function shallowEquals(a, b) {
    if (Object.is(a, b)) {
      return true;
    }
    if (a == null || b == null) {
      return false;
    }
    if (typeof a !== "object" || typeof b !== "object") {
      return false;
    }
    const keysA = Object.keys(a);
    const keysB = Object.keys(b);
    if (keysA.length !== keysB.length) {
      return false;
    }
    for (let i = 0; i < keysA.length; i += 1) {
      const key = keysA[i];
      if (!Object.hasOwn(b, key) || !Object.is(a[key], b[key])) {
        return false;
      }
    }
    return true;
  }

  // src/shouldRender.ts
  function shouldRender(component, nextProps, nextState, updateStrategy = "customDeep") {
    if (updateStrategy === "always") {
      return true;
    }
    if (updateStrategy === "shallow") {
      const { props: props2, state: state2 } = component;
      return !shallowEquals(props2, nextProps) || !shallowEquals(state2, nextState);
    }
    const { props, state } = component;
    return !deepEquals_default(props, nextProps) || !deepEquals_default(state, nextState);
  }
  function getSchemaTypesForXxxOf(schemas) {
    const allTypes = uniq(
      schemas.map((s) => isObject2(s) ? getSchemaType(s) : void 0).flat().filter((t) => t !== void 0)
    );
    return allTypes.length === 1 ? allTypes[0] : allTypes;
  }
  function shouldRenderOptionalField(registry, schema, required, uiSchema) {
    const { enableOptionalDataFieldForType = [] } = getUiOptions(uiSchema, registry.globalUiOptions);
    let schemaType;
    if (ANY_OF_KEY in schema && Array.isArray(schema[ANY_OF_KEY])) {
      schemaType = getSchemaTypesForXxxOf(schema[ANY_OF_KEY]);
    } else if (ONE_OF_KEY in schema && Array.isArray(schema[ONE_OF_KEY])) {
      schemaType = getSchemaTypesForXxxOf(schema[ONE_OF_KEY]);
    } else {
      schemaType = getSchemaType(schema);
    }
    return !isRootSchema(registry, schema) && !required && !!schemaType && !Array.isArray(schemaType) && !!enableOptionalDataFieldForType.find((val) => val === schemaType);
  }

  // src/toDateString.ts
  function toDateString(dateObject, time = true) {
    const { year, month, day, hour = 0, minute = 0, second = 0 } = dateObject;
    const utcTime = Date.UTC(year, month - 1, day, hour, minute, second);
    const datetime = new Date(utcTime).toJSON();
    return time ? datetime : datetime.slice(0, 10);
  }
  function toErrorList(errorSchema, fieldPath = []) {
    if (!errorSchema) {
      return [];
    }
    let errorList = [];
    if (ERRORS_KEY in errorSchema) {
      errorList = errorList.concat(
        errorSchema[ERRORS_KEY].map((message) => {
          const property = `.${fieldPath.join(".")}`;
          return {
            property,
            message,
            stack: `${property} ${message}`
          };
        })
      );
    }
    return Object.keys(errorSchema).reduce((currentList, key) => {
      if (key !== ERRORS_KEY) {
        const childSchema = errorSchema[key];
        if (isPlainObject2(childSchema)) {
          return currentList.concat(toErrorList(childSchema, [...fieldPath, key]));
        }
      }
      return currentList;
    }, errorList);
  }
  function toErrorSchema(errors) {
    const builder = new ErrorSchemaBuilder();
    if (errors.length) {
      errors.forEach((error) => {
        const { property, message } = error;
        const path = property === "." ? [] : toPath(property);
        if (path.length > 0 && path[0] === "") {
          path.splice(0, 1);
        }
        if (message) {
          builder.addErrors(message, path);
        }
      });
    }
    return builder.ErrorSchema;
  }

  // src/toFieldPathId.ts
  function toFieldPathId(fieldPath, globalFormOptions, parentPath, isMultiValue) {
    const basePath = Array.isArray(parentPath) ? parentPath : parentPath?.path;
    const childPath = fieldPath === "" ? [] : [fieldPath];
    const path = basePath ? basePath.concat(...childPath) : childPath;
    const id = [globalFormOptions.idPrefix, ...path].join(globalFormOptions.idSeparator);
    let name;
    if (globalFormOptions.nameGenerator && path.length > 0) {
      name = globalFormOptions.nameGenerator(path, globalFormOptions.idPrefix, isMultiValue);
    }
    return { path, [ID_KEY]: id, ...name !== void 0 && { name } };
  }
  function unwrapErrorHandler(errorHandler) {
    return Object.keys(errorHandler).reduce((acc, key) => {
      if (key === "addError") {
        return acc;
      }
      const childSchema = errorHandler[key];
      if (isPlainObject2(childSchema)) {
        return {
          ...acc,
          [key]: unwrapErrorHandler(childSchema)
        };
      }
      return { ...acc, [key]: childSchema };
    }, {});
  }
  function readyForChange(state) {
    return Object.values(state).every((value) => value !== -1);
  }
  function DateElement(props) {
    const {
      className = "form-control",
      type,
      range,
      value,
      select,
      rootId,
      name,
      disabled,
      readonly,
      autofocus,
      registry,
      onBlur,
      onFocus
    } = props;
    const id = `${rootId}_${type}`;
    const { SelectWidget } = registry.widgets;
    const onChange = react.useCallback((newValue) => select(type, newValue), [select, type]);
    return /* @__PURE__ */ jsxRuntime.jsx(
      SelectWidget,
      {
        schema: { type: "integer" },
        id,
        name,
        className,
        options: { enumOptions: dateRangeOptions(range[0], range[1]) },
        placeholder: type,
        value,
        disabled,
        readonly,
        autofocus,
        onChange,
        onBlur,
        onFocus,
        registry,
        label: "",
        "aria-describedby": ariaDescribedByIds(rootId)
      }
    );
  }
  function useAltDateWidgetProps(props) {
    const { time = false, disabled = false, readonly = false, options, onChange, value } = props;
    const [state, setState] = react.useState(parseDateString(value, time));
    react.useEffect(() => {
      setState(parseDateString(value, time));
    }, [time, value]);
    const handleChange = react.useCallback(
      (property, newValue) => {
        const nextState = {
          ...state,
          [property]: typeof newValue === "undefined" ? -1 : newValue
        };
        if (readyForChange(nextState)) {
          onChange(toDateString(nextState, time));
        } else {
          setState(nextState);
        }
      },
      [state, onChange, time]
    );
    const handleClear = react.useCallback(
      (event) => {
        event.preventDefault();
        if (disabled || readonly) {
          return;
        }
        onChange(void 0);
      },
      [disabled, readonly, onChange]
    );
    const handleSetNow = react.useCallback(
      (event) => {
        event.preventDefault();
        if (disabled || readonly) {
          return;
        }
        const nextState = parseDateString((/* @__PURE__ */ new Date()).toJSON(), time);
        onChange(toDateString(nextState, time));
      },
      [disabled, readonly, time, onChange]
    );
    const elements = react.useMemo(
      () => getDateElementProps(
        state,
        time,
        options.yearsRange,
        options.format
      ),
      [state, time, options.yearsRange, options.format]
    );
    return { elements, handleChange, handleClear, handleSetNow };
  }
  function useDeepCompareMemo(newValue) {
    const valueRef = react.useRef(newValue);
    if (!deepEquals_default(newValue, valueRef.current)) {
      valueRef.current = newValue;
    }
    return valueRef.current;
  }
  function addNameToDataURL(dataURL, name) {
    return dataURL.replace(";base64", `;name=${encodeURIComponent(name)};base64`);
  }
  function processFile(file) {
    const { name, size, type } = file;
    return new Promise((resolve, reject) => {
      const reader = new window.FileReader();
      reader.onerror = reject;
      reader.onload = (event) => {
        if (typeof event.target?.result === "string") {
          resolve({
            dataURL: addNameToDataURL(event.target.result, name),
            name,
            size,
            type
          });
        } else {
          resolve({
            dataURL: null,
            name,
            size,
            type
          });
        }
      };
      reader.readAsDataURL(file);
    });
  }
  function processFiles(files) {
    return Promise.all(Array.from(files).map(processFile));
  }
  function extractFileInfo(dataURLs) {
    return dataURLs.reduce((acc, dataURL) => {
      if (!dataURL) {
        return acc;
      }
      try {
        const { blob, name } = dataURItoBlob(dataURL);
        return [
          ...acc,
          {
            dataURL,
            name,
            size: blob.size,
            type: blob.type
          }
        ];
      } catch {
        return acc;
      }
    }, []);
  }
  function useFileWidgetProps(value, onChange, multiple = false) {
    const values = react.useMemo(() => {
      if (multiple && value) {
        return Array.isArray(value) ? value : [value];
      }
      return [];
    }, [value, multiple]);
    const filesInfo = react.useMemo(
      () => Array.isArray(value) ? extractFileInfo(value) : extractFileInfo([value || ""]),
      [value]
    );
    const handleChange = react.useCallback(
      async (files) => {
        const filesInfoEvent = await processFiles(files);
        const newValue = filesInfoEvent.map((fileInfo) => fileInfo.dataURL || null);
        if (multiple) {
          onChange(values.concat(...newValue));
        } else {
          onChange(newValue[0]);
        }
      },
      [values, multiple, onChange]
    );
    const handleRemove = react.useCallback(
      (index) => {
        if (multiple) {
          const newValue = values.filter((_, i) => i !== index);
          onChange(newValue);
        } else {
          onChange(void 0);
        }
      },
      [values, multiple, onChange]
    );
    return { filesInfo, handleChange, handleRemove };
  }

  // src/utcToLocal.ts
  function utcToLocal(jsonDate) {
    if (!jsonDate) {
      return "";
    }
    const date = new Date(jsonDate);
    const yyyy = pad(date.getFullYear(), 4);
    const MM = pad(date.getMonth() + 1, 2);
    const dd = pad(date.getDate(), 2);
    const hh = pad(date.getHours(), 2);
    const mm = pad(date.getMinutes(), 2);
    const ss = pad(date.getSeconds(), 2);
    const SSS = pad(date.getMilliseconds(), 3);
    return `${yyyy}-${MM}-${dd}T${hh}:${mm}:${ss}.${SSS}`;
  }
  function validationDataMerge(validationData, additionalErrorSchema, preventDuplicates = false) {
    if (!additionalErrorSchema) {
      return validationData;
    }
    const { errors: oldErrors, errorSchema: oldErrorSchema } = validationData;
    let errors = toErrorList(additionalErrorSchema);
    let errorSchema = additionalErrorSchema;
    if (!isEmpty4(oldErrorSchema)) {
      errorSchema = mergeObjects(
        oldErrorSchema,
        additionalErrorSchema,
        preventDuplicates ? "preventDuplicates" : true
      );
      errors = [...oldErrors].concat(errors);
    }
    return { errorSchema, errors };
  }
  function withIdRefPrefixObject(node) {
    const realObj = node;
    for (const key in realObj) {
      if (Object.hasOwn(realObj, key)) {
        const value = realObj[key];
        if (key === REF_KEY && typeof value === "string" && value.startsWith("#")) {
          realObj[key] = ROOT_SCHEMA_PREFIX + value;
        } else {
          realObj[key] = withIdRefPrefix(value);
        }
      }
    }
    return node;
  }
  function withIdRefPrefixArray(node) {
    return node.map((item) => withIdRefPrefix(item));
  }
  function withIdRefPrefix(schemaNode) {
    if (Array.isArray(schemaNode)) {
      return withIdRefPrefixArray([...schemaNode]);
    }
    if (isObject2(schemaNode)) {
      return withIdRefPrefixObject({ ...schemaNode });
    }
    return schemaNode;
  }

  // src/enums.ts
  var TranslatableString = /* @__PURE__ */ ((TranslatableString2) => {
    TranslatableString2["ArrayItemTitle"] = "Item";
    TranslatableString2["MissingItems"] = "Missing items definition";
    TranslatableString2["EmptyArray"] = "No items yet. Use the button below to add some.";
    TranslatableString2["YesLabel"] = "Yes";
    TranslatableString2["NoLabel"] = "No";
    TranslatableString2["CloseLabel"] = "Close";
    TranslatableString2["ErrorsLabel"] = "Errors";
    TranslatableString2["NewStringDefault"] = "New Value";
    TranslatableString2["AddButton"] = "Add";
    TranslatableString2["AddItemButton"] = "Add Item";
    TranslatableString2["CopyButton"] = "Copy";
    TranslatableString2["MoveDownButton"] = "Move down";
    TranslatableString2["MoveUpButton"] = "Move up";
    TranslatableString2["RemoveButton"] = "Remove";
    TranslatableString2["NowLabel"] = "Now";
    TranslatableString2["ClearLabel"] = "Clear";
    TranslatableString2["AriaDateLabel"] = "Select a date";
    TranslatableString2["PreviewLabel"] = "Preview";
    TranslatableString2["DecrementAriaLabel"] = "Decrease value by 1";
    TranslatableString2["IncrementAriaLabel"] = "Increase value by 1";
    TranslatableString2["OptionalObjectAdd"] = "Add data for optional field";
    TranslatableString2["OptionalObjectRemove"] = "Remove data for optional field";
    TranslatableString2["OptionalObjectEmptyMsg"] = "No data for optional field";
    TranslatableString2["Type"] = "Type";
    TranslatableString2["Value"] = "Value";
    TranslatableString2["ClearButton"] = "clear input";
    TranslatableString2["UnknownFieldType"] = "Unknown field type %1";
    TranslatableString2["OptionPrefix"] = "Option %1";
    TranslatableString2["TitleOptionPrefix"] = "%1 option %2";
    TranslatableString2["KeyLabel"] = "%1 Key";
    TranslatableString2["DeprecatedLabel"] = "%1 (deprecated)";
    TranslatableString2["InvalidObjectField"] = 'Invalid "%1" object field configuration: _%2_.';
    TranslatableString2["UnsupportedField"] = "Unsupported field schema.";
    TranslatableString2["UnsupportedFieldWithId"] = "Unsupported field schema for field `%1`.";
    TranslatableString2["UnsupportedFieldWithReason"] = "Unsupported field schema: _%1_.";
    TranslatableString2["UnsupportedFieldWithIdAndReason"] = "Unsupported field schema for field `%1`: _%2_.";
    TranslatableString2["FilesInfo"] = "**%1** (%2, %3 bytes)";
    return TranslatableString2;
  })(TranslatableString || {});
  var ParserValidator = class {
    /** Construct the ParserValidator for the given `rootSchema`. This `rootSchema` will be stashed in the `schemaMap`
     * first.
     *
     * @param rootSchema - The root schema against which this validator will be executed
     */
    constructor(rootSchema) {
      /** The map of schemas encountered by the ParserValidator */
      this.schemaMap = {};
      this.rootSchema = rootSchema;
      this.addSchema(rootSchema, hashForSchema(rootSchema));
    }
    /** Resets the internal AJV validator to clear schemas from it. Can be helpful for resetting the validator for tests.
     */
    reset() {
      this.schemaMap = {};
    }
    /** Adds the given `schema` to the `schemaMap` keyed by the `hash` or `ID_KEY` if present on the `schema`. If the
     * schema does not have an `ID_KEY`, then the `hash` will be added as the `ID_KEY` to allow the schema to be
     * associated with it's `hash` for future use (by a schema compiler).
     *
     * @param schema - The schema which is to be added to the map
     * @param hash - The hash value at which to map the schema
     */
    addSchema(schema, hash) {
      const key = get15(schema, ID_KEY, hash);
      const identifiedSchema = { ...schema, [ID_KEY]: key };
      const existing = this.schemaMap[key];
      if (!existing) {
        this.schemaMap[key] = identifiedSchema;
      } else if (!deepEquals_default(existing, identifiedSchema)) {
        console.error("existing schema:", JSON.stringify(existing, null, 2));
        console.error("new schema:", JSON.stringify(identifiedSchema, null, 2));
        throw new Error(
          `Two different schemas exist with the same key ${key}! What a bad coincidence. If possible, try adding an $id to one of the schemas`
        );
      }
    }
    /** Returns the current `schemaMap` to the caller
     */
    getSchemaMap() {
      return this.schemaMap;
    }
    /** Implements the `ValidatorType` `isValid()` method to capture the `schema` in the `schemaMap`. Throws an error when
     * the `rootSchema` is not the same as the root schema provided during construction.
     *
     * @param schema - The schema to record in the `schemaMap`
     * @param _formData - The formData parameter that is ignored
     * @param rootSchema - The root schema associated with the schema
     * @throws - Error when the given `rootSchema` differs from the root schema provided during construction
     */
    isValid(schema, _formData, rootSchema) {
      if (!deepEquals_default(rootSchema, this.rootSchema)) {
        throw new Error("Unexpectedly calling isValid() with a rootSchema that differs from the construction rootSchema");
      }
      this.addSchema(schema, hashForSchema(schema));
      return false;
    }
    /** Implements the `ValidatorType` `rawValidation()` method to throw an error since it is never supposed to be called
     *
     * @param _schema - The schema parameter that is ignored
     * @param _formData - The formData parameter that is ignored
     */
    rawValidation(_schema, _formData) {
      throw new Error("Unexpectedly calling the `rawValidation()` method during schema parsing");
    }
    /** Implements the `ValidatorType` `toErrorList()` method to throw an error since it is never supposed to be called
     *
     * @param _errorSchema - The error schema parameter that is ignored
     * @param _fieldPath - The field path parameter that is ignored
     */
    toErrorList(_errorSchema, _fieldPath) {
      throw new Error("Unexpectedly calling the `toErrorList()` method during schema parsing");
    }
    /** Implements the `ValidatorType` `validateFormData()` method to throw an error since it is never supposed to be
     * called
     *
     * @param _formData - The formData parameter that is ignored
     * @param _schema - The schema parameter that is ignored
     * @param _customValidate - The customValidate parameter that is ignored
     * @param _transformErrors - The transformErrors parameter that is ignored
     * @param _uiSchema - The uiSchema parameter that is ignored
     */
    validateFormData(_formData, _schema, _customValidate, _transformErrors, _uiSchema) {
      throw new Error("Unexpectedly calling the `validateFormData()` method during schema parsing");
    }
  };

  // src/parser/schemaParser.ts
  function parseSchema(validator, recurseList, rootSchema, schema) {
    const schemas = retrieveSchemaInternal(validator, schema, rootSchema, void 0, true);
    schemas.forEach((localSchema) => {
      const sameSchemaIndex = recurseList.findIndex((item) => deepEquals_default(item, localSchema));
      if (sameSchemaIndex === -1) {
        recurseList.push(localSchema);
        const allOptions = resolveAnyOrOneOfSchemas(validator, localSchema, rootSchema, true);
        allOptions.forEach((s) => {
          if (PROPERTIES_KEY in s && s[PROPERTIES_KEY]) {
            forEach(localSchema[PROPERTIES_KEY], (value) => {
              parseSchema(validator, recurseList, rootSchema, value);
            });
          }
        });
        if (ITEMS_KEY in localSchema && !Array.isArray(localSchema.items) && typeof localSchema.items !== "boolean") {
          parseSchema(validator, recurseList, rootSchema, localSchema.items);
        }
      }
    });
  }
  function schemaParser(rootSchema) {
    const validator = new ParserValidator(rootSchema);
    const recurseList = [];
    parseSchema(validator, recurseList, rootSchema, rootSchema);
    return validator.getSchemaMap();
  }

  exports.ADDITIONAL_PROPERTIES_KEY = ADDITIONAL_PROPERTIES_KEY;
  exports.ADDITIONAL_PROPERTY_FLAG = ADDITIONAL_PROPERTY_FLAG;
  exports.ALL_OF_KEY = ALL_OF_KEY;
  exports.ANY_OF_KEY = ANY_OF_KEY;
  exports.CONST_KEY = CONST_KEY;
  exports.DEFAULT_ID_PREFIX = DEFAULT_ID_PREFIX;
  exports.DEFAULT_ID_SEPARATOR = DEFAULT_ID_SEPARATOR;
  exports.DEFAULT_KEY = DEFAULT_KEY;
  exports.DEFINITIONS_KEY = DEFINITIONS_KEY;
  exports.DEPENDENCIES_KEY = DEPENDENCIES_KEY;
  exports.DISCRIMINATOR_PATH = DISCRIMINATOR_PATH;
  exports.DateElement = DateElement;
  exports.ENUM_KEY = ENUM_KEY;
  exports.ERRORS_KEY = ERRORS_KEY;
  exports.ErrorSchemaBuilder = ErrorSchemaBuilder;
  exports.FORM_CONTEXT_NAME = FORM_CONTEXT_NAME;
  exports.ID_KEY = ID_KEY;
  exports.IF_KEY = IF_KEY;
  exports.ITEMS_KEY = ITEMS_KEY;
  exports.JSON_SCHEMA_DRAFT_2019_09 = JSON_SCHEMA_DRAFT_2019_09;
  exports.JSON_SCHEMA_DRAFT_2020_12 = JSON_SCHEMA_DRAFT_2020_12;
  exports.JUNK_OPTION_ID = JUNK_OPTION_ID;
  exports.LOOKUP_MAP_NAME = LOOKUP_MAP_NAME;
  exports.NAME_KEY = NAME_KEY;
  exports.ONE_OF_KEY = ONE_OF_KEY;
  exports.PATTERN_PROPERTIES_KEY = PATTERN_PROPERTIES_KEY;
  exports.PROPERTIES_KEY = PROPERTIES_KEY;
  exports.READONLY_KEY = READONLY_KEY;
  exports.REF_KEY = REF_KEY;
  exports.REQUIRED_KEY = REQUIRED_KEY;
  exports.RJSF_ADDITIONAL_PROPERTIES_FLAG = RJSF_ADDITIONAL_PROPERTIES_FLAG;
  exports.RJSF_REF_KEY = RJSF_REF_KEY;
  exports.ROOT_SCHEMA_PREFIX = ROOT_SCHEMA_PREFIX;
  exports.SCHEMA_KEY = SCHEMA_KEY;
  exports.SUBMIT_BTN_OPTIONS_KEY = SUBMIT_BTN_OPTIONS_KEY;
  exports.TranslatableString = TranslatableString;
  exports.UI_DEFINITIONS_KEY = UI_DEFINITIONS_KEY;
  exports.UI_FIELD_KEY = UI_FIELD_KEY;
  exports.UI_GLOBAL_OPTIONS_KEY = UI_GLOBAL_OPTIONS_KEY;
  exports.UI_OPTIONS_KEY = UI_OPTIONS_KEY;
  exports.UI_WIDGET_KEY = UI_WIDGET_KEY;
  exports.allowAdditionalItems = allowAdditionalItems;
  exports.ariaDescribedByIds = ariaDescribedByIds;
  exports.asNumber = asNumber;
  exports.bracketNameGenerator = bracketNameGenerator;
  exports.buttonId = buttonId;
  exports.canExpand = canExpand;
  exports.createErrorHandler = createErrorHandler;
  exports.createSchemaUtils = createSchemaUtils;
  exports.dataURItoBlob = dataURItoBlob;
  exports.dateRangeOptions = dateRangeOptions;
  exports.deepEquals = deepEquals_default;
  exports.descriptionId = descriptionId;
  exports.dotNotationNameGenerator = dotNotationNameGenerator;
  exports.englishStringTranslator = englishStringTranslator;
  exports.enumOptionSelectedValue = enumOptionSelectedValue;
  exports.enumOptionValueDecoder = enumOptionValueDecoder;
  exports.enumOptionValueEncoder = enumOptionValueEncoder;
  exports.enumOptionsDeselectValue = enumOptionsDeselectValue;
  exports.enumOptionsIndexForValue = enumOptionsIndexForValue;
  exports.enumOptionsIsSelected = enumOptionsIsSelected;
  exports.enumOptionsSelectValue = enumOptionsSelectValue;
  exports.enumOptionsValueForIndex = enumOptionsValueForIndex;
  exports.errorId = errorId;
  exports.examplesId = examplesId;
  exports.findFieldInSchema = findFieldInSchema;
  exports.findSchemaDefinition = findSchemaDefinition;
  exports.findSelectedOptionInXxxOf = findSelectedOptionInXxxOf;
  exports.getChangedFields = getChangedFields;
  exports.getClosestMatchingOption = getClosestMatchingOption;
  exports.getDateElementProps = getDateElementProps;
  exports.getDefaultFormState = getDefaultFormState;
  exports.getDiscriminatorFieldFromSchema = getDiscriminatorFieldFromSchema;
  exports.getDisplayLabel = getDisplayLabel;
  exports.getFieldNames = getFieldNames;
  exports.getFirstMatchingOption = getFirstMatchingOption;
  exports.getFromSchema = getFromSchema;
  exports.getInputProps = getInputProps;
  exports.getOptionMatchingSimpleDiscriminator = getOptionMatchingSimpleDiscriminator;
  exports.getOptionValueFormat = getOptionValueFormat;
  exports.getSchemaType = getSchemaType;
  exports.getSubmitButtonOptions = getSubmitButtonOptions;
  exports.getTemplate = getTemplate;
  exports.getTestIds = getTestIds;
  exports.getUiOptions = getUiOptions;
  exports.getUsedFormData = getUsedFormData;
  exports.getWidget = getWidget;
  exports.guessType = guessType;
  exports.hasWidget = hasWidget;
  exports.hashForSchema = hashForSchema;
  exports.hashObject = hashObject;
  exports.hashString = hashString;
  exports.helpId = helpId;
  exports.isConstant = isConstant;
  exports.isCustomWidget = isCustomWidget;
  exports.isFilesArray = isFilesArray;
  exports.isFixedItems = isFixedItems;
  exports.isFormDataAvailable = isFormDataAvailable;
  exports.isMultiSelect = isMultiSelect;
  exports.isObject = isObject;
  exports.isRootSchema = isRootSchema;
  exports.isSelect = isSelect;
  exports.isValueEmpty = isValueEmpty;
  exports.labelValue = labelValue;
  exports.localToUTC = localToUTC;
  exports.lookupFromFormContext = lookupFromFormContext;
  exports.mergeDefaultsWithFormData = mergeDefaultsWithFormData;
  exports.mergeObjects = mergeObjects;
  exports.mergeSchemas = mergeSchemas;
  exports.omitExtraData = omitExtraData;
  exports.optionId = optionId;
  exports.optionalControlsId = optionalControlsId;
  exports.optionsList = optionsList;
  exports.orderProperties = orderProperties;
  exports.pad = pad;
  exports.parseDateString = parseDateString;
  exports.rangeSpec = rangeSpec;
  exports.relaxOptionsForScoring = relaxOptionsForScoring;
  exports.removeOptionalEmptyObjects = removeOptionalEmptyObjects;
  exports.replaceStringParameters = replaceStringParameters;
  exports.resolveUiSchema = resolveUiSchema;
  exports.retrieveSchema = retrieveSchema;
  exports.sanitizeDataForNewSchema = sanitizeDataForNewSchema;
  exports.schemaParser = schemaParser;
  exports.schemaRequiresTrueValue = schemaRequiresTrueValue;
  exports.shallowEquals = shallowEquals;
  exports.shouldRender = shouldRender;
  exports.shouldRenderOptionalField = shouldRenderOptionalField;
  exports.sortedJSONStringify = sortedJSONStringify;
  exports.titleId = titleId;
  exports.toConstant = toConstant;
  exports.toDateString = toDateString;
  exports.toErrorList = toErrorList;
  exports.toErrorSchema = toErrorSchema;
  exports.toFieldPathId = toFieldPathId;
  exports.toPathSchema = toPathSchema;
  exports.unwrapErrorHandler = unwrapErrorHandler;
  exports.useAltDateWidgetProps = useAltDateWidgetProps;
  exports.useDeepCompareMemo = useDeepCompareMemo;
  exports.useFileWidgetProps = useFileWidgetProps;
  exports.utcToLocal = utcToLocal;
  exports.validationDataMerge = validationDataMerge;
  exports.withIdRefPrefix = withIdRefPrefix;

}));
