// @rjsf/core/dist/core.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('react'), require('@rjsf/utils'), require('lodash/cloneDeep'), require('lodash/get'), require('lodash/isEmpty'), require('lodash/pick'), require('lodash/set'), require('lodash/toPath'), require('lodash/unset'), require('lodash/isObject'), require('lodash/uniqueId'), require('react/jsx-runtime'), require('lodash/each'), require('lodash/flatten'), require('lodash/has'), require('lodash/includes'), require('lodash/intersection'), require('lodash/isEqual'), require('lodash/isFunction'), require('lodash/isPlainObject'), require('lodash/isString'), require('lodash/isUndefined'), require('lodash/last'), require('lodash/noop'), require('lodash/omit'), require('markdown-to-jsx/react'), require('@rjsf/validator-ajv8')) :
  typeof define === 'function' && define.amd ? define(['exports', 'react', '@rjsf/utils', 'lodash/cloneDeep', 'lodash/get', 'lodash/isEmpty', 'lodash/pick', 'lodash/set', 'lodash/toPath', 'lodash/unset', 'lodash/isObject', 'lodash/uniqueId', 'react/jsx-runtime', 'lodash/each', 'lodash/flatten', 'lodash/has', 'lodash/includes', 'lodash/intersection', 'lodash/isEqual', 'lodash/isFunction', 'lodash/isPlainObject', 'lodash/isString', 'lodash/isUndefined', 'lodash/last', 'lodash/noop', 'lodash/omit', 'markdown-to-jsx/react', '@rjsf/validator-ajv8'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.JSONSchemaForm = {}, global.react, global.utils, global._cloneDeep, global.get, global.isEmpty, global._pick, global.set, global._toPath, global._unset, global.isObject, global.uniqueId, global.jsxRuntime, global.each, global.flatten, global.has, global.includes, global.intersection, global.isEqual, global.isFunction, global.isPlainObject, global.isString, global.isUndefined, global.last, global.noop, global.omit3, global.react$1, global.validator));
})(this, (function (exports, react, utils, _cloneDeep, get, isEmpty, _pick, set, _toPath, _unset, isObject, uniqueId, jsxRuntime, each, flatten, has, includes, intersection, isEqual, isFunction, isPlainObject, isString, isUndefined, last, noop, omit3, react$1, validator) { 'use strict';

  // src/components/Form.tsx
  function generateRowId() {
    return uniqueId("rjsf-array-item-");
  }
  function generateKeyedFormData(formData) {
    return !Array.isArray(formData) ? [] : formData.map((item) => ({
      key: generateRowId(),
      item
    }));
  }
  function keyedToPlainFormData(keyedFormData) {
    if (Array.isArray(keyedFormData)) {
      return keyedFormData.map((keyedItem) => keyedItem.item);
    }
    return [];
  }
  function isItemRequired(itemSchema) {
    if (Array.isArray(itemSchema.type)) {
      return !itemSchema.type.includes("null");
    }
    return itemSchema.type !== "null";
  }
  function canAddItem(registry, schema, formItems, uiSchema) {
    let { addable } = utils.getUiOptions(uiSchema, registry.globalUiOptions);
    if (addable !== false) {
      if (schema.maxItems !== void 0) {
        addable = formItems.length < schema.maxItems;
      } else {
        addable = true;
      }
    }
    return addable;
  }
  function computeItemUiSchema(uiSchema, item, index, formContext) {
    if (typeof uiSchema.items === "function") {
      try {
        const result = uiSchema.items(item, index, formContext);
        return result;
      } catch (e) {
        console.error(`Error executing dynamic uiSchema.items function for item at index ${index}:`, e);
        return void 0;
      }
    } else {
      return uiSchema.items;
    }
  }
  function getNewFormDataRow(registry, schema) {
    const { schemaUtils, globalFormOptions } = registry;
    let itemSchema = schema.items;
    if (globalFormOptions.useFallbackUiForUnsupportedType && !itemSchema) {
      itemSchema = {};
    } else if (utils.isFixedItems(schema) && utils.allowAdditionalItems(schema)) {
      itemSchema = schema.additionalItems;
    }
    return schemaUtils.getDefaultFormState(itemSchema);
  }
  function ArrayAsMultiSelect(props) {
    const {
      schema,
      fieldPathId,
      uiSchema,
      formData: items = [],
      disabled = false,
      readonly = false,
      autofocus = false,
      required = false,
      placeholder,
      onBlur,
      onFocus,
      registry,
      rawErrors,
      name,
      onSelectChange
    } = props;
    const { widgets: widgets2, schemaUtils, globalFormOptions, globalUiOptions } = registry;
    const itemsSchema = schemaUtils.retrieveSchema(schema.items, items);
    const itemsUiSchema = uiSchema?.items ?? uiSchema;
    const enumOptions = utils.optionsList(itemsSchema, itemsUiSchema);
    const { widget = "select", title: uiTitle, ...options } = utils.getUiOptions(uiSchema, globalUiOptions);
    const Widget = utils.getWidget(schema, widget, widgets2);
    const label = uiTitle ?? schema.title ?? name;
    const displayLabel = schemaUtils.getDisplayLabel(schema, uiSchema, globalUiOptions);
    const multiValueFieldPathId = utils.useDeepCompareMemo(utils.toFieldPathId("", globalFormOptions, fieldPathId, true));
    return /* @__PURE__ */ jsxRuntime.jsx(
      Widget,
      {
        id: multiValueFieldPathId[utils.ID_KEY],
        name,
        multiple: true,
        onChange: onSelectChange,
        onBlur,
        onFocus,
        options: { ...options, enumOptions },
        schema,
        uiSchema,
        registry,
        value: items,
        disabled,
        readonly,
        required,
        label,
        hideLabel: !displayLabel,
        placeholder,
        autofocus,
        rawErrors,
        htmlName: multiValueFieldPathId.name
      }
    );
  }
  function ArrayAsCustomWidget(props) {
    const {
      schema,
      fieldPathId,
      uiSchema,
      disabled = false,
      readonly = false,
      autofocus = false,
      required = false,
      hideError,
      placeholder,
      onBlur,
      onFocus,
      formData: items = [],
      registry,
      rawErrors,
      name,
      onSelectChange
    } = props;
    const { widgets: widgets2, schemaUtils, globalFormOptions, globalUiOptions } = registry;
    const { widget, title: uiTitle, ...options } = utils.getUiOptions(uiSchema, globalUiOptions);
    const Widget = utils.getWidget(schema, widget, widgets2);
    const label = uiTitle ?? schema.title ?? name;
    const displayLabel = schemaUtils.getDisplayLabel(schema, uiSchema, globalUiOptions);
    const multiValueFieldPathId = utils.useDeepCompareMemo(utils.toFieldPathId("", globalFormOptions, fieldPathId, true));
    return /* @__PURE__ */ jsxRuntime.jsx(
      Widget,
      {
        id: multiValueFieldPathId[utils.ID_KEY],
        name,
        multiple: true,
        onChange: onSelectChange,
        onBlur,
        onFocus,
        options,
        schema,
        uiSchema,
        registry,
        value: items,
        disabled,
        readonly,
        hideError,
        required,
        label,
        hideLabel: !displayLabel,
        placeholder,
        autofocus,
        rawErrors,
        htmlName: multiValueFieldPathId.name
      }
    );
  }
  function ArrayAsFiles(props) {
    const {
      schema,
      uiSchema,
      fieldPathId,
      name,
      disabled = false,
      readonly = false,
      autofocus = false,
      required = false,
      onBlur,
      onFocus,
      registry,
      formData: items = [],
      rawErrors,
      onSelectChange
    } = props;
    const { widgets: widgets2, schemaUtils, globalFormOptions, globalUiOptions } = registry;
    const { widget = "files", title: uiTitle, ...options } = utils.getUiOptions(uiSchema, globalUiOptions);
    const Widget = utils.getWidget(schema, widget, widgets2);
    const label = uiTitle ?? schema.title ?? name;
    const displayLabel = schemaUtils.getDisplayLabel(schema, uiSchema, globalUiOptions);
    const multiValueFieldPathId = utils.useDeepCompareMemo(utils.toFieldPathId("", globalFormOptions, fieldPathId, true));
    return /* @__PURE__ */ jsxRuntime.jsx(
      Widget,
      {
        options,
        id: multiValueFieldPathId[utils.ID_KEY],
        name,
        multiple: true,
        onChange: onSelectChange,
        onBlur,
        onFocus,
        schema,
        uiSchema,
        value: items,
        disabled,
        readonly,
        required,
        registry,
        autofocus,
        rawErrors,
        label,
        hideLabel: !displayLabel,
        htmlName: multiValueFieldPathId.name
      }
    );
  }
  function ArrayFieldItemInner(props) {
    const {
      itemKey,
      index,
      name,
      disabled,
      hideError,
      readonly,
      registry,
      uiOptions,
      parentUiSchema,
      canAdd,
      canRemove = true,
      canMoveUp,
      canMoveDown,
      rawItemSchema,
      itemData,
      itemUiSchema,
      parentFieldPathId,
      itemErrorSchema,
      autofocus,
      onBlur,
      onFocus,
      onChange,
      rawErrors,
      totalItems,
      title,
      handleAddItem,
      handleCopyItem,
      handleRemoveItem,
      handleReorderItems
    } = props;
    const {
      schemaUtils,
      fields: { ArraySchemaField, SchemaField: SchemaField2 },
      globalUiOptions,
      globalFormOptions
    } = registry;
    const itemSchema = react.useMemo(
      () => schemaUtils.retrieveSchema(rawItemSchema, itemData),
      [schemaUtils, rawItemSchema, itemData]
    );
    const fieldPathId = react.useMemo(
      () => utils.toFieldPathId(index, globalFormOptions, parentFieldPathId),
      [index, globalFormOptions, parentFieldPathId]
    );
    const ItemSchemaField = ArraySchemaField || SchemaField2;
    const ArrayFieldItemTemplate2 = utils.getTemplate(
      "ArrayFieldItemTemplate",
      registry,
      uiOptions
    );
    const displayLabel = schemaUtils.getDisplayLabel(itemSchema, itemUiSchema, globalUiOptions);
    const { description } = utils.getUiOptions(itemUiSchema);
    const hasDescription = !!description || !!itemSchema.description;
    const { orderable = true, removable = true, copyable = false } = uiOptions;
    const has4 = {
      moveUp: orderable && canMoveUp,
      moveDown: orderable && canMoveDown,
      copy: copyable && canAdd,
      remove: removable && canRemove,
      toolbar: false
    };
    has4.toolbar = Object.keys(has4).some((key) => has4[key]);
    const onAddItem = react.useCallback(
      (event) => {
        handleAddItem(event, index + 1);
      },
      [handleAddItem, index]
    );
    const onCopyItem = react.useCallback(
      (event) => {
        handleCopyItem(event, index);
      },
      [handleCopyItem, index]
    );
    const onRemoveItem = react.useCallback(
      (event) => {
        handleRemoveItem(event, index);
      },
      [handleRemoveItem, index]
    );
    const onMoveUpItem = react.useCallback(
      (event) => {
        handleReorderItems(event, index, index - 1);
      },
      [handleReorderItems, index]
    );
    const onMoveDownItem = react.useCallback(
      (event) => {
        handleReorderItems(event, index, index + 1);
      },
      [handleReorderItems, index]
    );
    const templateProps = {
      children: /* @__PURE__ */ jsxRuntime.jsx(
        ItemSchemaField,
        {
          name,
          title,
          index,
          schema: itemSchema,
          uiSchema: itemUiSchema,
          formData: itemData,
          errorSchema: itemErrorSchema,
          fieldPathId,
          required: isItemRequired(itemSchema),
          onChange,
          onBlur,
          onFocus,
          registry,
          disabled,
          readonly,
          hideError,
          autofocus,
          rawErrors
        }
      ),
      buttonsProps: {
        fieldPathId,
        disabled,
        readonly,
        canAdd,
        hasCopy: has4.copy,
        hasMoveUp: has4.moveUp,
        hasMoveDown: has4.moveDown,
        hasRemove: has4.remove,
        index,
        totalItems,
        onAddItem,
        onCopyItem,
        onRemoveItem,
        onMoveUpItem,
        onMoveDownItem,
        registry,
        schema: itemSchema,
        uiSchema: itemUiSchema
      },
      itemKey,
      className: "rjsf-array-item",
      disabled,
      hasToolbar: has4.toolbar,
      index,
      totalItems,
      readonly,
      registry,
      schema: itemSchema,
      uiSchema: itemUiSchema,
      parentUiSchema,
      displayLabel,
      hasDescription
    };
    return /* @__PURE__ */ jsxRuntime.jsx(ArrayFieldItemTemplate2, { ...templateProps });
  }
  var ArrayFieldItem = react.memo(ArrayFieldItemInner);
  function NormalArray(props) {
    const {
      schema,
      uiSchema = {},
      errorSchema,
      fieldPathId,
      formData: formDataFromProps,
      name,
      title,
      disabled = false,
      readonly = false,
      autofocus = false,
      required = false,
      hideError = false,
      registry,
      onBlur,
      onFocus,
      rawErrors,
      onChange,
      keyedFormData,
      handleAddItem,
      handleCopyItem,
      handleRemoveItem,
      handleReorderItems
    } = props;
    const fieldTitle = schema.title || title || name;
    const { schemaUtils, fields: fields2, formContext, globalUiOptions } = registry;
    const { OptionalDataControlsField: OptionalDataControlsField2 } = fields2;
    const uiOptions = react.useMemo(() => utils.getUiOptions(uiSchema, globalUiOptions), [uiSchema, globalUiOptions]);
    const schemaItems = react.useMemo(() => isObject(schema.items) ? schema.items : {}, [schema.items]);
    const itemsSchema = react.useMemo(() => schemaUtils.retrieveSchema(schemaItems), [schemaUtils, schemaItems]);
    const formData = react.useMemo(() => keyedToPlainFormData(keyedFormData), [keyedFormData]);
    const renderOptionalField = utils.shouldRenderOptionalField(registry, schema, required, uiSchema);
    const hasFormData = utils.isFormDataAvailable(formDataFromProps);
    const canAdd = react.useMemo(
      () => canAddItem(registry, schema, formData, uiSchema) && (!renderOptionalField || hasFormData),
      [registry, schema, formData, uiSchema, renderOptionalField, hasFormData]
    );
    const actualFormData = hasFormData ? keyedFormData : [];
    const extraClass = renderOptionalField ? " rjsf-optional-array-field" : "";
    const childFieldPathId = props.childFieldPathId ?? fieldPathId;
    const optionalDataControl = renderOptionalField ? /* @__PURE__ */ jsxRuntime.jsx(OptionalDataControlsField2, { ...props, fieldPathId: childFieldPathId }) : void 0;
    const arrayProps = {
      canAdd,
      items: actualFormData.map((keyedItem, index) => {
        const { key, item } = keyedItem;
        const itemCast = item;
        const itemErrorSchema = errorSchema ? errorSchema[index] : void 0;
        const itemUiSchema = computeItemUiSchema(uiSchema, item, index, formContext);
        const itemProps = {
          itemKey: key,
          index,
          name: name && `${name}-${index}`,
          registry,
          uiOptions,
          parentUiSchema: uiSchema,
          hideError,
          readonly,
          disabled,
          required,
          title: fieldTitle ? `${fieldTitle}-${index + 1}` : void 0,
          canAdd,
          canMoveUp: index > 0,
          canMoveDown: index < formData.length - 1,
          rawItemSchema: schemaItems,
          parentFieldPathId: childFieldPathId,
          itemErrorSchema,
          itemData: itemCast,
          itemUiSchema,
          autofocus: autofocus && index === 0,
          onBlur,
          onFocus,
          rawErrors,
          totalItems: keyedFormData.length,
          handleAddItem,
          handleCopyItem,
          handleRemoveItem,
          handleReorderItems,
          onChange
        };
        return /* @__PURE__ */ jsxRuntime.jsx(ArrayFieldItem, { ...itemProps }, key);
      }),
      className: `rjsf-field rjsf-field-array rjsf-field-array-of-${itemsSchema.type}${extraClass}`,
      disabled,
      fieldPathId,
      uiSchema,
      onAddClick: handleAddItem,
      readonly,
      required,
      schema,
      title: fieldTitle,
      formData,
      rawErrors,
      registry,
      optionalDataControl
    };
    const Template = utils.getTemplate("ArrayFieldTemplate", registry, uiOptions);
    return /* @__PURE__ */ jsxRuntime.jsx(Template, { ...arrayProps });
  }
  function FixedArray(props) {
    const {
      schema,
      uiSchema = {},
      formData,
      errorSchema,
      fieldPathId,
      name,
      title,
      disabled = false,
      readonly = false,
      autofocus = false,
      required = false,
      hideError = false,
      registry,
      onBlur,
      onFocus,
      rawErrors,
      keyedFormData,
      onChange,
      handleAddItem,
      handleCopyItem,
      handleRemoveItem,
      handleReorderItems
    } = props;
    let { formData: items = [] } = props;
    const fieldTitle = schema.title || title || name;
    const { fields: fields2, formContext, globalUiOptions } = registry;
    const uiOptions = react.useMemo(() => utils.getUiOptions(uiSchema, globalUiOptions), [uiSchema, globalUiOptions]);
    const { OptionalDataControlsField: OptionalDataControlsField2 } = fields2;
    const renderOptionalField = utils.shouldRenderOptionalField(registry, schema, required, uiSchema);
    const hasFormData = utils.isFormDataAvailable(formData);
    const schemaItems = react.useMemo(
      () => isObject(schema.items) ? schema.items : [],
      [schema.items]
    );
    const hasAdditionalItems = isObject(schema.additionalItems);
    const childFieldPathId = props.childFieldPathId ?? fieldPathId;
    if (items.length < schemaItems.length) {
      items = items.concat(new Array(schemaItems.length - items.length));
    }
    const actualFormData = hasFormData ? keyedFormData : [];
    const extraClass = renderOptionalField ? " rjsf-optional-array-field" : "";
    const optionalDataControl = renderOptionalField ? /* @__PURE__ */ jsxRuntime.jsx(OptionalDataControlsField2, { ...props, fieldPathId: childFieldPathId }) : void 0;
    const canAdd = canAddItem(registry, schema, items, uiSchema) && hasAdditionalItems && (!renderOptionalField || hasFormData);
    const arrayProps = {
      canAdd,
      className: `rjsf-field rjsf-field-array rjsf-field-array-fixed-items${extraClass}`,
      disabled,
      fieldPathId,
      formData,
      items: actualFormData.map((keyedItem, index) => {
        const { key, item } = keyedItem;
        const itemCast = item;
        const additional = index >= schemaItems.length;
        const rawItemSchema = (additional && isObject(schema.additionalItems) ? schema.additionalItems : schemaItems[index]) || {};
        let itemUiSchema;
        if (additional) {
          itemUiSchema = uiSchema.additionalItems;
        } else if (Array.isArray(uiSchema.items)) {
          itemUiSchema = uiSchema.items[index];
        } else {
          itemUiSchema = computeItemUiSchema(uiSchema, item, index, formContext);
        }
        const itemErrorSchema = errorSchema ? errorSchema[index] : void 0;
        const itemProps = {
          index,
          itemKey: key,
          name: name && `${name}-${index}`,
          registry,
          uiOptions,
          parentUiSchema: uiSchema,
          hideError,
          readonly,
          disabled,
          required,
          title: fieldTitle ? `${fieldTitle}-${index + 1}` : void 0,
          canAdd,
          canRemove: additional,
          canMoveUp: index >= schemaItems.length + 1,
          canMoveDown: additional && index < items.length - 1,
          rawItemSchema,
          itemData: itemCast,
          itemUiSchema,
          parentFieldPathId: childFieldPathId,
          itemErrorSchema,
          autofocus: autofocus && index === 0,
          onBlur,
          onFocus,
          rawErrors,
          totalItems: keyedFormData.length,
          onChange,
          handleAddItem,
          handleCopyItem,
          handleRemoveItem,
          handleReorderItems
        };
        return /* @__PURE__ */ jsxRuntime.jsx(ArrayFieldItem, { ...itemProps }, key);
      }),
      onAddClick: handleAddItem,
      readonly,
      required,
      registry,
      schema,
      uiSchema,
      title: fieldTitle,
      errorSchema,
      rawErrors,
      optionalDataControl
    };
    const Template = utils.getTemplate("ArrayFieldTemplate", registry, uiOptions);
    return /* @__PURE__ */ jsxRuntime.jsx(Template, { ...arrayProps });
  }
  function useKeyedFormData(formData = []) {
    const newHash = react.useMemo(() => utils.hashObject(formData), [formData]);
    const [state, setState] = react.useState(() => ({
      formDataHash: newHash,
      keyedFormData: generateKeyedFormData(formData)
    }));
    let { keyedFormData, formDataHash } = state;
    if (newHash !== formDataHash) {
      const nextFormData = Array.isArray(formData) ? formData : [];
      const previousKeyedFormData = keyedFormData || [];
      keyedFormData = nextFormData.length === previousKeyedFormData.length ? previousKeyedFormData.map((previousKeyedFormDatum, index) => ({
        key: previousKeyedFormDatum.key,
        item: nextFormData[index]
      })) : generateKeyedFormData(nextFormData);
      formDataHash = newHash;
      setState({ formDataHash, keyedFormData });
    }
    const updateKeyedFormData = react.useCallback((newData) => {
      const plainFormData = keyedToPlainFormData(newData);
      const updatedHash = utils.hashObject(plainFormData);
      setState({ formDataHash: updatedHash, keyedFormData: newData });
      return plainFormData;
    }, []);
    return { keyedFormData, updateKeyedFormData };
  }
  function ArrayField(props) {
    const { schema, uiSchema, errorSchema, fieldPathId, registry, formData, onChange } = props;
    const { globalFormOptions, schemaUtils, translateString } = registry;
    const { keyedFormData, updateKeyedFormData } = useKeyedFormData(formData);
    const keyedFormDataRef = react.useRef(keyedFormData);
    keyedFormDataRef.current = keyedFormData;
    const errorSchemaRef = react.useRef(errorSchema);
    errorSchemaRef.current = errorSchema;
    const childFieldPathId = props.childFieldPathId ?? fieldPathId;
    const handleAddItem = react.useCallback(
      (event, index) => {
        if (event) {
          event.preventDefault();
        }
        let newErrorSchema;
        if (errorSchemaRef.current) {
          newErrorSchema = {};
          for (const idx of Object.keys(errorSchemaRef.current)) {
            const i = parseInt(idx, 10);
            if (index === void 0 || i < index) {
              set(newErrorSchema, [i], errorSchemaRef.current[i]);
            } else if (i >= index) {
              set(newErrorSchema, [i + 1], errorSchemaRef.current[i]);
            }
          }
        }
        const newKeyedFormDataRow = {
          key: generateRowId(),
          item: getNewFormDataRow(registry, schema)
        };
        const newKeyedFormData = [...keyedFormDataRef.current];
        if (index !== void 0) {
          newKeyedFormData.splice(index, 0, newKeyedFormDataRow);
        } else {
          newKeyedFormData.push(newKeyedFormDataRow);
        }
        onChange(updateKeyedFormData(newKeyedFormData), childFieldPathId.path, newErrorSchema);
      },
      [registry, schema, onChange, updateKeyedFormData, childFieldPathId]
    );
    const handleCopyItem = react.useCallback(
      (event, index) => {
        if (event) {
          event.preventDefault();
        }
        let newErrorSchema;
        if (errorSchemaRef.current) {
          newErrorSchema = {};
          for (const idx of Object.keys(errorSchemaRef.current)) {
            const i = parseInt(idx, 10);
            if (i <= index) {
              set(newErrorSchema, [i], errorSchemaRef.current[i]);
            } else if (i > index) {
              set(newErrorSchema, [i + 1], errorSchemaRef.current[i]);
            }
          }
        }
        const newKeyedFormDataRow = {
          key: generateRowId(),
          item: _cloneDeep(keyedFormDataRef.current[index].item)
        };
        const newKeyedFormData = [...keyedFormDataRef.current];
        if (index !== void 0) {
          newKeyedFormData.splice(index + 1, 0, newKeyedFormDataRow);
        } else {
          newKeyedFormData.push(newKeyedFormDataRow);
        }
        onChange(updateKeyedFormData(newKeyedFormData), childFieldPathId.path, newErrorSchema);
      },
      [onChange, updateKeyedFormData, childFieldPathId]
    );
    const handleRemoveItem = react.useCallback(
      (event, index) => {
        if (event) {
          event.preventDefault();
        }
        let newErrorSchema;
        if (errorSchemaRef.current) {
          newErrorSchema = {};
          for (const idx of Object.keys(errorSchemaRef.current)) {
            const i = parseInt(idx, 10);
            if (i < index) {
              set(newErrorSchema, [i], errorSchemaRef.current[i]);
            } else if (i > index) {
              set(newErrorSchema, [i - 1], errorSchemaRef.current[i]);
            }
          }
        }
        const newKeyedFormData = keyedFormDataRef.current.filter((_, i) => i !== index);
        onChange(updateKeyedFormData(newKeyedFormData), childFieldPathId.path, newErrorSchema);
      },
      [onChange, updateKeyedFormData, childFieldPathId]
    );
    const handleReorderItems = react.useCallback(
      (event, index, newIndex) => {
        if (event) {
          event.preventDefault();
          event.currentTarget.blur();
        }
        let newErrorSchema;
        if (errorSchemaRef.current) {
          newErrorSchema = {};
          for (const idx of Object.keys(errorSchemaRef.current)) {
            const i = parseInt(idx, 10);
            if (i === index) {
              set(newErrorSchema, [newIndex], errorSchemaRef.current[index]);
            } else if (i === newIndex) {
              set(newErrorSchema, [index], errorSchemaRef.current[newIndex]);
            } else {
              set(newErrorSchema, [idx], errorSchemaRef.current[i]);
            }
          }
        }
        function reOrderArray() {
          const newKeyedFormData2 = keyedFormDataRef.current.slice();
          newKeyedFormData2.splice(index, 1);
          newKeyedFormData2.splice(newIndex, 0, keyedFormDataRef.current[index]);
          return newKeyedFormData2;
        }
        const newKeyedFormData = reOrderArray();
        onChange(updateKeyedFormData(newKeyedFormData), childFieldPathId.path, newErrorSchema);
      },
      [onChange, updateKeyedFormData, childFieldPathId]
    );
    const handleChange = react.useCallback(
      (value, path, newErrorSchema, id) => {
        const lastPathIsItemIndex = typeof path.at(-1) === "number";
        onChange(
          // We need to treat undefined items as nulls to have validation.
          // See https://github.com/tdegrunt/jsonschema/issues/206
          // Only set to null for array items, and not for object properties within array items
          lastPathIsItemIndex && value === void 0 ? null : value,
          path,
          newErrorSchema,
          id
        );
      },
      [onChange]
    );
    const onSelectChange = react.useCallback(
      (value) => {
        onChange(value, childFieldPathId.path, void 0, childFieldPathId?.[utils.ID_KEY]);
      },
      [onChange, childFieldPathId]
    );
    const arrayAsMultiProps = {
      ...props,
      formData,
      fieldPathId: childFieldPathId,
      onSelectChange
    };
    const arrayProps = {
      ...props,
      handleAddItem,
      handleCopyItem,
      handleRemoveItem,
      handleReorderItems,
      keyedFormData,
      onChange: handleChange
    };
    if (!(utils.ITEMS_KEY in schema)) {
      if (!globalFormOptions.useFallbackUiForUnsupportedType) {
        const uiOptions = utils.getUiOptions(uiSchema);
        const UnsupportedFieldTemplate = utils.getTemplate(
          "UnsupportedFieldTemplate",
          registry,
          uiOptions
        );
        return /* @__PURE__ */ jsxRuntime.jsx(
          UnsupportedFieldTemplate,
          {
            schema,
            fieldPathId,
            reason: translateString(utils.TranslatableString.MissingItems),
            registry
          }
        );
      }
      const fallbackSchema = { ...schema, [utils.ITEMS_KEY]: { type: void 0 } };
      arrayAsMultiProps.schema = fallbackSchema;
      arrayProps.schema = fallbackSchema;
    }
    if (schemaUtils.isMultiSelect(arrayAsMultiProps.schema)) {
      return /* @__PURE__ */ jsxRuntime.jsx(ArrayAsMultiSelect, { ...arrayAsMultiProps });
    }
    if (utils.isCustomWidget(uiSchema)) {
      return /* @__PURE__ */ jsxRuntime.jsx(ArrayAsCustomWidget, { ...arrayAsMultiProps });
    }
    if (utils.isFixedItems(arrayAsMultiProps.schema)) {
      return /* @__PURE__ */ jsxRuntime.jsx(FixedArray, { ...arrayProps });
    }
    if (schemaUtils.isFilesArray(arrayAsMultiProps.schema, uiSchema)) {
      return /* @__PURE__ */ jsxRuntime.jsx(ArrayAsFiles, { ...arrayAsMultiProps });
    }
    return /* @__PURE__ */ jsxRuntime.jsx(NormalArray, { ...arrayProps });
  }
  function BooleanField(props) {
    const {
      schema,
      name,
      uiSchema,
      fieldPathId,
      formData,
      registry,
      required,
      disabled,
      readonly,
      hideError,
      autofocus,
      title,
      onChange,
      onFocus,
      onBlur,
      rawErrors
    } = props;
    const { title: schemaTitle } = schema;
    const { widgets: widgets2, translateString, globalUiOptions } = registry;
    const {
      widget = "checkbox",
      title: uiTitle,
      // Unlike the other fields, don't use `getDisplayLabel()` since it always returns false for the boolean type
      label: displayLabel = true,
      enumNames,
      ...options
    } = utils.getUiOptions(uiSchema, globalUiOptions);
    const Widget = utils.getWidget(schema, widget, widgets2);
    const yes = translateString(utils.TranslatableString.YesLabel);
    const no = translateString(utils.TranslatableString.NoLabel);
    let enumOptions;
    const label = uiTitle ?? schemaTitle ?? title ?? name;
    if (Array.isArray(schema.oneOf)) {
      enumOptions = utils.optionsList(
        {
          oneOf: schema.oneOf.map((option) => {
            if (isObject(option)) {
              return {
                ...option,
                title: option.title || (option.const === true ? yes : no)
              };
            }
            return void 0;
          }).filter((o) => o)
          // cast away the error that typescript can't grok is fixed
        },
        uiSchema
      );
    } else {
      const enums = schema.enum ?? [true, false];
      if (!enumNames && enums.length === 2 && enums.every((v) => typeof v === "boolean")) {
        enumOptions = [
          {
            value: enums[0],
            label: enums[0] ? yes : no
          },
          {
            value: enums[1],
            label: enums[1] ? yes : no
          }
        ];
      } else {
        enumOptions = utils.optionsList({ enum: enums }, uiSchema);
      }
    }
    const onWidgetChange = react.useCallback(
      (value, errorSchema, id) => onChange(value, fieldPathId.path, errorSchema, id),
      [onChange, fieldPathId]
    );
    return /* @__PURE__ */ jsxRuntime.jsx(
      Widget,
      {
        options: { ...options, enumOptions },
        schema,
        uiSchema,
        id: fieldPathId.$id,
        name,
        onChange: onWidgetChange,
        onFocus,
        onBlur,
        label,
        hideLabel: !displayLabel,
        value: formData,
        required,
        disabled,
        readonly,
        hideError,
        registry,
        autofocus,
        rawErrors,
        htmlName: fieldPathId.name
      }
    );
  }
  var BooleanField_default = BooleanField;
  function getFallbackTypeSelectionSchema(title) {
    return {
      type: "string",
      enum: ["string", "number", "boolean", "object", "array"],
      default: "string",
      title
    };
  }
  function getTypeOfFormData(formData) {
    const dataType = typeof formData;
    if (dataType === "string" || dataType === "number" || dataType === "boolean") {
      return dataType;
    }
    if (dataType === "object") {
      return Array.isArray(formData) ? "array" : "object";
    }
    return "string";
  }
  function castToNewType(formData, newType) {
    switch (newType) {
      case "string":
        return String(formData);
      case "number": {
        const castedNumber = Number(formData);
        return Number.isNaN(castedNumber) ? 0 : castedNumber;
      }
      case "boolean":
        return Boolean(formData);
      default:
        return formData;
    }
  }
  function FallbackField(props) {
    const {
      id,
      formData,
      displayLabel = true,
      schema,
      name,
      uiSchema,
      required,
      disabled = false,
      readonly = false,
      onBlur,
      onFocus,
      registry,
      fieldPathId,
      onChange,
      errorSchema
    } = props;
    const { translateString, fields: fields2, globalFormOptions } = registry;
    const [type, setType] = react.useState(getTypeOfFormData(formData));
    const uiOptions = utils.getUiOptions(uiSchema);
    const typeSelectorInnerFieldPathId = utils.useDeepCompareMemo(
      utils.toFieldPathId("__internal_type_selector", globalFormOptions, fieldPathId)
    );
    const schemaTitle = translateString(utils.TranslatableString.Type);
    const typesOptionSchema = react.useMemo(() => getFallbackTypeSelectionSchema(schemaTitle), [schemaTitle]);
    const onTypeChange = (newType) => {
      if (newType != null) {
        setType(newType);
        onChange(castToNewType(formData, newType), fieldPathId.path, errorSchema, id);
      }
    };
    if (!globalFormOptions.useFallbackUiForUnsupportedType) {
      const { reason = translateString(utils.TranslatableString.UnknownFieldType, [String(schema.type)]) } = props;
      const UnsupportedFieldTemplate = utils.getTemplate(
        "UnsupportedFieldTemplate",
        registry,
        uiOptions
      );
      return /* @__PURE__ */ jsxRuntime.jsx(UnsupportedFieldTemplate, { schema, fieldPathId, reason, registry });
    }
    const FallbackFieldTemplate2 = utils.getTemplate(
      "FallbackFieldTemplate",
      registry,
      uiOptions
    );
    const { SchemaField: SchemaField2 } = fields2;
    return /* @__PURE__ */ jsxRuntime.jsx(
      FallbackFieldTemplate2,
      {
        schema,
        registry,
        typeSelector: /* @__PURE__ */ jsxRuntime.jsx(
          SchemaField2,
          {
            fieldPathId: typeSelectorInnerFieldPathId,
            name: `${name}__fallback_type`,
            schema: typesOptionSchema,
            formData: type,
            onChange: onTypeChange,
            onBlur,
            onFocus,
            registry,
            hideLabel: !displayLabel,
            disabled,
            readonly,
            required
          },
          formData ? utils.hashObject(formData) : "__empty__"
        ),
        schemaField: /* @__PURE__ */ jsxRuntime.jsx(
          SchemaField2,
          {
            ...props,
            schema: {
              type,
              title: translateString(utils.TranslatableString.Value),
              ...type === "object" && { additionalProperties: true }
            }
          }
        )
      }
    );
  }
  var LOOKUP_REGEX = /^\$lookup=(.+)/;
  var LAYOUT_GRID_UI_OPTION = "layoutGrid";
  function getNonNullishValue(value, fallback) {
    return value ?? fallback;
  }
  function isNumericIndex(str) {
    return /^\d+?$/.test(str);
  }
  var LAYOUT_GRID_FIELD_TEST_IDS = utils.getTestIds();
  function computeFieldUiSchema(field, uiProps, uiSchema, schemaReadonly, forceReadonly) {
    const globalUiOptions = get(uiSchema, [utils.UI_GLOBAL_OPTIONS_KEY], {});
    const localUiSchema = get(uiSchema, field);
    const localUiOptions = { ...get(localUiSchema, [utils.UI_OPTIONS_KEY], {}), ...uiProps, ...globalUiOptions };
    const fieldUiSchema = { ...localUiSchema };
    if (!isEmpty(localUiOptions)) {
      set(fieldUiSchema, [utils.UI_OPTIONS_KEY], localUiOptions);
    }
    if (!isEmpty(globalUiOptions)) {
      set(fieldUiSchema, [utils.UI_GLOBAL_OPTIONS_KEY], globalUiOptions);
    }
    let { readonly: uiReadonly } = utils.getUiOptions(fieldUiSchema);
    if (forceReadonly === true || isUndefined(uiReadonly) && schemaReadonly === true) {
      uiReadonly = true;
      if (has(localUiOptions, utils.READONLY_KEY)) {
        set(fieldUiSchema, [utils.UI_OPTIONS_KEY, utils.READONLY_KEY], true);
      } else {
        set(fieldUiSchema, `ui:${utils.READONLY_KEY}`, true);
      }
    }
    return { fieldUiSchema, uiReadonly };
  }
  function conditionMatches(operator, datum, value = "$0m3tH1nG Un3xP3cT3d") {
    const data = flatten([datum]).sort();
    const values = flatten([value]).sort();
    switch (operator) {
      case "all" /* ALL */:
        return isEqual(data, values);
      case "some" /* SOME */:
        return intersection(data, values).length > 0;
      case "none" /* NONE */:
        return intersection(data, values).length === 0;
      default:
        return false;
    }
  }
  function findChildrenAndProps(layoutGridSchema, schemaKey, registry) {
    let gridProps = {};
    let children = layoutGridSchema[schemaKey];
    if (isPlainObject(children)) {
      const { children: elements, className: toMapClassNames, ...otherProps } = children;
      children = elements;
      if (toMapClassNames) {
        const classes = toMapClassNames.split(" ");
        const className = classes.map((ele) => utils.lookupFromFormContext(registry, ele, ele)).join(" ");
        gridProps = { ...otherProps, className };
      } else {
        gridProps = otherProps;
      }
    }
    if (!Array.isArray(children)) {
      throw new TypeError(`Expected array for "${schemaKey}" in ${JSON.stringify(layoutGridSchema)}`);
    }
    return { children, gridProps };
  }
  function computeArraySchemasIfPresent(schema, fieldPathId, potentialIndex) {
    let rawSchema;
    let resultPathId = fieldPathId;
    if (isNumericIndex(potentialIndex) && schema && schema?.type === "array" && has(schema, utils.ITEMS_KEY)) {
      const index = Number(potentialIndex);
      const items = schema[utils.ITEMS_KEY];
      if (Array.isArray(items)) {
        if (index > items.length) {
          rawSchema = last(items);
        } else {
          rawSchema = items[index];
        }
      } else {
        rawSchema = items;
      }
      resultPathId = {
        [utils.ID_KEY]: fieldPathId[utils.ID_KEY],
        path: [...fieldPathId.path.slice(0, fieldPathId.path.length - 1), index]
      };
    }
    return { rawSchema, fieldPathId: resultPathId };
  }
  function getSchemaDetailsForField(registry, dottedPath, initialSchema, formData, initialFieldIdPath) {
    const { schemaUtils, globalFormOptions } = registry;
    let rawSchema = initialSchema;
    let fieldPathId = initialFieldIdPath;
    const parts = dottedPath.split(".");
    const leafPath = parts.pop();
    let schema = schemaUtils.retrieveSchema(rawSchema, formData);
    let innerData = formData;
    let isReadonly = schema.readOnly;
    parts.forEach((part) => {
      fieldPathId = utils.toFieldPathId(part, globalFormOptions, fieldPathId);
      if (has(schema, utils.PROPERTIES_KEY)) {
        rawSchema = get(schema, [utils.PROPERTIES_KEY, part], {});
      } else if (schema && (has(schema, utils.ONE_OF_KEY) || has(schema, utils.ANY_OF_KEY))) {
        const xxx = has(schema, utils.ONE_OF_KEY) ? utils.ONE_OF_KEY : utils.ANY_OF_KEY;
        const selectedSchema = schemaUtils.findSelectedOptionInXxxOf(schema, part, xxx, innerData);
        rawSchema = get(selectedSchema, [utils.PROPERTIES_KEY, part], {});
      } else {
        const result = computeArraySchemasIfPresent(schema, fieldPathId, part);
        rawSchema = result.rawSchema ?? {};
        fieldPathId = result.fieldPathId;
      }
      innerData = get(innerData, part, {});
      schema = schemaUtils.retrieveSchema(rawSchema, innerData);
      isReadonly = getNonNullishValue(schema.readOnly, isReadonly);
    });
    let optionsInfo;
    let isRequired2 = false;
    if (isEmpty(schema)) {
      schema = void 0;
    }
    if (schema && leafPath) {
      if (schema && (has(schema, utils.ONE_OF_KEY) || has(schema, utils.ANY_OF_KEY))) {
        const xxx = has(schema, utils.ONE_OF_KEY) ? utils.ONE_OF_KEY : utils.ANY_OF_KEY;
        schema = schemaUtils.findSelectedOptionInXxxOf(schema, leafPath, xxx, innerData);
      }
      fieldPathId = utils.toFieldPathId(leafPath, globalFormOptions, fieldPathId);
      isRequired2 = schema !== void 0 && Array.isArray(schema.required) && includes(schema.required, leafPath);
      const result = computeArraySchemasIfPresent(schema, fieldPathId, leafPath);
      if (result.rawSchema) {
        schema = result.rawSchema;
        fieldPathId = result.fieldPathId;
      } else {
        schema = get(schema, [utils.PROPERTIES_KEY, leafPath]);
        schema = schema ? schemaUtils.retrieveSchema(schema) : schema;
      }
      isReadonly = getNonNullishValue(schema?.readOnly, isReadonly);
      if (schema && (has(schema, utils.ONE_OF_KEY) || has(schema, utils.ANY_OF_KEY))) {
        const xxx = has(schema, utils.ONE_OF_KEY) ? utils.ONE_OF_KEY : utils.ANY_OF_KEY;
        const discriminator = utils.getDiscriminatorFieldFromSchema(schema);
        optionsInfo = { options: schema[xxx], hasDiscriminator: !!discriminator };
      }
    }
    return { schema, isRequired: isRequired2, isReadonly, optionsInfo, fieldPathId };
  }
  function getCustomRenderComponent(render, registry) {
    let customRenderer = render;
    if (isString(customRenderer)) {
      customRenderer = utils.lookupFromFormContext(registry, customRenderer);
    }
    if (isFunction(customRenderer)) {
      return customRenderer;
    }
    return null;
  }
  function computeUIComponentPropsFromGridSchema(registry, gridSchema) {
    let name;
    let UIComponent = null;
    let uiProps = {};
    let rendered;
    if (isString(gridSchema) || isUndefined(gridSchema)) {
      name = gridSchema ?? "";
    } else {
      const { name: innerName = "", render, ...innerProps } = gridSchema;
      name = innerName;
      uiProps = innerProps;
      if (!isEmpty(uiProps)) {
        each(uiProps, (prop, key) => {
          if (isString(prop)) {
            const match = LOOKUP_REGEX.exec(prop);
            if (Array.isArray(match) && match.length > 1) {
              const lookupName = match[1];
              uiProps[key] = utils.lookupFromFormContext(registry, lookupName, lookupName);
            }
          }
        });
      }
      UIComponent = getCustomRenderComponent(render, registry);
      if (!innerName && UIComponent) {
        rendered = /* @__PURE__ */ jsxRuntime.jsx(UIComponent, { ...innerProps, "data-testid": LAYOUT_GRID_FIELD_TEST_IDS.uiComponent });
      }
    }
    return { name, UIComponent, uiProps, rendered };
  }
  function LayoutGridFieldChildren(props) {
    const { childrenLayoutGridSchemaId, ...layoutGridFieldProps } = props;
    const { registry, schema: rawSchema, formData } = layoutGridFieldProps;
    const { schemaUtils } = registry;
    const schema = schemaUtils.retrieveSchema(rawSchema, formData);
    return childrenLayoutGridSchemaId.map((layoutGridSchema) => /* @__PURE__ */ react.createElement(
      LayoutGridField,
      {
        ...layoutGridFieldProps,
        key: `layoutGrid-${utils.hashObject(layoutGridSchema)}`,
        schema,
        layoutGridSchema
      }
    ));
  }
  function LayoutGridCondition(props) {
    const { layoutGridSchema, ...layoutGridFieldProps } = props;
    const { formData, registry } = layoutGridFieldProps;
    const { children, gridProps } = findChildrenAndProps(layoutGridSchema, "ui:condition" /* CONDITION */, registry);
    const { operator, field = "", value } = gridProps;
    const fieldData = get(formData, field, null);
    if (conditionMatches(operator, fieldData, value)) {
      return /* @__PURE__ */ jsxRuntime.jsx(LayoutGridFieldChildren, { ...layoutGridFieldProps, childrenLayoutGridSchemaId: children });
    }
    return null;
  }
  function LayoutGridCol(props) {
    const { layoutGridSchema, ...layoutGridFieldProps } = props;
    const { registry, uiSchema } = layoutGridFieldProps;
    const { children, gridProps } = findChildrenAndProps(layoutGridSchema, "ui:col" /* COLUMN */, registry);
    const uiOptions = utils.getUiOptions(uiSchema);
    const GridTemplate2 = utils.getTemplate("GridTemplate", registry, uiOptions);
    return /* @__PURE__ */ jsxRuntime.jsx(GridTemplate2, { column: true, "data-testid": LAYOUT_GRID_FIELD_TEST_IDS.col, ...gridProps, children: /* @__PURE__ */ jsxRuntime.jsx(LayoutGridFieldChildren, { ...layoutGridFieldProps, childrenLayoutGridSchemaId: children }) });
  }
  function LayoutGridColumns(props) {
    const { layoutGridSchema, ...layoutGridFieldProps } = props;
    const { registry, uiSchema } = layoutGridFieldProps;
    const { children, gridProps } = findChildrenAndProps(layoutGridSchema, "ui:columns" /* COLUMNS */, registry);
    const uiOptions = utils.getUiOptions(uiSchema);
    const GridTemplate2 = utils.getTemplate("GridTemplate", registry, uiOptions);
    return children.map((child) => /* @__PURE__ */ jsxRuntime.jsx(
      GridTemplate2,
      {
        column: true,
        "data-testid": LAYOUT_GRID_FIELD_TEST_IDS.col,
        ...gridProps,
        children: /* @__PURE__ */ jsxRuntime.jsx(LayoutGridFieldChildren, { ...layoutGridFieldProps, childrenLayoutGridSchemaId: [child] })
      },
      `column-${utils.hashObject(child)}`
    ));
  }
  function LayoutGridRow(props) {
    const { layoutGridSchema, ...layoutGridFieldProps } = props;
    const { registry, uiSchema } = layoutGridFieldProps;
    const { children, gridProps } = findChildrenAndProps(layoutGridSchema, "ui:row" /* ROW */, registry);
    const uiOptions = utils.getUiOptions(uiSchema);
    const GridTemplate2 = utils.getTemplate("GridTemplate", registry, uiOptions);
    return /* @__PURE__ */ jsxRuntime.jsx(GridTemplate2, { ...gridProps, "data-testid": LAYOUT_GRID_FIELD_TEST_IDS.row, children: /* @__PURE__ */ jsxRuntime.jsx(LayoutGridFieldChildren, { ...layoutGridFieldProps, childrenLayoutGridSchemaId: children }) });
  }
  function LayoutGridFieldComponent(props) {
    const {
      gridSchema,
      schema: initialSchema,
      uiSchema,
      errorSchema,
      fieldPathId,
      onBlur,
      onFocus,
      formData,
      readonly,
      registry,
      layoutGridSchema,
      // Used to pull this out of otherProps since we don't want to pass it through
      ...otherProps
    } = props;
    const { onChange } = otherProps;
    const { fields: fields2 } = registry;
    const { SchemaField: SchemaField2, LayoutMultiSchemaField: LayoutMultiSchemaField2 } = fields2;
    const uiComponentProps = computeUIComponentPropsFromGridSchema(registry, gridSchema);
    const { name, UIComponent, uiProps } = uiComponentProps;
    const {
      schema,
      isRequired: isRequired2,
      isReadonly,
      optionsInfo,
      fieldPathId: fieldIdSchema
    } = getSchemaDetailsForField(registry, name, initialSchema, formData, fieldPathId);
    const memoFieldPathId = utils.useDeepCompareMemo(fieldIdSchema);
    if (uiComponentProps.rendered) {
      return uiComponentProps.rendered;
    }
    if (schema) {
      const Field = optionsInfo?.hasDiscriminator ? LayoutMultiSchemaField2 : SchemaField2;
      const { fieldUiSchema, uiReadonly } = computeFieldUiSchema(name, uiProps, uiSchema, isReadonly, readonly);
      return /* @__PURE__ */ jsxRuntime.jsx(
        Field,
        {
          "data-testid": optionsInfo?.hasDiscriminator ? LAYOUT_GRID_FIELD_TEST_IDS.layoutMultiSchemaField : LAYOUT_GRID_FIELD_TEST_IDS.field,
          ...otherProps,
          name,
          required: isRequired2,
          readonly: uiReadonly,
          schema,
          uiSchema: fieldUiSchema,
          errorSchema: get(errorSchema, name),
          fieldPathId: memoFieldPathId,
          formData: get(formData, name),
          onChange,
          onBlur,
          onFocus,
          options: optionsInfo?.options,
          registry
        }
      );
    }
    if (UIComponent) {
      return /* @__PURE__ */ jsxRuntime.jsx(
        UIComponent,
        {
          "data-testid": LAYOUT_GRID_FIELD_TEST_IDS.uiComponent,
          ...otherProps,
          name,
          required: isRequired2,
          formData,
          readOnly: !!isReadonly || readonly,
          errorSchema,
          uiSchema,
          schema: initialSchema,
          fieldPathId,
          onBlur,
          onFocus,
          registry,
          ...uiProps
        }
      );
    }
    return null;
  }
  function LayoutGridField(props) {
    const { uiSchema } = props;
    let { layoutGridSchema } = props;
    const uiOptions = utils.getUiOptions(uiSchema);
    if (!layoutGridSchema && LAYOUT_GRID_UI_OPTION in uiOptions && isObject(uiOptions[LAYOUT_GRID_UI_OPTION])) {
      layoutGridSchema = uiOptions[LAYOUT_GRID_UI_OPTION];
    }
    if (isObject(layoutGridSchema)) {
      if ("ui:row" /* ROW */ in layoutGridSchema) {
        return /* @__PURE__ */ jsxRuntime.jsx(LayoutGridRow, { ...props, layoutGridSchema });
      }
      if ("ui:col" /* COLUMN */ in layoutGridSchema) {
        return /* @__PURE__ */ jsxRuntime.jsx(LayoutGridCol, { ...props, layoutGridSchema });
      }
      if ("ui:columns" /* COLUMNS */ in layoutGridSchema) {
        return /* @__PURE__ */ jsxRuntime.jsx(LayoutGridColumns, { ...props, layoutGridSchema });
      }
      if ("ui:condition" /* CONDITION */ in layoutGridSchema) {
        return /* @__PURE__ */ jsxRuntime.jsx(LayoutGridCondition, { ...props, layoutGridSchema });
      }
    }
    return /* @__PURE__ */ jsxRuntime.jsx(LayoutGridFieldComponent, { ...props, gridSchema: layoutGridSchema });
  }
  LayoutGridField.TEST_IDS = LAYOUT_GRID_FIELD_TEST_IDS;
  function LayoutHeaderField(props) {
    const { fieldPathId, title, schema, uiSchema, required, registry, name } = props;
    const options = utils.getUiOptions(uiSchema, registry.globalUiOptions);
    const { title: uiTitle } = options;
    const { title: schemaTitle } = schema;
    const fieldTitle = uiTitle || title || schemaTitle || name;
    if (!fieldTitle) {
      return null;
    }
    const TitleFieldTemplate = utils.getTemplate(
      "TitleFieldTemplate",
      registry,
      options
    );
    return /* @__PURE__ */ jsxRuntime.jsx(
      TitleFieldTemplate,
      {
        id: utils.titleId(fieldPathId),
        title: fieldTitle,
        required,
        schema,
        uiSchema,
        registry
      }
    );
  }
  function getSelectedOption(options, selectorField, value) {
    const defaultValue = "!@#!@$@#$!@$#";
    const schemaOptions = options.map(({ schema }) => schema);
    return schemaOptions.find((option) => {
      const selector = get(option, [utils.PROPERTIES_KEY, selectorField]);
      const result = get(selector, utils.DEFAULT_KEY, get(selector, utils.CONST_KEY, defaultValue));
      return result === value;
    });
  }
  function computeEnumOptions(schema, options, schemaUtils, uiSchema, formData) {
    const realOptions = options.map((opt) => schemaUtils.retrieveSchema(opt, formData));
    let tempSchema = schema;
    if (has(schema, utils.ONE_OF_KEY)) {
      tempSchema = { ...schema, [utils.ONE_OF_KEY]: realOptions };
    } else if (has(schema, utils.ANY_OF_KEY)) {
      tempSchema = { ...schema, [utils.ANY_OF_KEY]: realOptions };
    }
    const enumOptions = utils.optionsList(tempSchema, uiSchema);
    if (!enumOptions) {
      throw new Error(`No enumOptions were computed from the schema ${JSON.stringify(tempSchema)}`);
    }
    return enumOptions;
  }
  function LayoutMultiSchemaField(props) {
    const {
      name,
      baseType,
      disabled = false,
      formData,
      fieldPathId,
      onBlur,
      onChange,
      options,
      onFocus,
      registry,
      uiSchema,
      schema,
      autofocus,
      readonly,
      required,
      errorSchema,
      hideError = false
    } = props;
    const { widgets: widgets2, schemaUtils, globalUiOptions } = registry;
    const [enumOptions, setEnumOptions] = react.useState(computeEnumOptions(schema, options, schemaUtils, uiSchema, formData));
    const id = get(fieldPathId, utils.ID_KEY);
    const discriminator = utils.getDiscriminatorFieldFromSchema(schema);
    const FieldErrorTemplate2 = utils.getTemplate("FieldErrorTemplate", registry, options);
    const FieldTemplate2 = utils.getTemplate("FieldTemplate", registry, options);
    const schemaHash = utils.hashObject(schema);
    const optionsHash = utils.hashObject(options);
    const uiSchemaHash = uiSchema ? utils.hashObject(uiSchema) : "";
    const formDataHash = formData ? utils.hashObject(formData) : "";
    react.useEffect(() => {
      setEnumOptions(computeEnumOptions(schema, options, schemaUtils, uiSchema, formData));
    }, [schemaHash, optionsHash, schemaUtils, uiSchemaHash, formDataHash]);
    const {
      widget = discriminator ? "radio" : "select",
      title = "",
      placeholder = "",
      optionsSchemaSelector: selectorField = discriminator,
      hideError: uiSchemaHideError,
      ...uiOptions
    } = utils.getUiOptions(uiSchema);
    if (!selectorField) {
      throw new Error("No selector field provided for the LayoutMultiSchemaField");
    }
    const selectedOption = get(formData, selectorField);
    let optionSchema = get(enumOptions[0]?.schema, [utils.PROPERTIES_KEY, selectorField], {});
    const option = getSelectedOption(enumOptions, selectorField, selectedOption);
    optionSchema = optionSchema?.type ? optionSchema : { ...optionSchema, type: option?.type || baseType };
    const Widget = utils.getWidget(optionSchema, widget, widgets2);
    const hideFieldError = uiSchemaHideError === void 0 ? hideError : Boolean(uiSchemaHideError);
    const rawErrors = get(errorSchema, [utils.ERRORS_KEY], []);
    const fieldErrorSchema = omit3(errorSchema, [utils.ERRORS_KEY]);
    const displayLabel = schemaUtils.getDisplayLabel(schema, uiSchema, globalUiOptions);
    const onOptionChange = (opt) => {
      if (disabled || readonly) {
        return;
      }
      const newOption = getSelectedOption(enumOptions, selectorField, opt);
      const oldOption = getSelectedOption(enumOptions, selectorField, selectedOption);
      let newFormData = schemaUtils.sanitizeDataForNewSchema(newOption, oldOption, formData);
      if (newFormData && newOption) {
        newFormData = schemaUtils.getDefaultFormState(newOption, newFormData, "excludeObjectChildren");
      }
      if (newFormData) {
        set(newFormData, selectorField, opt);
      }
      onChange(newFormData, fieldPathId.path, void 0, id);
    };
    const widgetOptions = { enumOptions, ...uiOptions };
    const errors = !hideFieldError && rawErrors.length > 0 ? /* @__PURE__ */ jsxRuntime.jsx(FieldErrorTemplate2, { fieldPathId, schema, errors: rawErrors, registry }) : void 0;
    return /* @__PURE__ */ jsxRuntime.jsx(
      FieldTemplate2,
      {
        fieldPathId,
        id,
        schema,
        label: (title || schema.title) ?? "",
        disabled: disabled || Array.isArray(enumOptions) && isEmpty(enumOptions),
        uiSchema,
        required,
        readonly: !!readonly,
        registry,
        displayLabel,
        errors,
        onChange,
        onKeyRename: noop,
        onKeyRenameBlur: noop,
        onRemoveProperty: noop,
        children: /* @__PURE__ */ jsxRuntime.jsx(
          Widget,
          {
            id,
            name,
            schema,
            label: (title || schema.title) ?? "",
            disabled: disabled || Array.isArray(enumOptions) && isEmpty(enumOptions),
            uiSchema,
            autofocus,
            readonly,
            required,
            registry,
            multiple: false,
            rawErrors,
            hideError: hideFieldError,
            hideLabel: !displayLabel,
            errorSchema: fieldErrorSchema,
            placeholder,
            onChange: onOptionChange,
            onBlur,
            onFocus,
            value: selectedOption,
            options: widgetOptions,
            htmlName: fieldPathId.name
          }
        )
      }
    );
  }
  var AnyOfField = class extends react.Component {
    /** Flag to skip the formData-change-driven option recalculation when the user just selected an option.
     * Set to true in the setState callback of onOptionChange (after onChange is called), consumed and reset in
     * componentDidUpdate. This prevents the matching-option recalculation from overriding a user's explicit choice
     * when getDefaultFormState populates undefined properties that make deepEquals see a false formData change.
     */
    skipNextOptionRecalculation = false;
    /** Constructs an `AnyOfField` with the given `props` to initialize the initially selected option in state
     *
     * @param props - The `FieldProps` for this template
     */
    constructor(props) {
      super(props);
      const {
        formData,
        options,
        registry: { schemaUtils }
      } = this.props;
      const retrievedOptions = options.map((opt) => schemaUtils.retrieveSchema(opt, formData));
      this.state = {
        retrievedOptions,
        selectedOption: this.getMatchingOption(0, formData, retrievedOptions)
      };
    }
    /** React lifecycle method that is called when the props and/or state for this component is updated. It recomputes the
     * currently selected option based on the overall `formData`
     *
     * @param prevProps - The previous `FieldProps` for this template
     * @param prevState - The previous `AnyOfFieldState` for this template
     */
    componentDidUpdate(prevProps, prevState) {
      const { formData, options, fieldPathId } = this.props;
      const { selectedOption } = this.state;
      let newState = this.state;
      if (!utils.deepEquals(prevProps.options, options)) {
        const {
          registry: { schemaUtils }
        } = this.props;
        const retrievedOptions = options.map((opt) => schemaUtils.retrieveSchema(opt, formData));
        newState = { selectedOption, retrievedOptions };
      }
      if (!utils.deepEquals(formData, prevProps.formData) && fieldPathId.$id === prevProps.fieldPathId.$id) {
        if (this.skipNextOptionRecalculation) {
          this.skipNextOptionRecalculation = false;
        } else {
          const { retrievedOptions } = newState;
          const matchingOption = this.getMatchingOption(selectedOption, formData, retrievedOptions);
          if (prevState && matchingOption !== selectedOption) {
            newState = { selectedOption: matchingOption, retrievedOptions };
          }
        }
      }
      if (newState !== this.state) {
        this.setState(newState);
      }
    }
    /** Determines the best matching option for the given `formData` and `options`.
     *
     * @param formData - The new formData
     * @param options - The list of options to choose from
     * @return - The index of the `option` that best matches the `formData`
     */
    getMatchingOption(selectedOption, formData, options) {
      const {
        schema,
        registry: { schemaUtils }
      } = this.props;
      const discriminator = utils.getDiscriminatorFieldFromSchema(schema);
      const option = schemaUtils.getClosestMatchingOption(formData, options, selectedOption, discriminator);
      return option;
    }
    /** Callback handler to remember what the currently selected option is. In addition to that the `formData` is updated
     * to remove properties that are not part of the newly selected option schema, and then the updated data is passed to
     * the `onChange` handler.
     *
     * @param option - The new option value being selected
     */
    onOptionChange = (option) => {
      const { selectedOption, retrievedOptions } = this.state;
      const { disabled = false, formData, onChange, readonly = false, registry, fieldPathId } = this.props;
      if (disabled || readonly) {
        return;
      }
      const { schemaUtils } = registry;
      const intOption = option !== void 0 ? parseInt(option, 10) : -1;
      if (intOption === selectedOption) {
        return;
      }
      const newOption = intOption >= 0 ? retrievedOptions[intOption] : void 0;
      const oldOption = selectedOption >= 0 ? retrievedOptions[selectedOption] : void 0;
      let newFormData = schemaUtils.sanitizeDataForNewSchema(newOption, oldOption, formData);
      if (newOption) {
        newFormData = schemaUtils.getDefaultFormState(newOption, newFormData, "excludeObjectChildren");
      }
      this.setState({ selectedOption: intOption }, () => {
        this.skipNextOptionRecalculation = true;
        onChange(newFormData, fieldPathId.path, void 0, this.getFieldId());
      });
    };
    getFieldId() {
      const { fieldPathId, schema } = this.props;
      return `${fieldPathId.$id}${schema.oneOf ? "__oneof_select" : "__anyof_select"}`;
    }
    /** Renders the `AnyOfField` selector along with a `SchemaField` for the value of the `formData`
     */
    render() {
      const {
        name,
        disabled = false,
        errorSchema = {},
        formData,
        onBlur,
        onFocus,
        readonly,
        required = false,
        registry,
        schema,
        uiSchema
      } = this.props;
      const { widgets: widgets2, fields: fields2, translateString, globalUiOptions, schemaUtils } = registry;
      const { SchemaField: SchemaFieldComponent } = fields2;
      const MultiSchemaFieldTemplate2 = utils.getTemplate(
        "MultiSchemaFieldTemplate",
        registry,
        globalUiOptions
      );
      const isOptionalRender = utils.shouldRenderOptionalField(registry, schema, required, uiSchema);
      const hasFormData = utils.isFormDataAvailable(formData);
      const { selectedOption, retrievedOptions } = this.state;
      const {
        widget = "select",
        placeholder,
        autofocus,
        autocomplete,
        title = schema.title,
        ...uiOptions
      } = utils.getUiOptions(uiSchema, globalUiOptions);
      const Widget = utils.getWidget({ type: "number" }, widget, widgets2);
      const rawErrors = get(errorSchema, utils.ERRORS_KEY, []);
      const fieldErrorSchema = omit3(errorSchema, [utils.ERRORS_KEY]);
      const displayLabel = schemaUtils.getDisplayLabel(schema, uiSchema, globalUiOptions);
      const option = selectedOption >= 0 ? retrievedOptions[selectedOption] || null : null;
      let optionSchema;
      if (option) {
        const { required: schemaRequired } = schema;
        optionSchema = schemaRequired ? utils.mergeSchemas({ required: schemaRequired }, option) : option;
      }
      let optionsUiSchema = [];
      if (utils.ONE_OF_KEY in schema && uiSchema && utils.ONE_OF_KEY in uiSchema) {
        if (Array.isArray(uiSchema[utils.ONE_OF_KEY])) {
          optionsUiSchema = uiSchema[utils.ONE_OF_KEY];
        } else {
          console.warn(`uiSchema.oneOf is not an array for "${title || name}"`);
        }
      } else if (utils.ANY_OF_KEY in schema && uiSchema && utils.ANY_OF_KEY in uiSchema) {
        if (Array.isArray(uiSchema[utils.ANY_OF_KEY])) {
          optionsUiSchema = uiSchema[utils.ANY_OF_KEY];
        } else {
          console.warn(`uiSchema.anyOf is not an array for "${title || name}"`);
        }
      }
      let optionUiSchema = uiSchema;
      if (selectedOption >= 0 && optionsUiSchema.length > selectedOption) {
        optionUiSchema = optionsUiSchema[selectedOption];
      }
      const translateEnum = title ? utils.TranslatableString.TitleOptionPrefix : utils.TranslatableString.OptionPrefix;
      const translateParams = title ? [title] : [];
      const enumOptions = retrievedOptions.map((opt, index) => {
        const { title: uiTitle = opt.title } = utils.getUiOptions(optionsUiSchema[index]);
        return {
          label: uiTitle || translateString(translateEnum, translateParams.concat(String(index + 1))),
          value: index
        };
      });
      const selector = !isOptionalRender || hasFormData ? /* @__PURE__ */ jsxRuntime.jsx(
        Widget,
        {
          id: this.getFieldId(),
          name: `${name}${schema.oneOf ? "__oneof_select" : "__anyof_select"}`,
          schema: { type: "number", default: 0 },
          onChange: this.onOptionChange,
          onBlur,
          onFocus,
          disabled: disabled || isEmpty(enumOptions),
          multiple: false,
          rawErrors,
          errorSchema: fieldErrorSchema,
          value: selectedOption >= 0 ? selectedOption : void 0,
          options: { enumOptions, ...uiOptions },
          registry,
          placeholder,
          autocomplete,
          autofocus,
          label: title ?? name,
          hideLabel: !displayLabel,
          readonly
        }
      ) : void 0;
      const optionsSchemaField = optionSchema && optionSchema.type !== "null" && /* @__PURE__ */ jsxRuntime.jsx(SchemaFieldComponent, { ...this.props, schema: optionSchema, uiSchema: optionUiSchema }) || null;
      return /* @__PURE__ */ jsxRuntime.jsx(
        MultiSchemaFieldTemplate2,
        {
          schema,
          registry,
          uiSchema,
          selector,
          optionSchemaField: optionsSchemaField
        }
      );
    }
  };
  var MultiSchemaField_default = AnyOfField;
  function NullField(props) {
    const { formData, onChange, fieldPathId } = props;
    react.useEffect(() => {
      if (formData === void 0) {
        onChange(null, fieldPathId.path);
      }
    }, [fieldPathId, formData, onChange]);
    return null;
  }
  var NullField_default = NullField;
  var trailingCharMatcherWithPrefix = /\.([0-9]*0)*$/;
  var trailingCharMatcher = /[0.]0*$/;
  function NumberField(props) {
    const { registry, onChange, formData, value: initialValue } = props;
    const [lastValue, setLastValue] = react.useState(initialValue);
    const { StringField: StringField2 } = registry.fields;
    let value = formData;
    const handleChange = react.useCallback(
      (newValue, path, errorSchema, id) => {
        setLastValue(newValue);
        const normalizedValue = `${newValue}`.startsWith(".") ? `0${newValue}` : newValue;
        const processed = typeof normalizedValue === "string" && normalizedValue.match(trailingCharMatcherWithPrefix) ? utils.asNumber(normalizedValue.replace(trailingCharMatcher, "")) : utils.asNumber(normalizedValue);
        onChange(processed, path, errorSchema, id);
      },
      [onChange]
    );
    if (typeof lastValue === "string" && typeof value === "number") {
      const re = new RegExp(`^(${String(value).replace(".", "\\.")})?\\.?0*$`);
      if (lastValue.match(re)) {
        value = lastValue;
      }
    }
    return /* @__PURE__ */ jsxRuntime.jsx(StringField2, { ...props, formData: value, onChange: handleChange });
  }
  var NumberField_default = NumberField;

  // src/components/constants.ts
  var ADDITIONAL_PROPERTY_KEY_REMOVE = /* @__PURE__ */ Symbol("remove-this-key");
  var IS_RESET = /* @__PURE__ */ Symbol("reset");
  function isRequired(schema, name) {
    return Array.isArray(schema.required) && schema.required.includes(name);
  }
  function getDefaultValue(translateString, type) {
    switch (type) {
      case "array":
        return [];
      case "boolean":
        return false;
      case "null":
        return null;
      case "number":
        return 0;
      case "object":
        return {};
      case "string":
      default:
        return translateString(utils.TranslatableString.NewStringDefault);
    }
  }
  function ObjectFieldPropertyFn(props) {
    const {
      fieldPathId,
      schema,
      registry,
      uiSchema,
      errorSchema,
      formData,
      onChange,
      onBlur,
      onFocus,
      disabled,
      readonly,
      required,
      hideError,
      propertyName,
      handleKeyRename,
      handleRemoveProperty,
      addedByAdditionalProperties
    } = props;
    const [wasPropertyKeyModified, setWasPropertyKeyModified] = react.useState(false);
    const { globalFormOptions, fields: fields2 } = registry;
    const { SchemaField: SchemaField2 } = fields2;
    const innerFieldIdPathId = utils.useDeepCompareMemo(
      utils.toFieldPathId(propertyName, globalFormOptions, fieldPathId.path)
    );
    const onPropertyChange = react.useCallback(
      (value, path, newErrorSchema, id) => {
        let normalizedValue = value;
        if (value === void 0 && addedByAdditionalProperties) {
          normalizedValue = "";
        }
        onChange(normalizedValue, path, newErrorSchema, id);
      },
      [onChange, addedByAdditionalProperties]
    );
    const onKeyRename = react.useCallback(
      (value) => {
        if (propertyName !== value) {
          setWasPropertyKeyModified(true);
        }
        handleKeyRename(propertyName, value);
      },
      [propertyName, handleKeyRename]
    );
    const onKeyRenameBlur = react.useCallback(
      (event) => {
        const {
          target: { value }
        } = event;
        onKeyRename(value);
      },
      [onKeyRename]
    );
    const onRemoveProperty = react.useCallback(() => {
      handleRemoveProperty(propertyName);
    }, [propertyName, handleRemoveProperty]);
    return /* @__PURE__ */ jsxRuntime.jsx(
      SchemaField2,
      {
        name: propertyName,
        required,
        schema,
        uiSchema,
        errorSchema,
        fieldPathId: innerFieldIdPathId,
        formData,
        wasPropertyKeyModified,
        onKeyRename,
        onKeyRenameBlur,
        onRemoveProperty,
        onChange: onPropertyChange,
        onBlur,
        onFocus,
        registry,
        disabled,
        readonly,
        hideError
      }
    );
  }
  var ObjectFieldProperty = react.memo(ObjectFieldPropertyFn);
  function ObjectField(props) {
    const {
      schema: rawSchema,
      uiSchema = {},
      formData,
      errorSchema,
      fieldPathId,
      name,
      required = false,
      disabled,
      readonly,
      hideError,
      onBlur,
      onFocus,
      onChange,
      registry,
      title
    } = props;
    const { fields: fields2, schemaUtils, translateString, globalUiOptions } = registry;
    const { OptionalDataControlsField: OptionalDataControlsField2 } = fields2;
    const formDataRef = react.useRef(formData);
    formDataRef.current = formData;
    const schema = react.useMemo(
      () => schemaUtils.retrieveSchema(rawSchema, formData, true),
      [schemaUtils, rawSchema, formData]
    );
    const uiOptions = react.useMemo(() => utils.getUiOptions(uiSchema, globalUiOptions), [uiSchema, globalUiOptions]);
    const { properties: schemaProperties = {} } = schema;
    const childFieldPathId = props.childFieldPathId ?? fieldPathId;
    const lastRenamedProperty = react.useRef({ previousKey: "", currentKey: void 0 });
    const templateTitle = uiOptions.title ?? schema.title ?? title ?? name;
    const description = uiOptions.description ?? schema.description;
    const renderOptionalField = utils.shouldRenderOptionalField(registry, schema, required, uiSchema);
    const hasFormData = utils.isFormDataAvailable(formData);
    let orderedProperties = [];
    const getAvailableKey = react.useCallback(
      (preferredKey, existingFormData) => {
        const { duplicateKeySuffixSeparator = "-" } = utils.getUiOptions(uiSchema, globalUiOptions);
        let index = 0;
        let newKey = preferredKey;
        while (has(existingFormData, newKey)) {
          index += 1;
          newKey = `${preferredKey}${duplicateKeySuffixSeparator}${index}`;
        }
        return newKey;
      },
      [uiSchema, globalUiOptions]
    );
    const onAddProperty = react.useCallback(() => {
      if (!(schema.additionalProperties || schema.patternProperties)) {
        return;
      }
      const newFormData = { ...formData };
      const newKey = getAvailableKey("newKey", newFormData);
      if (schema.patternProperties) {
        set(newFormData, newKey, null);
      } else {
        let type = void 0;
        let constValue = void 0;
        let defaultValue = void 0;
        if (isObject(schema.additionalProperties)) {
          type = schema.additionalProperties.type;
          constValue = schema.additionalProperties.const;
          defaultValue = schema.additionalProperties.default;
          let apSchema = schema.additionalProperties;
          if (utils.REF_KEY in apSchema) {
            apSchema = schemaUtils.retrieveSchema({ [utils.REF_KEY]: apSchema[utils.REF_KEY] }, formData);
            type = apSchema.type;
            constValue = apSchema.const;
            defaultValue = apSchema.default;
          }
          if (!type && (utils.ANY_OF_KEY in apSchema || utils.ONE_OF_KEY in apSchema)) {
            type = "object";
          }
        }
        const newValue = constValue ?? defaultValue ?? getDefaultValue(translateString, type);
        set(newFormData, newKey, newValue);
      }
      if (lastRenamedProperty.current.previousKey === newKey) {
        lastRenamedProperty.current.currentKey = newKey;
        lastRenamedProperty.current.previousKey = getAvailableKey(newKey, newFormData);
      }
      onChange(newFormData, childFieldPathId.path);
    }, [formData, onChange, translateString, schemaUtils, childFieldPathId, getAvailableKey, schema]);
    const handleKeyRename = react.useCallback(
      (oldKey, newKey) => {
        if (oldKey !== newKey) {
          const currentFormData = formDataRef.current;
          const actualNewKey = getAvailableKey(newKey, currentFormData);
          const newFormData = {
            ...currentFormData
          };
          const newKeys = { [oldKey]: actualNewKey };
          const keyValues = Object.keys(newFormData).map((key) => {
            const mappedKey = Object.hasOwn(newKeys, key) ? newKeys[key] : key;
            return { [mappedKey]: newFormData[key] };
          });
          const renamedObj = Object.assign({}, ...keyValues);
          formDataRef.current = renamedObj;
          if (oldKey !== lastRenamedProperty.current.currentKey) {
            lastRenamedProperty.current.previousKey = oldKey;
          }
          lastRenamedProperty.current.currentKey = actualNewKey;
          onChange(renamedObj, childFieldPathId.path);
        }
      },
      [onChange, childFieldPathId, getAvailableKey]
    );
    const handleRemoveProperty = react.useCallback(
      (key) => {
        onChange(ADDITIONAL_PROPERTY_KEY_REMOVE, [...childFieldPathId.path, key]);
      },
      [onChange, childFieldPathId]
    );
    const getStableKey = react.useCallback((property) => {
      if (lastRenamedProperty.current.currentKey === property) {
        return lastRenamedProperty.current.previousKey;
      }
      return property;
    }, []);
    if (!renderOptionalField || hasFormData) {
      try {
        const properties = Object.keys(schemaProperties);
        orderedProperties = utils.orderProperties(properties, uiOptions.order);
      } catch (err) {
        return /* @__PURE__ */ jsxRuntime.jsxs("div", { children: [
          /* @__PURE__ */ jsxRuntime.jsx("p", { className: "rjsf-config-error", style: { color: "red" }, children: /* @__PURE__ */ jsxRuntime.jsx(react$1.Markdown, { options: { disableParsingRawHTML: true }, children: translateString(utils.TranslatableString.InvalidObjectField, [name || "root", err.message]) }) }),
          /* @__PURE__ */ jsxRuntime.jsx("pre", { children: JSON.stringify(schema) })
        ] });
      }
    }
    const Template = utils.getTemplate("ObjectFieldTemplate", registry, uiOptions);
    const optionalDataControl = renderOptionalField ? /* @__PURE__ */ jsxRuntime.jsx(OptionalDataControlsField2, { ...props, fieldPathId: childFieldPathId, schema }) : void 0;
    const templateProps = {
      // getDisplayLabel() always returns false for object types, so just check the `uiOptions.label`
      title: uiOptions.label === false ? "" : templateTitle,
      description: uiOptions.label === false ? void 0 : description,
      properties: orderedProperties.map((propertyName) => {
        const addedByAdditionalProperties = has(schema, [utils.PROPERTIES_KEY, propertyName, utils.ADDITIONAL_PROPERTY_FLAG]);
        const fieldUiSchema = addedByAdditionalProperties ? uiSchema.additionalProperties : uiSchema[propertyName];
        const hidden = utils.getUiOptions(fieldUiSchema).widget === "hidden";
        const content = /* @__PURE__ */ jsxRuntime.jsx(
          ObjectFieldProperty,
          {
            propertyName,
            required: isRequired(schema, propertyName),
            schema: get(schema, [utils.PROPERTIES_KEY, propertyName], {}),
            uiSchema: fieldUiSchema,
            errorSchema: get(errorSchema, [propertyName]),
            fieldPathId: childFieldPathId,
            formData: get(formData, [propertyName]),
            handleKeyRename,
            handleRemoveProperty,
            addedByAdditionalProperties,
            onChange,
            onBlur,
            onFocus,
            registry,
            disabled,
            readonly,
            hideError
          },
          getStableKey(propertyName)
        );
        return {
          content,
          name: propertyName,
          readonly,
          disabled,
          required,
          hidden
        };
      }),
      readonly,
      disabled,
      required,
      fieldPathId,
      uiSchema,
      errorSchema,
      schema,
      formData,
      registry,
      optionalDataControl,
      className: renderOptionalField ? "rjsf-optional-object-field" : void 0
    };
    return /* @__PURE__ */ jsxRuntime.jsx(Template, { ...templateProps, onAddProperty });
  }
  function OptionalDataControlsField(props) {
    const {
      schema,
      uiSchema = {},
      formData,
      disabled = false,
      readonly = false,
      onChange,
      errorSchema,
      fieldPathId,
      registry
    } = props;
    const { globalUiOptions = {}, schemaUtils, translateString } = registry;
    const uiOptions = utils.getUiOptions(uiSchema, globalUiOptions);
    const OptionalDataControlsTemplate2 = utils.getTemplate(
      "OptionalDataControlsTemplate",
      registry,
      uiOptions
    );
    const hasFormData = utils.isFormDataAvailable(formData);
    let id;
    let label;
    let onAddClick;
    let onRemoveClick;
    if (disabled || readonly) {
      id = utils.optionalControlsId(fieldPathId, "Msg");
      label = hasFormData ? void 0 : translateString(utils.TranslatableString.OptionalObjectEmptyMsg);
    } else {
      const labelEnum = hasFormData ? utils.TranslatableString.OptionalObjectRemove : utils.TranslatableString.OptionalObjectAdd;
      label = translateString(labelEnum);
      if (hasFormData) {
        id = utils.optionalControlsId(fieldPathId, "Remove");
        onRemoveClick = () => onChange(void 0, fieldPathId.path, errorSchema);
      } else {
        id = utils.optionalControlsId(fieldPathId, "Add");
        onAddClick = () => {
          let newFormData = schemaUtils.getDefaultFormState(schema, formData, "excludeObjectChildren");
          if (newFormData === void 0) {
            newFormData = utils.getSchemaType(schema) === "array" ? [] : {};
          }
          onChange(newFormData, fieldPathId.path, errorSchema);
        };
      }
    }
    return label && /* @__PURE__ */ jsxRuntime.jsx(
      OptionalDataControlsTemplate2,
      {
        id,
        registry,
        schema,
        uiSchema,
        label,
        onAddClick,
        onRemoveClick
      }
    );
  }
  var COMPONENT_TYPES = {
    array: "ArrayField",
    boolean: "BooleanField",
    integer: "NumberField",
    number: "NumberField",
    object: "ObjectField",
    string: "StringField",
    null: "NullField"
  };
  function getFieldComponent(schema, uiOptions, registry) {
    const { field } = uiOptions;
    const { fields: fields2 } = registry;
    if (typeof field === "function") {
      return field;
    }
    if (typeof field === "string" && field in fields2) {
      return fields2[field];
    }
    const schemaType = utils.getSchemaType(schema);
    const type = Array.isArray(schemaType) ? schemaType[0] : schemaType || "";
    const schemaId = schema.$id;
    let componentName = COMPONENT_TYPES[type];
    if (schemaId && schemaId in fields2) {
      componentName = schemaId;
    }
    if (!componentName && (schema.anyOf || schema.oneOf)) {
      return () => null;
    }
    return componentName in fields2 ? fields2[componentName] : fields2.FallbackField;
  }
  function SchemaFieldRender(props) {
    const {
      schema: _schema,
      fieldPathId,
      uiSchema: _uiSchema,
      formData,
      errorSchema,
      name,
      onChange,
      onKeyRename,
      onKeyRenameBlur,
      onRemoveProperty,
      required = false,
      registry,
      wasPropertyKeyModified = false
    } = props;
    const { schemaUtils, globalFormOptions, globalUiOptions, fields: fields2 } = registry;
    const { AnyOfField: _AnyOfField, OneOfField: _OneOfField } = fields2;
    const uiSchema = utils.resolveUiSchema(_schema, _uiSchema, registry);
    const uiOptions = utils.getUiOptions(uiSchema, globalUiOptions);
    const FieldTemplate2 = utils.getTemplate("FieldTemplate", registry, uiOptions);
    const DescriptionFieldTemplate = utils.getTemplate(
      "DescriptionFieldTemplate",
      registry,
      uiOptions
    );
    const FieldHelpTemplate2 = utils.getTemplate("FieldHelpTemplate", registry, uiOptions);
    const FieldErrorTemplate2 = utils.getTemplate("FieldErrorTemplate", registry, uiOptions);
    const schema = schemaUtils.retrieveSchema(_schema, formData);
    const fieldId = fieldPathId[utils.ID_KEY];
    const handleFieldComponentChange = react.useCallback(
      (newFormData, path, newErrorSchema, id2) => {
        const theId = id2 || fieldId;
        return onChange(newFormData, path, newErrorSchema, theId);
      },
      [fieldId, onChange]
    );
    const FieldComponent = getFieldComponent(schema, uiOptions, registry);
    const isDeprecated = Boolean(schema.deprecated);
    const deprecatedHandling = isDeprecated ? uiOptions.deprecatedHandling ?? "label" : void 0;
    const disabled = Boolean(uiOptions.disabled ?? props.disabled) || deprecatedHandling === "disable";
    const readonly = Boolean(uiOptions.readonly ?? (props.readonly || props.schema.readOnly || schema.readOnly));
    const uiSchemaHideError = uiOptions.hideError;
    const hideError = uiSchemaHideError === void 0 ? props.hideError : Boolean(uiSchemaHideError);
    const autofocus = Boolean(uiOptions.autofocus ?? props.autofocus);
    if (Object.keys(schema).length === 0) {
      return null;
    }
    let displayLabel = schemaUtils.getDisplayLabel(schema, uiSchema, globalUiOptions);
    const isReplacingAnyOrOneOf = uiOptions.field && uiOptions.fieldReplacesAnyOrOneOf === true;
    let XxxOfField;
    let XxxOfOptions;
    let fieldPathIdProps = { fieldPathId };
    if ((utils.ANY_OF_KEY in schema || utils.ONE_OF_KEY in schema) && !isReplacingAnyOrOneOf && !schemaUtils.isSelect(schema)) {
      if (schema[utils.ANY_OF_KEY]) {
        XxxOfField = _AnyOfField;
        XxxOfOptions = schema[utils.ANY_OF_KEY].map(
          (xxxOfSchema) => schemaUtils.retrieveSchema(isObject(xxxOfSchema) ? xxxOfSchema : {}, formData)
        );
      } else if (schema[utils.ONE_OF_KEY]) {
        XxxOfField = _OneOfField;
        XxxOfOptions = schema[utils.ONE_OF_KEY].map(
          (xxxOfSchema) => schemaUtils.retrieveSchema(isObject(xxxOfSchema) ? xxxOfSchema : {}, formData)
        );
      }
      const isOptionalRender = utils.shouldRenderOptionalField(registry, schema, required, uiSchema);
      const hasFormData = utils.isFormDataAvailable(formData);
      displayLabel = displayLabel && (!isOptionalRender || hasFormData);
      fieldPathIdProps = {
        childFieldPathId: fieldPathId,
        // The main FieldComponent will add `XxxOf` onto the fieldPathId to avoid duplication with the rendering of the
        // same FieldComponent by the `XxxOfField`
        fieldPathId: utils.toFieldPathId("XxxOf", globalFormOptions, fieldPathId)
      };
    }
    const { __errors, ...fieldErrorSchema } = errorSchema || {};
    const fieldUiSchema = omit3(uiSchema, ["ui:classNames", "classNames", "ui:style"]);
    if (utils.UI_OPTIONS_KEY in fieldUiSchema) {
      fieldUiSchema[utils.UI_OPTIONS_KEY] = omit3(fieldUiSchema[utils.UI_OPTIONS_KEY], ["classNames", "style"]);
    }
    const field = /* @__PURE__ */ jsxRuntime.jsx(
      FieldComponent,
      {
        ...props,
        onChange: handleFieldComponentChange,
        ...fieldPathIdProps,
        schema,
        uiSchema: fieldUiSchema,
        disabled,
        readonly,
        hideError,
        autofocus,
        errorSchema: fieldErrorSchema,
        rawErrors: __errors
      }
    );
    const id = fieldPathId[utils.ID_KEY];
    let label;
    if (wasPropertyKeyModified) {
      label = name;
    } else {
      label = utils.ADDITIONAL_PROPERTY_FLAG in schema ? name : uiOptions.title || props.schema.title || schema.title || props.title || name;
    }
    if (deprecatedHandling === "label") {
      label = registry.translateString(utils.TranslatableString.DeprecatedLabel, [label]);
    }
    const description = uiOptions.description || props.schema.description || schema.description || "";
    const { help } = uiOptions;
    const hidden = uiOptions.widget === "hidden" || deprecatedHandling === "hide";
    const classNames = ["rjsf-field", `rjsf-field-${utils.getSchemaType(schema)}`];
    if (!hideError && __errors && __errors.length > 0) {
      classNames.push("rjsf-field-error");
    }
    if (uiOptions.classNames) {
      classNames.push(uiOptions.classNames);
    }
    const helpComponent = /* @__PURE__ */ jsxRuntime.jsx(
      FieldHelpTemplate2,
      {
        help,
        fieldPathId,
        schema,
        uiSchema,
        hasErrors: !hideError && __errors && __errors.length > 0,
        registry
      }
    );
    const errorsComponent = hideError || XxxOfField && !schemaUtils.isSelect(schema) ? void 0 : /* @__PURE__ */ jsxRuntime.jsx(
      FieldErrorTemplate2,
      {
        errors: __errors,
        errorSchema,
        fieldPathId,
        schema,
        uiSchema,
        registry
      }
    );
    const fieldProps = {
      description: /* @__PURE__ */ jsxRuntime.jsx(
        DescriptionFieldTemplate,
        {
          id: utils.descriptionId(id),
          description,
          schema,
          uiSchema,
          registry
        }
      ),
      rawDescription: description,
      help: helpComponent,
      rawHelp: typeof help === "string" ? help : void 0,
      errors: errorsComponent,
      rawErrors: hideError ? void 0 : __errors,
      fieldPathId,
      id,
      label,
      hidden,
      onChange,
      onKeyRename,
      onKeyRenameBlur,
      onRemoveProperty,
      required,
      disabled,
      readonly,
      hideError,
      displayLabel,
      classNames: classNames.join(" ").trim(),
      style: uiOptions.style,
      formData,
      schema,
      uiSchema,
      registry
    };
    return /* @__PURE__ */ jsxRuntime.jsx(FieldTemplate2, { ...fieldProps, children: /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
      field,
      XxxOfField && /* @__PURE__ */ jsxRuntime.jsx(
        XxxOfField,
        {
          name,
          disabled,
          readonly,
          hideError,
          errorSchema,
          formData,
          fieldPathId,
          onBlur: props.onBlur,
          onChange: props.onChange,
          onFocus: props.onFocus,
          options: XxxOfOptions,
          registry,
          required,
          schema,
          uiSchema
        }
      )
    ] }) });
  }
  var SchemaField = class extends react.Component {
    shouldComponentUpdate(nextProps) {
      const {
        registry: { globalFormOptions }
      } = this.props;
      const { experimental_componentUpdateStrategy = "customDeep" } = globalFormOptions;
      return utils.shouldRender(this, nextProps, this.state, experimental_componentUpdateStrategy);
    }
    render() {
      return /* @__PURE__ */ jsxRuntime.jsx(SchemaFieldRender, { ...this.props });
    }
  };
  var SchemaField_default = SchemaField;
  function StringField(props) {
    const {
      schema,
      name,
      uiSchema,
      fieldPathId,
      formData,
      required,
      disabled = false,
      readonly = false,
      autofocus = false,
      onChange,
      onBlur,
      onFocus,
      registry,
      rawErrors,
      hideError,
      title
    } = props;
    const { title: schemaTitle, format } = schema;
    const { widgets: widgets2, schemaUtils, globalUiOptions } = registry;
    const enumOptions = schemaUtils.isSelect(schema) ? utils.optionsList(schema, uiSchema) : void 0;
    let defaultWidget = enumOptions ? "select" : "text";
    if (format && utils.hasWidget(schema, format, widgets2)) {
      defaultWidget = format;
    }
    const { widget = defaultWidget, placeholder = "", title: uiTitle, ...options } = utils.getUiOptions(uiSchema);
    const displayLabel = schemaUtils.getDisplayLabel(schema, uiSchema, globalUiOptions);
    const label = uiTitle ?? title ?? schemaTitle ?? name;
    const Widget = utils.getWidget(schema, widget, widgets2);
    const onWidgetChange = react.useCallback(
      (value, errorSchema, id) => onChange(value, fieldPathId.path, errorSchema, id),
      [onChange, fieldPathId]
    );
    return /* @__PURE__ */ jsxRuntime.jsx(
      Widget,
      {
        options: { ...options, enumOptions },
        schema,
        uiSchema,
        id: fieldPathId.$id,
        name,
        label,
        hideLabel: !displayLabel,
        hideError,
        value: formData,
        onChange: onWidgetChange,
        onBlur,
        onFocus,
        required,
        disabled,
        readonly,
        autofocus,
        registry,
        placeholder,
        rawErrors,
        htmlName: fieldPathId.name
      }
    );
  }
  var StringField_default = StringField;

  // src/components/fields/index.ts
  function fields() {
    return {
      AnyOfField: MultiSchemaField_default,
      ArrayField,
      // ArrayField falls back to SchemaField if ArraySchemaField is not defined, which it isn't by default
      BooleanField: BooleanField_default,
      FallbackField,
      LayoutGridField,
      LayoutHeaderField,
      LayoutMultiSchemaField,
      NumberField: NumberField_default,
      ObjectField,
      OneOfField: MultiSchemaField_default,
      OptionalDataControlsField,
      SchemaField: SchemaField_default,
      StringField: StringField_default,
      NullField: NullField_default
    };
  }
  var fields_default = fields;
  function ArrayFieldDescriptionTemplate(props) {
    const { fieldPathId, description, registry, schema, uiSchema } = props;
    const options = utils.getUiOptions(uiSchema, registry.globalUiOptions);
    const { label: displayLabel = true } = options;
    if (!description || !displayLabel) {
      return null;
    }
    const DescriptionFieldTemplate = utils.getTemplate(
      "DescriptionFieldTemplate",
      registry,
      options
    );
    return /* @__PURE__ */ jsxRuntime.jsx(
      DescriptionFieldTemplate,
      {
        id: utils.descriptionId(fieldPathId),
        description,
        schema,
        uiSchema,
        registry
      }
    );
  }
  function ArrayFieldItemButtonsTemplate(props) {
    const {
      disabled,
      hasCopy,
      hasMoveDown,
      hasMoveUp,
      hasRemove,
      fieldPathId,
      onCopyItem,
      onRemoveItem,
      onMoveDownItem,
      onMoveUpItem,
      readonly,
      registry,
      uiSchema
    } = props;
    const { CopyButton: CopyButton2, MoveDownButton: MoveDownButton2, MoveUpButton: MoveUpButton2, RemoveButton: RemoveButton2 } = registry.templates.ButtonTemplates;
    return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
      (hasMoveUp || hasMoveDown) && /* @__PURE__ */ jsxRuntime.jsx(
        MoveUpButton2,
        {
          id: utils.buttonId(fieldPathId, "moveUp"),
          className: "rjsf-array-item-move-up",
          disabled: disabled || readonly || !hasMoveUp,
          onClick: onMoveUpItem,
          uiSchema,
          registry
        }
      ),
      (hasMoveUp || hasMoveDown) && /* @__PURE__ */ jsxRuntime.jsx(
        MoveDownButton2,
        {
          id: utils.buttonId(fieldPathId, "moveDown"),
          className: "rjsf-array-item-move-down",
          disabled: disabled || readonly || !hasMoveDown,
          onClick: onMoveDownItem,
          uiSchema,
          registry
        }
      ),
      hasCopy && /* @__PURE__ */ jsxRuntime.jsx(
        CopyButton2,
        {
          id: utils.buttonId(fieldPathId, "copy"),
          className: "rjsf-array-item-copy",
          disabled: disabled || readonly,
          onClick: onCopyItem,
          uiSchema,
          registry
        }
      ),
      hasRemove && /* @__PURE__ */ jsxRuntime.jsx(
        RemoveButton2,
        {
          id: utils.buttonId(fieldPathId, "remove"),
          className: "rjsf-array-item-remove",
          disabled: disabled || readonly,
          onClick: onRemoveItem,
          uiSchema,
          registry
        }
      )
    ] });
  }
  function ArrayFieldItemTemplate(props) {
    const { children, className, buttonsProps, displayLabel, hasDescription, hasToolbar, registry, uiSchema } = props;
    const uiOptions = utils.getUiOptions(uiSchema);
    const ArrayFieldItemButtonsTemplate2 = utils.getTemplate(
      "ArrayFieldItemButtonsTemplate",
      registry,
      uiOptions
    );
    const btnStyle = {
      flex: 1,
      paddingLeft: 6,
      paddingRight: 6,
      fontWeight: "bold"
    };
    const margin = hasDescription ? 31 : 9;
    const containerStyle = { display: "flex", alignItems: displayLabel ? "center" : "baseline" };
    const toolbarStyle = { display: "flex", justifyContent: "flex-end", marginTop: displayLabel ? `${margin}px` : 0 };
    return /* @__PURE__ */ jsxRuntime.jsxs("div", { className, style: containerStyle, children: [
      /* @__PURE__ */ jsxRuntime.jsx("div", { className: hasToolbar ? "col-xs-9 col-md-10 col-xl-11" : "col-xs-12", children }),
      hasToolbar && /* @__PURE__ */ jsxRuntime.jsx("div", { className: "col-xs-3 col-md-2 col-xl-1 array-item-toolbox", children: /* @__PURE__ */ jsxRuntime.jsx("div", { className: "btn-group", style: toolbarStyle, children: /* @__PURE__ */ jsxRuntime.jsx(ArrayFieldItemButtonsTemplate2, { ...buttonsProps, style: btnStyle }) }) })
    ] });
  }
  function ArrayFieldTemplate(props) {
    const {
      canAdd,
      className,
      disabled,
      fieldPathId,
      uiSchema,
      items,
      optionalDataControl,
      onAddClick,
      readonly,
      registry,
      required,
      schema,
      title
    } = props;
    const uiOptions = utils.getUiOptions(uiSchema);
    const ArrayFieldDescriptionTemplate2 = utils.getTemplate(
      "ArrayFieldDescriptionTemplate",
      registry,
      uiOptions
    );
    const ArrayFieldTitleTemplate2 = utils.getTemplate(
      "ArrayFieldTitleTemplate",
      registry,
      uiOptions
    );
    const showOptionalDataControlInTitle = !readonly && !disabled;
    const {
      ButtonTemplates: { AddButton: AddButton2 }
    } = registry.templates;
    return /* @__PURE__ */ jsxRuntime.jsxs("fieldset", { className, id: fieldPathId.$id, children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        ArrayFieldTitleTemplate2,
        {
          fieldPathId,
          title: uiOptions.title || title,
          required,
          schema,
          uiSchema,
          registry,
          optionalDataControl: showOptionalDataControlInTitle ? optionalDataControl : void 0
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        ArrayFieldDescriptionTemplate2,
        {
          fieldPathId,
          description: uiOptions.description || schema.description,
          schema,
          uiSchema,
          registry
        }
      ),
      !showOptionalDataControlInTitle ? optionalDataControl : void 0,
      /* @__PURE__ */ jsxRuntime.jsx("div", { className: "row array-item-list", children: items }),
      canAdd && /* @__PURE__ */ jsxRuntime.jsx(
        AddButton2,
        {
          id: utils.buttonId(fieldPathId, "add"),
          className: "rjsf-array-item-add",
          onClick: onAddClick,
          disabled: disabled || readonly,
          uiSchema,
          registry
        }
      )
    ] });
  }
  function ArrayFieldTitleTemplate(props) {
    const { fieldPathId, title, schema, uiSchema, required, registry, optionalDataControl } = props;
    const options = utils.getUiOptions(uiSchema, registry.globalUiOptions);
    const { label: displayLabel = true } = options;
    if (!title || !displayLabel) {
      return null;
    }
    const TitleFieldTemplate = utils.getTemplate(
      "TitleFieldTemplate",
      registry,
      options
    );
    return /* @__PURE__ */ jsxRuntime.jsx(
      TitleFieldTemplate,
      {
        id: utils.titleId(fieldPathId),
        title,
        required,
        schema,
        uiSchema,
        registry,
        optionalDataControl
      }
    );
  }
  function SchemaExamples(props) {
    const { id, schema } = props;
    const { examples, default: schemaDefault } = schema;
    if (!Array.isArray(examples)) {
      return null;
    }
    return /* @__PURE__ */ jsxRuntime.jsx("datalist", { id: utils.examplesId(id), children: examples.concat(
      schemaDefault !== void 0 && !examples.map(String).includes(String(schemaDefault)) ? [schemaDefault] : []
    ).map((example) => (
      // oxlint-disable-next-line jsx-a11y/control-has-associated-label
      /* @__PURE__ */ jsxRuntime.jsx("option", { value: example }, String(example))
    )) }, `datalist_${id}`);
  }
  function BaseInputTemplate(props) {
    const {
      id,
      name,
      // remove this from ...rest
      htmlName,
      value,
      readonly,
      disabled,
      autofocus,
      onBlur,
      onFocus,
      onChange,
      onChangeOverride,
      options,
      schema,
      uiSchema,
      registry,
      rawErrors,
      type,
      hideLabel,
      // remove this from ...rest
      hideError,
      // remove this from ...rest
      ...rest
    } = props;
    const { ClearButton: ClearButton2 } = registry.templates.ButtonTemplates;
    if (!id) {
      console.log("No id for", props);
      throw new Error(`no id for props ${JSON.stringify(props)}`);
    }
    const inputProps = {
      ...rest,
      ...utils.getInputProps(schema, type, options)
    };
    let inputValue;
    if (inputProps.type === "number" || inputProps.type === "integer") {
      inputValue = value || value === 0 ? value : "";
    } else {
      inputValue = value == null ? "" : value;
    }
    const handleChange = react.useCallback(
      ({ target: { value: newValue } }) => onChange(newValue === "" ? options.emptyValue : newValue),
      [onChange, options]
    );
    const handleBlur = react.useCallback(
      ({ target }) => onBlur(id, target && target.value),
      [onBlur, id]
    );
    const handleFocus = react.useCallback(
      ({ target }) => onFocus(id, target && target.value),
      [onFocus, id]
    );
    const handleClear = react.useCallback(
      (e) => {
        e.preventDefault();
        e.stopPropagation();
        onChange(options.emptyValue ?? "");
      },
      [onChange, options.emptyValue]
    );
    return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        "input",
        {
          id,
          name: htmlName || id,
          className: "form-control",
          readOnly: readonly,
          disabled,
          autoFocus: autofocus,
          value: inputValue,
          ...inputProps,
          list: schema.examples ? utils.examplesId(id) : void 0,
          onChange: onChangeOverride || handleChange,
          onBlur: handleBlur,
          onFocus: handleFocus,
          "aria-describedby": utils.ariaDescribedByIds(id, !!schema.examples)
        }
      ),
      options.allowClearTextInputs && !readonly && !disabled && inputValue && /* @__PURE__ */ jsxRuntime.jsx(ClearButton2, { registry, onClick: handleClear }),
      /* @__PURE__ */ jsxRuntime.jsx(SchemaExamples, { id, schema })
    ] });
  }
  function IconButtonFn(props) {
    const { iconType = "default", icon, className, uiSchema, registry, ...otherProps } = props;
    return /* @__PURE__ */ jsxRuntime.jsx("button", { type: "button", className: `btn btn-${iconType} ${className}`, ...otherProps, children: /* @__PURE__ */ jsxRuntime.jsx("i", { className: `glyphicon glyphicon-${icon}` }) });
  }
  var IconButton = react.memo(IconButtonFn);
  var IconButton_default = IconButton;
  function CopyButtonFn(props) {
    const {
      registry: { translateString }
    } = props;
    return /* @__PURE__ */ jsxRuntime.jsx(IconButton, { title: translateString(utils.TranslatableString.CopyButton), ...props, icon: "copy" });
  }
  var CopyButton = react.memo(CopyButtonFn);
  function MoveDownButtonFn(props) {
    const {
      registry: { translateString }
    } = props;
    return /* @__PURE__ */ jsxRuntime.jsx(IconButton, { title: translateString(utils.TranslatableString.MoveDownButton), ...props, icon: "arrow-down" });
  }
  var MoveDownButton = react.memo(MoveDownButtonFn);
  function MoveUpButtonFn(props) {
    const {
      registry: { translateString }
    } = props;
    return /* @__PURE__ */ jsxRuntime.jsx(IconButton, { title: translateString(utils.TranslatableString.MoveUpButton), ...props, icon: "arrow-up" });
  }
  var MoveUpButton = react.memo(MoveUpButtonFn);
  function RemoveButtonFn(props) {
    const {
      registry: { translateString }
    } = props;
    return /* @__PURE__ */ jsxRuntime.jsx(IconButton, { title: translateString(utils.TranslatableString.RemoveButton), ...props, iconType: "danger", icon: "remove" });
  }
  var RemoveButton = react.memo(RemoveButtonFn);
  function ClearButtonFn({
    id,
    className,
    onClick,
    disabled,
    registry,
    ...props
  }) {
    const { translateString } = registry;
    return /* @__PURE__ */ jsxRuntime.jsx(
      IconButton,
      {
        id,
        iconType: "default",
        icon: "remove",
        className: "btn-clear col-xs-12",
        title: translateString(utils.TranslatableString.ClearButton),
        onClick,
        disabled,
        registry,
        ...props
      }
    );
  }
  var ClearButton = react.memo(ClearButtonFn);
  function AddButton({
    id,
    className,
    onClick,
    disabled,
    registry
  }) {
    const { translateString } = registry;
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "row", children: /* @__PURE__ */ jsxRuntime.jsx(
      "p",
      {
        className: `col-xs-4 col-sm-2 col-lg-1 col-xs-offset-8 col-sm-offset-10 col-lg-offset-11 text-right ${className}`,
        children: /* @__PURE__ */ jsxRuntime.jsx(
          IconButton_default,
          {
            id,
            iconType: "info",
            icon: "plus",
            className: "btn-add col-xs-12",
            title: translateString(utils.TranslatableString.AddButton),
            onClick,
            disabled,
            registry
          }
        )
      }
    ) });
  }
  function SubmitButton({ uiSchema }) {
    const { submitText, norender, props: submitButtonProps = {} } = utils.getSubmitButtonOptions(uiSchema);
    if (norender) {
      return null;
    }
    return /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsx("button", { type: "submit", ...submitButtonProps, className: `btn btn-info ${submitButtonProps.className || ""}`, children: submitText }) });
  }

  // src/components/templates/ButtonTemplates/index.ts
  function buttonTemplates() {
    return {
      SubmitButton,
      AddButton,
      CopyButton,
      MoveDownButton,
      MoveUpButton,
      RemoveButton,
      ClearButton
    };
  }
  var ButtonTemplates_default = buttonTemplates;
  var TEST_IDS = utils.getTestIds();
  function RichDescription({ description, registry, uiSchema = {} }) {
    const { globalUiOptions } = registry;
    const uiOptions = utils.getUiOptions(uiSchema, globalUiOptions);
    if (uiOptions.enableMarkdownInDescription && typeof description === "string") {
      return /* @__PURE__ */ jsxRuntime.jsx(react$1.Markdown, { options: { disableParsingRawHTML: true }, "data-testid": TEST_IDS.markdown, children: description });
    }
    return description;
  }
  RichDescription.TEST_IDS = TEST_IDS;
  function DescriptionField(props) {
    const { id, description, registry, uiSchema } = props;
    if (!description) {
      return null;
    }
    return /* @__PURE__ */ jsxRuntime.jsx("div", { id, className: "field-description", children: /* @__PURE__ */ jsxRuntime.jsx(RichDescription, { description, registry, uiSchema }) });
  }
  function ErrorList({
    errors,
    registry
  }) {
    const { translateString } = registry;
    return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "panel panel-danger errors", children: [
      /* @__PURE__ */ jsxRuntime.jsx("div", { className: "panel-heading", children: /* @__PURE__ */ jsxRuntime.jsx("h3", { className: "panel-title", children: translateString(utils.TranslatableString.ErrorsLabel) }) }),
      /* @__PURE__ */ jsxRuntime.jsx("ul", { className: "list-group", children: errors.map((error, i) => (
        // oxlint-disable-next-line react/no-array-index-key
        /* @__PURE__ */ jsxRuntime.jsx("li", { className: "list-group-item text-danger", children: error.stack }, i)
      )) })
    ] });
  }
  function FallbackFieldTemplate(props) {
    const { schema, registry, typeSelector, schemaField } = props;
    const MultiSchemaFieldTemplate2 = utils.getTemplate(
      "MultiSchemaFieldTemplate",
      registry
    );
    return /* @__PURE__ */ jsxRuntime.jsx(
      MultiSchemaFieldTemplate2,
      {
        selector: typeSelector,
        optionSchemaField: schemaField,
        schema,
        registry
      }
    );
  }
  function FieldErrorTemplate(props) {
    const { errors = [], fieldPathId } = props;
    if (errors.length === 0) {
      return null;
    }
    const id = utils.errorId(fieldPathId);
    return /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsx("ul", { id, className: "error-detail bs-callout bs-callout-info", children: errors.filter((elem) => !!elem).map((error, index) => (
      // oxlint-disable-next-line react/no-array-index-key
      /* @__PURE__ */ jsxRuntime.jsx("li", { className: "text-danger", children: error }, index)
    )) }) });
  }
  var TEST_IDS2 = utils.getTestIds();
  function RichHelp({
    help,
    registry,
    uiSchema = {}
  }) {
    const { globalUiOptions } = registry;
    const uiOptions = utils.getUiOptions(uiSchema, globalUiOptions);
    if (uiOptions.enableMarkdownInHelp && typeof help === "string") {
      return /* @__PURE__ */ jsxRuntime.jsx(react$1.Markdown, { options: { disableParsingRawHTML: true }, "data-testid": TEST_IDS2.markdown, children: help });
    }
    return help;
  }
  RichHelp.TEST_IDS = TEST_IDS2;
  function FieldHelpTemplate(props) {
    const { fieldPathId, help, uiSchema, registry } = props;
    if (!help) {
      return null;
    }
    return /* @__PURE__ */ jsxRuntime.jsx("div", { id: utils.helpId(fieldPathId), className: "help-block", children: /* @__PURE__ */ jsxRuntime.jsx(RichHelp, { help, registry, uiSchema }) });
  }
  var REQUIRED_FIELD_SYMBOL = "*";
  function Label(props) {
    const { label, required, id } = props;
    if (!label) {
      return null;
    }
    return /* @__PURE__ */ jsxRuntime.jsxs("label", { className: "control-label", htmlFor: id, children: [
      label,
      required && /* @__PURE__ */ jsxRuntime.jsx("span", { className: "required", children: REQUIRED_FIELD_SYMBOL })
    ] });
  }
  function FieldTemplate(props) {
    const { id, label, children, errors, help, description, hidden, required, displayLabel, registry, uiSchema } = props;
    const uiOptions = utils.getUiOptions(uiSchema);
    const WrapIfAdditionalTemplate2 = utils.getTemplate(
      "WrapIfAdditionalTemplate",
      registry,
      uiOptions
    );
    if (hidden) {
      return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "hidden", children });
    }
    const isCheckbox = uiOptions.widget === "checkbox";
    return /* @__PURE__ */ jsxRuntime.jsxs(WrapIfAdditionalTemplate2, { ...props, children: [
      displayLabel && !isCheckbox && /* @__PURE__ */ jsxRuntime.jsx(Label, { label, required, id }),
      displayLabel && description ? description : null,
      children,
      errors,
      help
    ] });
  }

  // src/components/templates/FieldTemplate/index.ts
  var FieldTemplate_default = FieldTemplate;
  function GridTemplate(props) {
    const { children, column, className, ...rest } = props;
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className, ...rest, children });
  }
  function MultiSchemaFieldTemplate(props) {
    const { selector, optionSchemaField } = props;
    return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "panel panel-default panel-body", children: [
      /* @__PURE__ */ jsxRuntime.jsx("div", { className: "form-group", children: selector }),
      optionSchemaField
    ] });
  }
  function ObjectFieldTemplate(props) {
    const {
      className,
      description,
      disabled,
      formData,
      fieldPathId,
      onAddProperty,
      optionalDataControl,
      properties,
      readonly,
      registry,
      required,
      schema,
      title,
      uiSchema
    } = props;
    const options = utils.getUiOptions(uiSchema);
    const TitleFieldTemplate = utils.getTemplate("TitleFieldTemplate", registry, options);
    const DescriptionFieldTemplate = utils.getTemplate(
      "DescriptionFieldTemplate",
      registry,
      options
    );
    const isPureUnionSchema = (schema.oneOf || schema.anyOf) && !schema.properties && properties.length === 0;
    if (isPureUnionSchema) {
      return null;
    }
    const showOptionalDataControlInTitle = !readonly && !disabled;
    const {
      ButtonTemplates: { AddButton: AddButton2 }
    } = registry.templates;
    return /* @__PURE__ */ jsxRuntime.jsxs("fieldset", { className, id: fieldPathId.$id, children: [
      title && /* @__PURE__ */ jsxRuntime.jsx(
        TitleFieldTemplate,
        {
          id: utils.titleId(fieldPathId),
          title,
          required,
          schema,
          uiSchema,
          registry,
          optionalDataControl: showOptionalDataControlInTitle ? optionalDataControl : void 0
        }
      ),
      description && /* @__PURE__ */ jsxRuntime.jsx(
        DescriptionFieldTemplate,
        {
          id: utils.descriptionId(fieldPathId),
          description,
          schema,
          uiSchema,
          registry
        }
      ),
      !showOptionalDataControlInTitle ? optionalDataControl : void 0,
      properties.map((prop) => prop.content),
      utils.canExpand(schema, uiSchema, formData) && /* @__PURE__ */ jsxRuntime.jsx(
        AddButton2,
        {
          id: utils.buttonId(fieldPathId, "add"),
          className: "rjsf-object-property-expand",
          onClick: onAddProperty,
          disabled: disabled || readonly,
          uiSchema,
          registry
        }
      )
    ] });
  }
  function OptionalDataControlsTemplate(props) {
    const { id, registry, label, onAddClick, onRemoveClick } = props;
    if (onAddClick) {
      return /* @__PURE__ */ jsxRuntime.jsx(
        IconButton_default,
        {
          id,
          registry,
          icon: "plus",
          className: "rjsf-add-optional-data btn-sm",
          onClick: onAddClick,
          title: label
        }
      );
    }
    if (onRemoveClick) {
      return /* @__PURE__ */ jsxRuntime.jsx(
        IconButton_default,
        {
          id,
          registry,
          icon: "remove",
          className: "rjsf-remove-optional-data btn-sm",
          onClick: onRemoveClick,
          title: label
        }
      );
    }
    return /* @__PURE__ */ jsxRuntime.jsx("em", { id, children: label });
  }
  var REQUIRED_FIELD_SYMBOL2 = "*";
  function TitleField(props) {
    const { id, title, required, optionalDataControl } = props;
    return /* @__PURE__ */ jsxRuntime.jsxs("legend", { id, children: [
      title,
      required && /* @__PURE__ */ jsxRuntime.jsx("span", { className: "required", children: REQUIRED_FIELD_SYMBOL2 }),
      optionalDataControl && /* @__PURE__ */ jsxRuntime.jsx("span", { className: "pull-right", style: { marginBottom: "2px" }, children: optionalDataControl })
    ] });
  }
  function UnsupportedField(props) {
    const { schema, fieldPathId, reason, registry } = props;
    const { translateString } = registry;
    let translateEnum = utils.TranslatableString.UnsupportedField;
    const translateParams = [];
    if (fieldPathId && fieldPathId.$id) {
      translateEnum = utils.TranslatableString.UnsupportedFieldWithId;
      translateParams.push(fieldPathId.$id);
    }
    if (reason) {
      translateEnum = translateEnum === utils.TranslatableString.UnsupportedField ? utils.TranslatableString.UnsupportedFieldWithReason : utils.TranslatableString.UnsupportedFieldWithIdAndReason;
      translateParams.push(reason);
    }
    return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "unsupported-field", children: [
      /* @__PURE__ */ jsxRuntime.jsx("p", { children: /* @__PURE__ */ jsxRuntime.jsx(react$1.Markdown, { options: { disableParsingRawHTML: true }, children: translateString(translateEnum, translateParams) }) }),
      schema && /* @__PURE__ */ jsxRuntime.jsx("pre", { children: JSON.stringify(schema, null, 2) })
    ] });
  }
  var UnsupportedField_default = UnsupportedField;
  function WrapIfAdditionalTemplate(props) {
    const {
      id,
      classNames,
      style,
      disabled,
      displayLabel,
      label,
      onKeyRenameBlur,
      onRemoveProperty,
      rawDescription,
      readonly,
      required,
      schema,
      hideError,
      rawErrors,
      children,
      uiSchema,
      registry
    } = props;
    const { templates: templates2, translateString } = registry;
    const { RemoveButton: RemoveButton2 } = templates2.ButtonTemplates;
    const keyLabel = translateString(utils.TranslatableString.KeyLabel, [label]);
    const additional = utils.ADDITIONAL_PROPERTY_FLAG in schema;
    const hasDescription = !!rawDescription;
    const classNamesList = ["form-group", classNames];
    if (!hideError && rawErrors && rawErrors.length > 0) {
      classNamesList.push("has-error has-danger");
    }
    const uiClassNames = classNamesList.join(" ").trim();
    if (!additional) {
      return /* @__PURE__ */ jsxRuntime.jsx("div", { className: uiClassNames, style, children });
    }
    const margin = hasDescription ? 46 : 26;
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: uiClassNames, style, children: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "row", children: [
      /* @__PURE__ */ jsxRuntime.jsx("div", { className: "col-xs-5 form-additional", children: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "form-group", children: [
        displayLabel && /* @__PURE__ */ jsxRuntime.jsx(Label, { label: keyLabel, required, id: `${id}-key` }),
        displayLabel && rawDescription && /* @__PURE__ */ jsxRuntime.jsx("div", { children: "\xA0" }),
        /* @__PURE__ */ jsxRuntime.jsx(
          "input",
          {
            className: "form-control",
            type: "text",
            id: `${id}-key`,
            onBlur: onKeyRenameBlur,
            defaultValue: label
          },
          label
        )
      ] }) }),
      /* @__PURE__ */ jsxRuntime.jsx("div", { className: "form-additional form-group col-xs-5", children }),
      /* @__PURE__ */ jsxRuntime.jsx("div", { className: "col-xs-2", style: { marginTop: displayLabel ? `${margin}px` : void 0 }, children: /* @__PURE__ */ jsxRuntime.jsx(
        RemoveButton2,
        {
          id: utils.buttonId(id, "remove"),
          className: "rjsf-object-property-remove btn-block",
          style: { border: "0" },
          disabled: disabled || readonly,
          onClick: onRemoveProperty,
          uiSchema,
          registry
        }
      ) })
    ] }) });
  }

  // src/components/templates/index.ts
  function templates() {
    return {
      ArrayFieldDescriptionTemplate,
      ArrayFieldItemTemplate,
      ArrayFieldItemButtonsTemplate,
      ArrayFieldTemplate,
      ArrayFieldTitleTemplate,
      ButtonTemplates: ButtonTemplates_default(),
      BaseInputTemplate,
      DescriptionFieldTemplate: DescriptionField,
      ErrorListTemplate: ErrorList,
      FallbackFieldTemplate,
      FieldTemplate: FieldTemplate_default,
      FieldErrorTemplate,
      FieldHelpTemplate,
      GridTemplate,
      MultiSchemaFieldTemplate,
      ObjectFieldTemplate,
      OptionalDataControlsTemplate,
      TitleFieldTemplate: TitleField,
      UnsupportedFieldTemplate: UnsupportedField_default,
      WrapIfAdditionalTemplate
    };
  }
  var templates_default = templates;
  function AltDateTimeWidget({
    time = true,
    ...props
  }) {
    const { AltDateWidget: AltDateWidget2 } = props.registry.widgets;
    return /* @__PURE__ */ jsxRuntime.jsx(AltDateWidget2, { time, ...props });
  }
  var AltDateTimeWidget_default = AltDateTimeWidget;
  function AltDateWidget(props) {
    const { disabled = false, readonly = false, autofocus = false, options, id, name, registry, onBlur, onFocus } = props;
    const { translateString } = registry;
    const { elements, handleChange, handleClear, handleSetNow } = utils.useAltDateWidgetProps(props);
    return /* @__PURE__ */ jsxRuntime.jsxs("ul", { className: "list-inline", children: [
      elements.map((elemProps, i) => (
        // oxlint-disable-next-line react/no-array-index-key
        /* @__PURE__ */ jsxRuntime.jsx("li", { className: "list-inline-item", children: /* @__PURE__ */ jsxRuntime.jsx(
          utils.DateElement,
          {
            rootId: id,
            name,
            select: handleChange,
            ...elemProps,
            disabled,
            readonly,
            registry,
            onBlur,
            onFocus,
            autofocus: autofocus && i === 0
          }
        ) }, i)
      )),
      (options.hideNowButton !== "undefined" ? !options.hideNowButton : true) && /* @__PURE__ */ jsxRuntime.jsx("li", { className: "list-inline-item", children: /* @__PURE__ */ jsxRuntime.jsx("button", { type: "button", className: "btn btn-info btn-now", onClick: handleSetNow, children: translateString(utils.TranslatableString.NowLabel) }) }),
      (options.hideClearButton !== "undefined" ? !options.hideClearButton : true) && /* @__PURE__ */ jsxRuntime.jsx("li", { className: "list-inline-item", children: /* @__PURE__ */ jsxRuntime.jsx("button", { type: "button", className: "btn btn-warning btn-clear", onClick: handleClear, children: translateString(utils.TranslatableString.ClearLabel) }) })
    ] });
  }
  var AltDateWidget_default = AltDateWidget;
  function CheckboxesWidget({
    id,
    disabled,
    options,
    value,
    autofocus = false,
    readonly,
    onChange,
    onBlur,
    onFocus,
    htmlName
  }) {
    const { inline = false, enumOptions, enumDisabled, emptyValue } = options;
    const optionValueFormat = utils.getOptionValueFormat(options);
    const checkboxesValues = Array.isArray(value) ? value : [value];
    const handleBlur = react.useCallback(
      ({ target }) => onBlur(id, utils.enumOptionValueDecoder(target && target.value, enumOptions, optionValueFormat, emptyValue)),
      [onBlur, id, enumOptions, emptyValue, optionValueFormat]
    );
    const handleFocus = react.useCallback(
      ({ target }) => onFocus(id, utils.enumOptionValueDecoder(target && target.value, enumOptions, optionValueFormat, emptyValue)),
      [onFocus, id, enumOptions, emptyValue, optionValueFormat]
    );
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "checkboxes", id, children: Array.isArray(enumOptions) && enumOptions.map((option, index) => {
      const checked = utils.enumOptionsIsSelected(option.value, checkboxesValues);
      const itemDisabled = Array.isArray(enumDisabled) && enumDisabled.includes(option.value);
      const disabledCls = disabled || itemDisabled || readonly ? "disabled" : "";
      const handleChange = (event) => {
        if (event.target.checked) {
          onChange(utils.enumOptionsSelectValue(index, checkboxesValues, enumOptions));
        } else {
          onChange(utils.enumOptionsDeselectValue(index, checkboxesValues, enumOptions));
        }
      };
      const checkbox = /* @__PURE__ */ jsxRuntime.jsxs("span", { children: [
        /* @__PURE__ */ jsxRuntime.jsx(
          "input",
          {
            type: "checkbox",
            id: utils.optionId(id, index),
            name: htmlName || id,
            checked,
            value: utils.enumOptionValueEncoder(option.value, index, optionValueFormat),
            disabled: disabled || itemDisabled || readonly,
            autoFocus: autofocus && index === 0,
            onChange: handleChange,
            onBlur: handleBlur,
            onFocus: handleFocus,
            "aria-describedby": utils.ariaDescribedByIds(id)
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx("span", { children: option.label })
      ] });
      return inline ? /* @__PURE__ */ jsxRuntime.jsx("label", { className: `checkbox-inline ${disabledCls}`, children: checkbox }, String(option.value)) : /* @__PURE__ */ jsxRuntime.jsx("div", { className: `checkbox ${disabledCls}`, children: /* @__PURE__ */ jsxRuntime.jsx("label", { children: checkbox }) }, String(option.value));
    }) });
  }
  var CheckboxesWidget_default = CheckboxesWidget;
  function CheckboxWidget({
    schema,
    uiSchema,
    options,
    id,
    value,
    disabled,
    readonly,
    label,
    hideLabel,
    autofocus = false,
    onBlur,
    onFocus,
    onChange,
    registry,
    htmlName
  }) {
    const DescriptionFieldTemplate = utils.getTemplate(
      "DescriptionFieldTemplate",
      registry,
      options
    );
    const required = utils.schemaRequiresTrueValue(schema);
    const handleChange = react.useCallback(
      (event) => onChange(event.target.checked),
      [onChange]
    );
    const handleBlur = react.useCallback(
      (event) => onBlur(id, event.target.checked),
      [onBlur, id]
    );
    const handleFocus = react.useCallback(
      (event) => onFocus(id, event.target.checked),
      [onFocus, id]
    );
    const uiOptions = utils.getUiOptions(uiSchema);
    const isCheckboxWidget = uiOptions.widget === "checkbox";
    const description = isCheckboxWidget ? void 0 : options.description ?? schema.description;
    return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: `checkbox ${disabled || readonly ? "disabled" : ""}`, children: [
      !hideLabel && description && /* @__PURE__ */ jsxRuntime.jsx(
        DescriptionFieldTemplate,
        {
          id: utils.descriptionId(id),
          description,
          schema,
          uiSchema,
          registry
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsxs("label", { children: [
        /* @__PURE__ */ jsxRuntime.jsx(
          "input",
          {
            type: "checkbox",
            id,
            name: htmlName || id,
            checked: typeof value === "undefined" ? false : value,
            required,
            disabled: disabled || readonly,
            autoFocus: autofocus,
            onChange: handleChange,
            onBlur: handleBlur,
            onFocus: handleFocus,
            "aria-describedby": utils.ariaDescribedByIds(id)
          }
        ),
        utils.labelValue(/* @__PURE__ */ jsxRuntime.jsx("span", { children: label }), hideLabel)
      ] })
    ] });
  }
  var CheckboxWidget_default = CheckboxWidget;
  function ColorWidget(props) {
    const { disabled, readonly, options, registry } = props;
    const BaseInputTemplate2 = utils.getTemplate("BaseInputTemplate", registry, options);
    return /* @__PURE__ */ jsxRuntime.jsx(BaseInputTemplate2, { type: "color", ...props, disabled: disabled || readonly });
  }
  function DateTimeWidget(props) {
    const { onChange, value, options, registry } = props;
    const BaseInputTemplate2 = utils.getTemplate("BaseInputTemplate", registry, options);
    return /* @__PURE__ */ jsxRuntime.jsx(
      BaseInputTemplate2,
      {
        type: "datetime-local",
        ...props,
        value: utils.utcToLocal(value),
        onChange: (newValue) => onChange(utils.localToUTC(newValue))
      }
    );
  }
  function DateWidget(props) {
    const { onChange, options, registry } = props;
    const BaseInputTemplate2 = utils.getTemplate("BaseInputTemplate", registry, options);
    const handleChange = react.useCallback((value) => onChange(value || void 0), [onChange]);
    return /* @__PURE__ */ jsxRuntime.jsx(BaseInputTemplate2, { type: "date", ...props, onChange: handleChange });
  }
  function EmailWidget(props) {
    const { options, registry } = props;
    const BaseInputTemplate2 = utils.getTemplate("BaseInputTemplate", registry, options);
    return /* @__PURE__ */ jsxRuntime.jsx(BaseInputTemplate2, { type: "email", ...props });
  }
  function FileInfoPreview({
    fileInfo,
    registry
  }) {
    const { translateString } = registry;
    const { dataURL, type, name } = fileInfo;
    if (!dataURL) {
      return null;
    }
    const previewLabel = translateString(utils.TranslatableString.PreviewLabel);
    if (["image/jpeg", "image/png"].includes(type)) {
      return /* @__PURE__ */ jsxRuntime.jsx("img", { src: dataURL, alt: previewLabel, style: { maxWidth: "100%" }, className: "file-preview" });
    }
    return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
      " ",
      /* @__PURE__ */ jsxRuntime.jsx("a", { download: `preview-${name}`, href: dataURL, className: "file-download", children: previewLabel })
    ] });
  }
  function FilesInfo({
    filesInfo,
    registry,
    preview,
    onRemove,
    options
  }) {
    if (filesInfo.length === 0) {
      return null;
    }
    const { translateString } = registry;
    const { RemoveButton: RemoveButton2 } = utils.getTemplate("ButtonTemplates", registry, options);
    return /* @__PURE__ */ jsxRuntime.jsx("ul", { className: "file-info", children: filesInfo.map((fileInfo, key) => {
      const { name, size, type } = fileInfo;
      const handleRemove = () => onRemove(key);
      return (
        // oxlint-disable-next-line react/no-array-index-key
        /* @__PURE__ */ jsxRuntime.jsxs("li", { children: [
          /* @__PURE__ */ jsxRuntime.jsx(react$1.Markdown, { children: translateString(utils.TranslatableString.FilesInfo, [name, type, String(size)]) }),
          preview && /* @__PURE__ */ jsxRuntime.jsx(FileInfoPreview, { fileInfo, registry }),
          /* @__PURE__ */ jsxRuntime.jsx(RemoveButton2, { onClick: handleRemove, registry })
        ] }, key)
      );
    }) });
  }
  function FileWidget(props) {
    const { disabled, readonly, required, multiple, onChange, value, options, registry } = props;
    const { filesInfo, handleChange, handleRemove } = utils.useFileWidgetProps(value, onChange, multiple);
    const BaseInputTemplate2 = utils.getTemplate("BaseInputTemplate", registry, options);
    const handleOnChangeEvent = (event) => {
      if (event.target.files) {
        void handleChange(event.target.files);
      }
    };
    return /* @__PURE__ */ jsxRuntime.jsxs("div", { children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        BaseInputTemplate2,
        {
          ...props,
          disabled: disabled || readonly,
          type: "file",
          required: value ? false : required,
          onChangeOverride: handleOnChangeEvent,
          value: "",
          accept: options.accept ? String(options.accept) : void 0
        }
      ),
      /* @__PURE__ */ jsxRuntime.jsx(
        FilesInfo,
        {
          filesInfo,
          onRemove: handleRemove,
          registry,
          preview: options.filePreview,
          options
        }
      )
    ] });
  }
  var FileWidget_default = FileWidget;
  function HiddenWidget({
    id,
    value,
    htmlName
  }) {
    return /* @__PURE__ */ jsxRuntime.jsx("input", { type: "hidden", id, name: htmlName || id, value: typeof value === "undefined" ? "" : value });
  }
  var HiddenWidget_default = HiddenWidget;
  function PasswordWidget(props) {
    const { options, registry } = props;
    const BaseInputTemplate2 = utils.getTemplate("BaseInputTemplate", registry, options);
    return /* @__PURE__ */ jsxRuntime.jsx(BaseInputTemplate2, { type: "password", ...props });
  }
  function RadioWidget({
    options,
    value,
    required,
    disabled,
    readonly,
    autofocus = false,
    onBlur,
    onFocus,
    onChange,
    id,
    htmlName
  }) {
    const { enumOptions, enumDisabled, inline, emptyValue } = options;
    const optionValueFormat = utils.getOptionValueFormat(options);
    const handleBlur = react.useCallback(
      ({ target }) => onBlur(id, utils.enumOptionValueDecoder(target && target.value, enumOptions, optionValueFormat, emptyValue)),
      [onBlur, enumOptions, emptyValue, id, optionValueFormat]
    );
    const handleFocus = react.useCallback(
      ({ target }) => onFocus(id, utils.enumOptionValueDecoder(target && target.value, enumOptions, optionValueFormat, emptyValue)),
      [onFocus, enumOptions, emptyValue, id, optionValueFormat]
    );
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "field-radio-group", id, role: "radiogroup", children: Array.isArray(enumOptions) && enumOptions.map((option, i) => {
      const checked = utils.enumOptionsIsSelected(option.value, value);
      const itemDisabled = Array.isArray(enumDisabled) && enumDisabled.includes(option.value);
      const disabledCls = disabled || itemDisabled || readonly ? "disabled" : "";
      const handleChange = () => onChange(option.value);
      const radio = /* @__PURE__ */ jsxRuntime.jsxs("span", { children: [
        /* @__PURE__ */ jsxRuntime.jsx(
          "input",
          {
            type: "radio",
            id: utils.optionId(id, i),
            checked,
            name: htmlName || id,
            required,
            value: utils.enumOptionValueEncoder(option.value, i, optionValueFormat),
            disabled: disabled || itemDisabled || readonly,
            autoFocus: autofocus && i === 0,
            onChange: handleChange,
            onBlur: handleBlur,
            onFocus: handleFocus,
            "aria-describedby": utils.ariaDescribedByIds(id)
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx("span", { children: option.label })
      ] });
      return inline ? /* @__PURE__ */ jsxRuntime.jsx("label", { className: `radio-inline ${disabledCls}`, children: radio }, String(option.value)) : /* @__PURE__ */ jsxRuntime.jsx("div", { className: `radio ${disabledCls}`, children: /* @__PURE__ */ jsxRuntime.jsx("label", { children: radio }) }, String(option.value));
    }) });
  }
  var RadioWidget_default = RadioWidget;
  function RangeWidget(props) {
    const {
      value,
      registry: {
        templates: { BaseInputTemplate: BaseInputTemplate2 }
      }
    } = props;
    return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "field-range-wrapper", children: [
      /* @__PURE__ */ jsxRuntime.jsx(BaseInputTemplate2, { type: "range", ...props }),
      /* @__PURE__ */ jsxRuntime.jsx("span", { className: "range-view", children: value })
    ] });
  }
  function RatingWidget({
    id,
    value,
    required,
    disabled,
    readonly,
    autofocus,
    onChange,
    onFocus,
    onBlur,
    schema,
    options,
    htmlName
  }) {
    const { stars = 5, shape = "star" } = options;
    const numStars = schema.maximum ? Math.min(schema.maximum, 5) : Math.min(Math.max(stars, 1), 5);
    const min = schema.minimum || 0;
    const handleStarClick = react.useCallback(
      (starValue) => {
        if (!disabled && !readonly) {
          onChange(starValue);
        }
      },
      [onChange, disabled, readonly]
    );
    const handleFocus = react.useCallback(
      (event) => {
        if (onFocus) {
          const starValue = Number(event.target.dataset.value);
          onFocus(id, starValue);
        }
      },
      [onFocus, id]
    );
    const handleBlur = react.useCallback(
      (event) => {
        if (onBlur) {
          const starValue = Number(event.target.dataset.value);
          onBlur(id, starValue);
        }
      },
      [onBlur, id]
    );
    const getSymbol = (isFilled) => {
      if (shape === "heart") {
        return isFilled ? "\u2665" : "\u2661";
      }
      return isFilled ? "\u2605" : "\u2606";
    };
    return /* @__PURE__ */ jsxRuntime.jsxs(
      "div",
      {
        className: "rating-widget",
        style: {
          display: "inline-flex",
          fontSize: "1.5rem",
          cursor: disabled || readonly ? "default" : "pointer"
        },
        children: [
          [...Array(numStars)].map((_, index) => {
            const starValue = min + index;
            const isFilled = starValue <= value;
            return /* @__PURE__ */ jsxRuntime.jsx(
              "span",
              {
                onClick: () => handleStarClick(starValue),
                onKeyDown: (e) => (e.key === "Enter" || e.key === " ") && handleStarClick(starValue),
                onFocus: handleFocus,
                onBlur: handleBlur,
                "data-value": starValue,
                tabIndex: disabled || readonly ? -1 : 0,
                role: "radio",
                "aria-checked": starValue === value,
                "aria-label": `${starValue} ${shape === "heart" ? "heart" : "star"}${starValue === 1 ? "" : "s"}`,
                style: {
                  color: isFilled ? "#FFD700" : "#ccc",
                  padding: "0 0.2rem",
                  transition: "color 0.2s",
                  userSelect: "none"
                },
                children: getSymbol(isFilled)
              },
              index
            );
          }),
          /* @__PURE__ */ jsxRuntime.jsx(
            "input",
            {
              type: "hidden",
              id,
              name: htmlName || id,
              value: value || "",
              required,
              disabled: disabled || readonly,
              "aria-hidden": "true"
            }
          )
        ]
      }
    );
  }
  function getValue(event, multiple) {
    if (multiple) {
      return Array.from(event.target.options).slice().filter((o) => o.selected).map((o) => o.value);
    }
    return event.target.value;
  }
  function SelectWidget({
    schema,
    id,
    options,
    value,
    required,
    disabled,
    readonly,
    multiple = false,
    autofocus = false,
    onChange,
    onBlur,
    onFocus,
    placeholder,
    htmlName
  }) {
    const { enumOptions, enumDisabled, emptyValue: optEmptyVal } = options;
    const emptyValue = multiple ? [] : "";
    const optionValueFormat = utils.getOptionValueFormat(options);
    const handleFocus = react.useCallback(
      (event) => {
        const newValue = getValue(event, multiple);
        return onFocus(id, utils.enumOptionValueDecoder(newValue, enumOptions, optionValueFormat, optEmptyVal));
      },
      [onFocus, id, multiple, enumOptions, optEmptyVal, optionValueFormat]
    );
    const handleBlur = react.useCallback(
      (event) => {
        const newValue = getValue(event, multiple);
        return onBlur(id, utils.enumOptionValueDecoder(newValue, enumOptions, optionValueFormat, optEmptyVal));
      },
      [onBlur, id, multiple, enumOptions, optEmptyVal, optionValueFormat]
    );
    const handleChange = react.useCallback(
      (event) => {
        const newValue = getValue(event, multiple);
        return onChange(utils.enumOptionValueDecoder(newValue, enumOptions, optionValueFormat, optEmptyVal));
      },
      [onChange, multiple, enumOptions, optEmptyVal, optionValueFormat]
    );
    const selectValue = utils.enumOptionSelectedValue(value, enumOptions, multiple, optionValueFormat, emptyValue);
    const showPlaceholderOption = !multiple && schema.default === void 0;
    return (
      // oxlint-disable-next-line jsx-a11y/no-autofocus
      /* @__PURE__ */ jsxRuntime.jsxs(
        "select",
        {
          id,
          name: htmlName || id,
          multiple,
          className: "form-control",
          value: selectValue,
          required,
          disabled: disabled || readonly,
          autoFocus: autofocus,
          onBlur: handleBlur,
          onFocus: handleFocus,
          onChange: handleChange,
          "aria-describedby": utils.ariaDescribedByIds(id),
          children: [
            showPlaceholderOption && /* @__PURE__ */ jsxRuntime.jsx("option", { value: "", children: placeholder }),
            Array.isArray(enumOptions) && enumOptions.map(({ value: enumValue, label: enumLabel }, i) => {
              const isDisabled = enumDisabled && enumDisabled.includes(enumValue);
              return /* @__PURE__ */ jsxRuntime.jsx(
                "option",
                {
                  value: utils.enumOptionValueEncoder(enumValue, i, optionValueFormat),
                  disabled: isDisabled,
                  children: enumLabel
                },
                String(enumValue)
              );
            })
          ]
        }
      )
    );
  }
  var SelectWidget_default = SelectWidget;
  function TextareaWidget({
    id,
    options = {},
    placeholder,
    value,
    required,
    disabled,
    readonly,
    autofocus = false,
    onChange,
    onBlur,
    onFocus,
    htmlName
  }) {
    const handleChange = react.useCallback(
      ({ target: { value: newValue } }) => onChange(newValue === "" ? options.emptyValue : newValue),
      [onChange, options.emptyValue]
    );
    const handleBlur = react.useCallback(
      ({ target }) => onBlur(id, target && target.value),
      [onBlur, id]
    );
    const handleFocus = react.useCallback(
      ({ target }) => onFocus(id, target && target.value),
      [id, onFocus]
    );
    return /* @__PURE__ */ jsxRuntime.jsx(
      "textarea",
      {
        id,
        name: htmlName || id,
        className: "form-control",
        value: value || "",
        placeholder,
        required,
        disabled,
        readOnly: readonly,
        autoFocus: autofocus,
        rows: options.rows,
        onBlur: handleBlur,
        onFocus: handleFocus,
        onChange: handleChange,
        "aria-describedby": utils.ariaDescribedByIds(id)
      }
    );
  }
  var TextareaWidget_default = TextareaWidget;
  function TextWidget(props) {
    const { options, registry } = props;
    const BaseInputTemplate2 = utils.getTemplate("BaseInputTemplate", registry, options);
    return /* @__PURE__ */ jsxRuntime.jsx(BaseInputTemplate2, { ...props });
  }
  function TimeWidget(props) {
    const { onChange, options, registry } = props;
    const BaseInputTemplate2 = utils.getTemplate("BaseInputTemplate", registry, options);
    const handleChange = react.useCallback((value) => onChange(value ? `${value}:00` : void 0), [onChange]);
    return /* @__PURE__ */ jsxRuntime.jsx(BaseInputTemplate2, { type: "time", ...props, onChange: handleChange });
  }
  function UpDownWidget(props) {
    const { options, registry } = props;
    const BaseInputTemplate2 = utils.getTemplate("BaseInputTemplate", registry, options);
    return /* @__PURE__ */ jsxRuntime.jsx(BaseInputTemplate2, { type: "number", ...props });
  }
  function URLWidget(props) {
    const { options, registry } = props;
    const BaseInputTemplate2 = utils.getTemplate("BaseInputTemplate", registry, options);
    return /* @__PURE__ */ jsxRuntime.jsx(BaseInputTemplate2, { type: "url", ...props });
  }

  // src/components/widgets/index.ts
  function widgets() {
    return {
      AltDateWidget: AltDateWidget_default,
      AltDateTimeWidget: AltDateTimeWidget_default,
      CheckboxWidget: CheckboxWidget_default,
      CheckboxesWidget: CheckboxesWidget_default,
      ColorWidget,
      DateWidget,
      DateTimeWidget,
      EmailWidget,
      FileWidget: FileWidget_default,
      HiddenWidget: HiddenWidget_default,
      PasswordWidget,
      RadioWidget: RadioWidget_default,
      RangeWidget,
      RatingWidget,
      SelectWidget: SelectWidget_default,
      TextWidget,
      TextareaWidget: TextareaWidget_default,
      TimeWidget,
      UpDownWidget,
      URLWidget
    };
  }
  var widgets_default = widgets;

  // src/getDefaultRegistry.ts
  function getDefaultRegistry() {
    return {
      fields: fields_default(),
      templates: templates_default(),
      widgets: widgets_default(),
      rootSchema: {},
      formContext: {},
      translateString: utils.englishStringTranslator,
      globalFormOptions: {
        idPrefix: utils.DEFAULT_ID_PREFIX,
        idSeparator: utils.DEFAULT_ID_SEPARATOR,
        useFallbackUiForUnsupportedType: false
      }
    };
  }
  function toIChangeEvent(state, status) {
    return {
      ..._pick(state, ["schema", "uiSchema", "fieldPathId", "schemaUtils", "formData", "edit", "errors", "errorSchema"]),
      ...status !== void 0 && { status }
    };
  }
  var Form = class _Form extends react.Component {
    /** The ref used to hold the `form` element, this needs to be `any` because `tagName` or `_internalFormWrapper` can
     * provide any possible type here
     */
    formElement;
    /** The list of pending changes
     */
    pendingChanges = [];
    /** Flag to track when we're processing a user-initiated field change.
     * This prevents componentDidUpdate from reverting oneOf/anyOf option switches.
     */
    isProcessingUserChange = false;
    /** When the `extraErrors` prop changes, re-merges `schemaValidationErrors` + `extraErrors` + `customErrors` into
     * state before render, ensuring the updated errors are visible immediately in a single render cycle.
     *
     * @param props - The current props
     * @param state - The current state
     * @returns Partial state with re-merged errors if `extraErrors` changed, or `null` if no update is needed
     */
    static getDerivedStateFromProps(props, state) {
      if (props.extraErrors !== state.prevExtraErrors) {
        const baseErrors = {
          errors: state.schemaValidationErrors || [],
          errorSchema: state.schemaValidationErrorSchema || {}
        };
        let { errors, errorSchema } = baseErrors;
        if (props.extraErrors) {
          ({ errors, errorSchema } = utils.validationDataMerge(baseErrors, props.extraErrors));
        }
        if (state.customErrors) {
          ({ errors, errorSchema } = utils.validationDataMerge(
            { errors, errorSchema },
            state.customErrors.ErrorSchema,
            true
          ));
        }
        return { prevExtraErrors: props.extraErrors, errors, errorSchema };
      }
      return null;
    }
    /** Constructs the `Form` from the `props`. Will setup the initial state from the props. It will also call the
     * `onChange` handler if the initially provided `formData` is modified to add missing default values as part of the
     * state construction.
     *
     * @param props - The initial props for the `Form`
     */
    constructor(props) {
      super(props);
      if (!props.validator) {
        throw new Error("A validator is required for Form functionality to work");
      }
      const { formData: propsFormData, initialFormData, onChange } = props;
      const formData = propsFormData ?? initialFormData;
      this.state = {
        ...this.getStateFromProps(props, formData, void 0, void 0, void 0, true),
        prevExtraErrors: props.extraErrors
      };
      if (onChange && !utils.deepEquals(this.state.formData, formData)) {
        onChange(toIChangeEvent(this.state));
      }
      this.formElement = react.createRef();
    }
    /**
     * `getSnapshotBeforeUpdate` is a React lifecycle method that is invoked right before the most recently rendered
     * output is committed to the DOM. It enables your component to capture current values (e.g., scroll position) before
     * they are potentially changed.
     *
     * In this case, it checks if the props have changed since the last render. If they have, it computes the next state
     * of the component using `getStateFromProps` method and returns it along with a `shouldUpdate` flag set to `true` IF
     * the `nextState` and `prevState` are different, otherwise `false`. This ensures that we have the most up-to-date
     * state ready to be applied in `componentDidUpdate`.
     *
     * If `formData` hasn't changed, it simply returns an object with `shouldUpdate` set to `false`, indicating that a
     * state update is not necessary.
     *
     * @param prevProps - The previous set of props before the update.
     * @param prevState - The previous state before the update.
     * @returns Either an object containing the next state and a flag indicating that an update should occur, or an object
     *        with a flag indicating that an update is not necessary.
     */
    getSnapshotBeforeUpdate(prevProps, prevState) {
      if (!utils.deepEquals(this.props, prevProps)) {
        const formDataChangedFields = utils.getChangedFields(this.props.formData, prevProps.formData);
        const stateDataChangedFields = utils.getChangedFields(this.props.formData, this.state.formData);
        const isSchemaChanged = !utils.deepEquals(prevProps.schema, this.props.schema);
        const isFormDataChanged = formDataChangedFields.length > 0 || !utils.deepEquals(prevProps.formData, this.props.formData);
        const isStateDataChanged = stateDataChangedFields.length > 0 || !utils.deepEquals(this.state.formData, this.props.formData);
        const nextState = this.getStateFromProps(
          this.props,
          this.props.formData,
          // If the `schema` has changed, we need to update the retrieved schema.
          // Or if the `formData` changes, for example in the case of a schema with dependencies that need to
          //  match one of the subSchemas, the retrieved schema must be updated.
          isSchemaChanged || isFormDataChanged ? void 0 : this.state.retrievedSchema,
          isSchemaChanged,
          formDataChangedFields,
          // Skip live validation for this request if no form data has changed from the last state
          !isStateDataChanged
        );
        const shouldUpdate = !utils.deepEquals(nextState, prevState);
        return { nextState, shouldUpdate };
      }
      return { shouldUpdate: false };
    }
    /**
     * `componentDidUpdate` is a React lifecycle method that is invoked immediately after updating occurs. This method is
     * not called for the initial render.
     *
     * Here, it checks if an update is necessary based on the `shouldUpdate` flag received from `getSnapshotBeforeUpdate`.
     * If an update is required, it applies the next state and, if needed, triggers the `onChange` handler to inform about
     * changes.
     *
     * @param _ - The previous set of props.
     * @param prevState - The previous state of the component before the update.
     * @param snapshot - The value returned from `getSnapshotBeforeUpdate`.
     */
    componentDidUpdate(_, prevState, snapshot) {
      if (snapshot.shouldUpdate) {
        const { nextState } = snapshot;
        const nextStateDiffersFromProps = !utils.deepEquals(nextState.formData, this.props.formData);
        const wasProcessingUserChange = this.isProcessingUserChange;
        this.isProcessingUserChange = false;
        if (wasProcessingUserChange && nextStateDiffersFromProps) {
          return;
        }
        if (nextStateDiffersFromProps && !utils.deepEquals(nextState.formData, prevState.formData) && this.props.onChange) {
          this.props.onChange(toIChangeEvent(nextState));
        }
        this.setState(nextState);
      }
    }
    /** Extracts the updated state from the given `props` and `inputFormData`. As part of this process, the
     * `inputFormData` is first processed to add any missing required defaults. After that, the data is run through the
     * validation process IF required by the `props`.
     *
     * @param props - The props passed to the `Form`
     * @param inputFormData - The new or current data for the `Form`
     * @param retrievedSchema - An expanded schema, if not provided, it will be retrieved from the `schema` and `formData`.
     * @param [isSchemaChanged=false] - A flag indicating whether the schema has changed.
     * @param [formDataChangedFields=[]] - The changed fields of `formData`
     * @param [skipLiveValidate=false] - Optional flag, if true, means that we are not running live validation
     * @param [shouldSanitize=false] - Optional flag, if true, means that we should attempt to sanitize formData
     * @returns - The new state for the `Form`
     */
    getStateFromProps(props, inputFormData, retrievedSchema, isSchemaChanged = false, formDataChangedFields = [], skipLiveValidate = false, shouldSanitize = false) {
      const state = this.state || {};
      const schema = "schema" in props ? props.schema : this.props.schema;
      const validator2 = "validator" in props ? props.validator : this.props.validator;
      const uiSchema = ("uiSchema" in props ? props.uiSchema : this.props.uiSchema) || {};
      const isUncontrolled = props.formData === void 0 && this.props.formData === void 0;
      const edit = typeof inputFormData !== "undefined";
      const liveValidate = "liveValidate" in props ? props.liveValidate : this.props.liveValidate;
      const mustValidate = edit && !props.noValidate && liveValidate;
      const experimental_defaultFormStateBehavior = "experimental_defaultFormStateBehavior" in props ? props.experimental_defaultFormStateBehavior : this.props.experimental_defaultFormStateBehavior;
      const experimental_customMergeAllOf = "experimental_customMergeAllOf" in props ? props.experimental_customMergeAllOf : this.props.experimental_customMergeAllOf;
      let { schemaUtils } = state;
      if (!schemaUtils || schemaUtils.doesSchemaUtilsDiffer(
        validator2,
        schema,
        experimental_defaultFormStateBehavior,
        experimental_customMergeAllOf
      )) {
        schemaUtils = utils.createSchemaUtils(
          validator2,
          schema,
          experimental_defaultFormStateBehavior,
          experimental_customMergeAllOf
        );
      }
      const rootSchema = schemaUtils.getRootSchema();
      let defaultsFormData = inputFormData;
      if (inputFormData === IS_RESET) {
        defaultsFormData = void 0;
      } else if (inputFormData === void 0 && isUncontrolled) {
        defaultsFormData = state.formData;
      }
      let formData;
      let computedRetrievedSchema;
      let wasSanitized = false;
      const preventInfiniteSanitize = [];
      do {
        formData = schemaUtils.getDefaultFormState(
          rootSchema,
          defaultsFormData,
          false,
          state.initialDefaultsGenerated
        );
        const formHash = shouldSanitize ? utils.hashObject({ formData }) : "";
        computedRetrievedSchema = this.updateRetrievedSchema(
          retrievedSchema ?? schemaUtils.retrieveSchema(rootSchema, formData)
        );
        if (shouldSanitize && !preventInfiniteSanitize.includes(formHash) && !utils.deepEquals(computedRetrievedSchema, state.retrievedSchema)) {
          const sanitizedFormData = schemaUtils.sanitizeDataForNewSchema(
            computedRetrievedSchema,
            state.retrievedSchema,
            formData
          );
          wasSanitized = !utils.deepEquals(sanitizedFormData, formData);
          if (wasSanitized) {
            formData = sanitizedFormData;
            defaultsFormData = sanitizedFormData;
            const sanitizedFormHash = utils.hashObject({ formData: sanitizedFormData });
            wasSanitized = !preventInfiniteSanitize.includes(sanitizedFormHash);
            preventInfiniteSanitize.push(sanitizedFormHash);
          }
          preventInfiniteSanitize.push(formHash);
        } else {
          wasSanitized = false;
        }
      } while (wasSanitized);
      const getCurrentErrors = () => {
        if (props.noValidate || isSchemaChanged) {
          return { errors: [], errorSchema: {} };
        }
        if (!props.liveValidate) {
          return {
            errors: state.schemaValidationErrors || [],
            errorSchema: state.schemaValidationErrorSchema || {}
          };
        }
        return {
          errors: state.errors || [],
          errorSchema: state.errorSchema || {}
        };
      };
      let errors;
      let errorSchema;
      let { schemaValidationErrors, schemaValidationErrorSchema } = state;
      if (mustValidate && !skipLiveValidate) {
        const liveValidation = this.liveValidate(
          rootSchema,
          schemaUtils,
          state.errorSchema,
          formData,
          void 0,
          state.customErrors,
          retrievedSchema,
          // If retrievedSchema is undefined which means the schema or formData has changed, we do not merge state.
          // Else in the case where it hasn't changed,
          retrievedSchema !== void 0
        );
        errors = liveValidation.errors;
        errorSchema = liveValidation.errorSchema;
        schemaValidationErrors = liveValidation.schemaValidationErrors;
        schemaValidationErrorSchema = liveValidation.schemaValidationErrorSchema;
      } else {
        const currentErrors = getCurrentErrors();
        errors = currentErrors.errors;
        errorSchema = currentErrors.errorSchema;
        if (formDataChangedFields.length > 0 && !mustValidate) {
          const newErrorSchema = formDataChangedFields.reduce((acc, key) => {
            acc[key] = void 0;
            return acc;
          }, {});
          schemaValidationErrorSchema = utils.mergeObjects(
            currentErrors.errorSchema,
            newErrorSchema,
            "preventDuplicates"
          );
          errorSchema = schemaValidationErrorSchema;
        }
        const mergedErrors = _Form.mergeErrors({ errorSchema, errors }, props.extraErrors, state.customErrors);
        errors = mergedErrors.errors;
        errorSchema = mergedErrors.errorSchema;
      }
      const newRegistry = _Form.getRegistry(props, rootSchema, schemaUtils);
      const registry = utils.deepEquals(state.registry, newRegistry) ? state.registry : newRegistry;
      const fieldPathId = state.fieldPathId && state.fieldPathId?.[utils.ID_KEY] === registry.globalFormOptions.idPrefix ? state.fieldPathId : utils.toFieldPathId("", registry.globalFormOptions);
      const nextState = {
        schemaUtils,
        schema: rootSchema,
        uiSchema,
        fieldPathId,
        formData,
        edit,
        errors,
        errorSchema,
        schemaValidationErrors,
        schemaValidationErrorSchema,
        retrievedSchema: computedRetrievedSchema,
        initialDefaultsGenerated: true,
        registry
      };
      return nextState;
    }
    /** React lifecycle method that is used to determine whether component should be updated.
     *
     * @param nextProps - The next version of the props
     * @param nextState - The next version of the state
     * @returns - True if the component should be updated, false otherwise
     */
    shouldComponentUpdate(nextProps, nextState) {
      const { experimental_componentUpdateStrategy = "customDeep" } = this.props;
      return utils.shouldRender(this, nextProps, nextState, experimental_componentUpdateStrategy);
    }
    /** Validates the `formData` against the `schema` using the `altSchemaUtils` (if provided otherwise it uses the
     * `schemaUtils` in the state), returning the results.
     *
     * @param formData - The new form data to validate
     * @param schema - The schema used to validate against
     * @param [altSchemaUtils] - The alternate schemaUtils to use for validation
     * @param [retrievedSchema] - An optionally retrieved schema for per
     */
    validate(formData, schema = this.state.schema, altSchemaUtils, retrievedSchema) {
      const schemaUtils = altSchemaUtils || this.state.schemaUtils;
      const { customValidate, transformErrors, uiSchema } = this.props;
      const validationSchema = retrievedSchema ?? schema;
      return schemaUtils.getValidator().validateFormData(formData, validationSchema, customValidate, transformErrors, uiSchema);
    }
    /** Renders any errors contained in the `state` in using the `ErrorList`, if not disabled by `showErrorList`. */
    renderErrors(registry) {
      const { errors, errorSchema, schema, uiSchema } = this.state;
      const options = utils.getUiOptions(uiSchema);
      const ErrorListTemplate = utils.getTemplate("ErrorListTemplate", registry, options);
      if (errors && errors.length) {
        return /* @__PURE__ */ jsxRuntime.jsx(
          ErrorListTemplate,
          {
            errors,
            errorSchema: errorSchema || {},
            schema,
            uiSchema,
            registry
          }
        );
      }
      return null;
    }
    /** Merges any `extraErrors` or `customErrors` into the given `schemaValidation` object, returning the result
     *
     * @param schemaValidation - The `ValidationData` object into which additional errors are merged
     * @param [extraErrors] - The extra errors from the props
     * @param [customErrors] - The customErrors from custom components
     * @return - The `extraErrors` and `customErrors` merged into the `schemaValidation`
     * @private
     */
    static mergeErrors(schemaValidation, extraErrors, customErrors) {
      let { errorSchema, errors } = schemaValidation;
      if (extraErrors) {
        const merged = utils.validationDataMerge(schemaValidation, extraErrors);
        errorSchema = merged.errorSchema;
        errors = merged.errors;
      }
      if (customErrors) {
        const merged = utils.validationDataMerge({ errors, errorSchema }, customErrors.ErrorSchema, true);
        errorSchema = merged.errorSchema;
        errors = merged.errors;
      }
      return { errors, errorSchema };
    }
    /** Performs live validation and then updates and returns the errors and error schemas by potentially merging in
     * `extraErrors` and `customErrors`.
     *
     * @param rootSchema - The `rootSchema` from the state
     * @param schemaUtils - The `SchemaUtilsType` from the state
     * @param originalErrorSchema - The original `ErrorSchema` from the state
     * @param [formData] - The new form data to validate
     * @param [extraErrors] - The extra errors from the props
     * @param [customErrors] - The customErrors from custom components
     * @param [retrievedSchema] - An expanded schema, if not provided, it will be retrieved from the `schema` and `formData`
     * @param [mergeIntoOriginalErrorSchema=false] - Optional flag indicating whether we merge into original schema
     * @returns - An object containing `errorSchema`, `errors`, `schemaValidationErrors` and `schemaValidationErrorSchema`
     * @private
     */
    liveValidate(rootSchema, schemaUtils, originalErrorSchema, formData, extraErrors, customErrors, retrievedSchema, mergeIntoOriginalErrorSchema = false) {
      const schemaValidation = this.validate(formData, rootSchema, schemaUtils, retrievedSchema);
      const { errors } = schemaValidation;
      let { errorSchema } = schemaValidation;
      if (mergeIntoOriginalErrorSchema) {
        errorSchema = utils.mergeObjects(
          originalErrorSchema,
          schemaValidation.errorSchema,
          "preventDuplicates"
        );
      }
      const schemaValidationErrors = errors;
      const schemaValidationErrorSchema = errorSchema;
      const mergedErrors = _Form.mergeErrors({ errorSchema, errors }, extraErrors, customErrors);
      return { ...mergedErrors, schemaValidationErrors, schemaValidationErrorSchema };
    }
    /** Returns the `formData` with only the elements specified in the `fields` list
     *
     * @param formData - The data for the `Form`
     * @param fields - The fields to keep while filtering
     * @deprecated - To be removed as an exported `Form` function in a future release; there isn't a planned replacement
     */
    // oxlint-disable-next-line class-methods-use-this, typescript/no-deprecated
    getUsedFormData = (formData, fields2) => utils.getUsedFormData(formData, fields2);
    /** Returns the list of field names from inspecting the `pathSchema` as well as using the `formData`
     *
     * @param pathSchema - The `PathSchema` object for the form
     * @param [formData] - The form data to use while checking for empty objects/arrays
     * @deprecated - To be removed as an exported `Form` function in a future release; there isn't a planned replacement
     */
    // oxlint-disable-next-line class-methods-use-this, typescript/no-deprecated
    getFieldNames = (pathSchema, formData) => utils.getFieldNames(pathSchema, formData);
    /** Returns the `formData` after filtering to remove any extra data not in a form field
     *
     * @param formData - The data for the `Form`
     * @returns The `formData` after omitting extra data
     * @deprecated - To be removed as an exported `Form` function in a future release, use `SchemaUtils.omitExtraData`
     *               instead.
     */
    omitExtraData = (formData) => {
      const { schema, schemaUtils } = this.state;
      return schemaUtils.omitExtraData(schema, formData);
    };
    /** Allows a user to set a value for the provided `fieldPath`, which must be either a dotted path to the field OR a
     * `FieldPathList`. To set the root element, used either `''` or `[]` for the path. Passing undefined will clear the
     * value in the field.
     *
     * @param fieldPath - Either a dotted path to the field or the `FieldPathList` to the field
     * @param [newValue] - The new value for the field
     */
    setFieldValue = (fieldPath, newValue) => {
      const { registry } = this.state;
      const path = Array.isArray(fieldPath) ? fieldPath : fieldPath.split(".");
      const fieldPathId = utils.toFieldPathId("", registry.globalFormOptions, path);
      this.onChange(newValue, path, void 0, fieldPathId[utils.ID_KEY]);
    };
    /** Pushes the given change information into the `pendingChanges` array and then calls `processPendingChanges()` if
     * the array only contains a single pending change.
     *
     * @param newValue - The new form data from a change to a field
     * @param path - The path to the change into which to set the formData
     * @param [newErrorSchema] - The new `ErrorSchema` based on the field change
     * @param [id] - The id of the field that caused the change
     */
    onChange = (newValue, path, newErrorSchema, id) => {
      this.pendingChanges.push({ newValue, path, newErrorSchema, id });
      if (this.pendingChanges.length === 1) {
        this.processPendingChange();
      }
    };
    /** Function to handle changes made to a field in the `Form`. This handler gets the first change from the
     * `pendingChanges` list, containing the `newValue` for the `formData` and the `path` at which the `newValue` is to be
     * updated, along with a new, optional `ErrorSchema` for that same `path` and potentially the `id` of the field being
     * changed. It will first update the `formData` with any missing default fields and then, if `omitExtraData` and
     * `liveOmit` are turned on, the `formData` will be filtered to remove any extra data not in a form field. Then, the
     * resulting `formData` will be validated if required. The state will be updated with the new updated (potentially
     * filtered) `formData`, any errors that resulted from validation. Finally the `onChange` callback will be called, if
     * specified, with the updated state and the `processPendingChange()` function is called again.
     */
    processPendingChange() {
      if (this.pendingChanges.length === 0) {
        return;
      }
      this.isProcessingUserChange = true;
      const { newValue, path, id } = this.pendingChanges[0];
      const { newErrorSchema } = this.pendingChanges[0];
      const { extraErrors, omitExtraData, liveOmit, noValidate, liveValidate, onChange, disabled, readonly } = this.props;
      const { formData: oldFormData, schemaUtils, schema, fieldPathId, schemaValidationErrorSchema, errors } = this.state;
      let { customErrors, retrievedSchema } = this.state;
      let mergeBaseErrorSchema = schemaValidationErrorSchema;
      const rootPathId = fieldPathId.path[0] || "";
      const isRootPath = !path || path.length === 0 || path.length === 1 && path[0] === rootPathId;
      let formData = isRootPath ? newValue : _cloneDeep(oldFormData);
      const hasOnlyUndefinedValues = utils.isObject(formData) && Object.keys(formData).length > 0 && Object.values(formData).every((v) => v === void 0);
      const wasPreviouslyNull = oldFormData === null || oldFormData === void 0;
      const inputForDefaults = hasOnlyUndefinedValues && wasPreviouslyNull ? void 0 : formData;
      if (utils.isObject(formData) || Array.isArray(formData)) {
        if (newValue === ADDITIONAL_PROPERTY_KEY_REMOVE) {
          _unset(formData, path);
        } else if (!isRootPath) {
          let unsetPath = false;
          let valueForPath = newValue;
          if (newValue === void 0) {
            const lastSegment = path[path.length - 1];
            if (typeof lastSegment === "number") {
              valueForPath = null;
            } else {
              const { field } = schemaUtils.findFieldInSchema(schema, path, oldFormData);
              const leaf = field;
              const isOneOfOrAnyOfLeaf = leaf && (utils.ONE_OF_KEY in leaf || utils.ANY_OF_KEY in leaf);
              if (!isOneOfOrAnyOfLeaf && leaf !== void 0) {
                unsetPath = true;
              }
            }
          }
          if (unsetPath) {
            _unset(formData, path);
          } else {
            set(formData, path, valueForPath);
          }
        }
        const shouldSanitize = retrievedSchema && !isRootPath && !utils.isObject(newValue) && !Array.isArray(newValue) && !disabled && !readonly;
        const newState = this.getStateFromProps(
          this.props,
          inputForDefaults,
          void 0,
          void 0,
          void 0,
          true,
          shouldSanitize
        );
        formData = newState.formData;
        retrievedSchema = newState.retrievedSchema;
      }
      const mustValidate = !noValidate && (liveValidate === true || liveValidate === "onChange");
      let state = { formData, retrievedSchema };
      let newFormData = formData;
      if (omitExtraData === true && (liveOmit === true || liveOmit === "onChange")) {
        newFormData = this.omitExtraData(formData);
        state = { ...state, formData: newFormData };
      }
      if (newErrorSchema) {
        const oldValidationError = !isRootPath ? get(schemaValidationErrorSchema, path) : schemaValidationErrorSchema;
        if (!isEmpty(oldValidationError)) {
          if (!isRootPath) {
            mergeBaseErrorSchema = _cloneDeep(schemaValidationErrorSchema);
            set(mergeBaseErrorSchema, path, newErrorSchema);
          } else {
            mergeBaseErrorSchema = newErrorSchema;
          }
        } else {
          if (!customErrors) {
            customErrors = new utils.ErrorSchemaBuilder();
          }
          if (isRootPath) {
            const pathErrors = get(newErrorSchema, utils.ERRORS_KEY);
            if (pathErrors) {
              customErrors.setErrors(pathErrors);
            }
          } else {
            set(customErrors.ErrorSchema, path, newErrorSchema);
          }
        }
      } else if (customErrors && get(customErrors.ErrorSchema, [...path, utils.ERRORS_KEY])) {
        customErrors.clearErrors(path);
      }
      if (mustValidate && this.pendingChanges.length === 1) {
        const liveValidation = this.liveValidate(
          schema,
          schemaUtils,
          mergeBaseErrorSchema,
          newFormData,
          extraErrors,
          customErrors,
          retrievedSchema
        );
        state = { ...state, formData: newFormData, ...liveValidation, customErrors };
      } else if (!noValidate && newErrorSchema) {
        const mergedErrors = _Form.mergeErrors(
          { errorSchema: mergeBaseErrorSchema, errors },
          extraErrors,
          customErrors
        );
        state = { ...state, formData: newFormData, ...mergedErrors, customErrors };
      }
      this.setState(state, () => {
        if (onChange) {
          onChange(toIChangeEvent({ ...this.state, ...state }), id);
        }
        this.pendingChanges.shift();
        this.processPendingChange();
      });
    }
    /**
     * If the retrievedSchema has changed the new retrievedSchema is returned.
     * Otherwise, the old retrievedSchema is returned to persist reference.
     * -  This ensures that AJV retrieves the schema from the cache when it has not changed,
     *    avoiding the performance cost of recompiling the schema.
     *
     * @param retrievedSchema The new retrieved schema.
     * @returns The new retrieved schema if it has changed, else the old retrieved schema.
     */
    updateRetrievedSchema(retrievedSchema) {
      const isTheSame = utils.deepEquals(retrievedSchema, this.state?.retrievedSchema);
      return isTheSame ? this.state.retrievedSchema : retrievedSchema;
    }
    /**
     * Callback function to handle reset form data.
     * - Reset all fields with default values.
     * - Reset validations and errors
     *
     */
    reset = () => {
      const { formData: propsFormData, initialFormData = IS_RESET, onChange } = this.props;
      const newState = this.getStateFromProps(
        this.props,
        propsFormData ?? initialFormData,
        void 0,
        void 0,
        void 0,
        true
      );
      const newFormData = newState.formData;
      const state = {
        formData: newFormData,
        errorSchema: {},
        errors: [],
        schemaValidationErrors: [],
        schemaValidationErrorSchema: {},
        initialDefaultsGenerated: false,
        customErrors: void 0
      };
      this.setState(state, () => onChange && onChange(toIChangeEvent({ ...this.state, ...state })));
    };
    /** Callback function to handle when a field on the form is blurred. Calls the `onBlur` callback for the `Form` if it
     * was provided. Also runs any live validation and/or live omit operations if the flags indicate they should happen
     * during `onBlur`.
     *
     * @param id - The unique `id` of the field that was blurred
     * @param data - The data associated with the field that was blurred
     */
    onBlur = (id, data) => {
      const { onBlur, omitExtraData, liveOmit, liveValidate } = this.props;
      if (onBlur) {
        onBlur(id, data);
      }
      if (omitExtraData === true && liveOmit === "onBlur" || liveValidate === "onBlur") {
        const { onChange, extraErrors } = this.props;
        const { formData } = this.state;
        let newFormData = formData;
        let state = { formData: newFormData };
        if (omitExtraData === true && liveOmit === "onBlur") {
          newFormData = this.omitExtraData(formData);
          state = { formData: newFormData };
        }
        if (liveValidate === "onBlur") {
          const { schema, schemaUtils, errorSchema, customErrors, retrievedSchema } = this.state;
          const liveValidation = this.liveValidate(
            schema,
            schemaUtils,
            errorSchema,
            newFormData,
            extraErrors,
            customErrors,
            retrievedSchema
          );
          state = { formData: newFormData, ...liveValidation, customErrors };
        }
        const hasChanges = Object.keys(state).filter((key) => !key.startsWith("schemaValidation")).some((key) => {
          const oldData = get(this.state, key);
          const newData = get(state, key);
          return !utils.deepEquals(oldData, newData);
        });
        this.setState(state, () => {
          if (onChange && hasChanges) {
            onChange(toIChangeEvent({ ...this.state, ...state }), id);
          }
        });
      }
    };
    /** Callback function to handle when a field on the form is focused. Calls the `onFocus` callback for the `Form` if it
     * was provided.
     *
     * @param id - The unique `id` of the field that was focused
     * @param data - The data associated with the field that was focused
     */
    onFocus = (id, data) => {
      const { onFocus } = this.props;
      if (onFocus) {
        onFocus(id, data);
      }
    };
    /** Callback function to handle when the form is submitted. First, it prevents the default event behavior. Nothing
     * happens if the target and currentTarget of the event are not the same. It will omit any extra data in the
     * `formData` in the state if `omitExtraData` is true. It will validate the resulting `formData`, reporting errors
     * via the `onError()` callback unless validation is disabled. Finally, it will add in any `extraErrors` and then call
     * back the `onSubmit` callback if it was provided.
     *
     * @param event - The submit HTML form event
     */
    onSubmit = (event) => {
      event.preventDefault();
      if (event.target !== event.currentTarget) {
        return;
      }
      event.persist();
      const { omitExtraData, extraErrors, noValidate, onSubmit } = this.props;
      let { formData: newFormData } = this.state;
      if (omitExtraData === true) {
        newFormData = this.omitExtraData(newFormData);
      }
      if (noValidate || this.validateFormWithFormData(newFormData)) {
        const errorSchema = extraErrors || {};
        const errors = extraErrors ? utils.toErrorList(extraErrors) : [];
        this.setState(
          {
            formData: newFormData,
            errors,
            errorSchema,
            schemaValidationErrors: [],
            schemaValidationErrorSchema: {}
          },
          () => {
            if (onSubmit) {
              onSubmit(toIChangeEvent({ ...this.state, formData: newFormData }, "submitted"), event);
            }
          }
        );
      }
    };
    /** Extracts the `GlobalFormOptions` from the given Form `props`
     *
     * @param props - The form props to extract the global form options from
     * @returns - The `GlobalFormOptions` from the props
     * @private
     */
    static getGlobalFormOptions(props) {
      const {
        uiSchema = {},
        experimental_componentUpdateStrategy,
        idSeparator = utils.DEFAULT_ID_SEPARATOR,
        idPrefix = utils.DEFAULT_ID_PREFIX,
        nameGenerator,
        useFallbackUiForUnsupportedType = false
      } = props;
      const rootFieldId = uiSchema["ui:rootFieldId"];
      return {
        idPrefix: rootFieldId || idPrefix,
        idSeparator,
        useFallbackUiForUnsupportedType,
        ...experimental_componentUpdateStrategy !== void 0 && { experimental_componentUpdateStrategy },
        ...nameGenerator !== void 0 && { nameGenerator }
      };
    }
    /** Computed the registry for the form using the given `props`, `schema` and `schemaUtils` */
    static getRegistry(props, schema, schemaUtils) {
      const { translateString: customTranslateString, uiSchema = {} } = props;
      const { fields: fields2, templates: templates2, widgets: widgets2, formContext, translateString } = getDefaultRegistry();
      return {
        fields: { ...fields2, ...props.fields },
        templates: {
          ...templates2,
          ...props.templates,
          ButtonTemplates: {
            ...templates2.ButtonTemplates,
            ...props.templates?.ButtonTemplates
          }
        },
        widgets: { ...widgets2, ...props.widgets },
        rootSchema: schema,
        formContext: props.formContext || formContext,
        schemaUtils,
        translateString: customTranslateString || translateString,
        globalUiOptions: uiSchema[utils.UI_GLOBAL_OPTIONS_KEY],
        globalFormOptions: _Form.getGlobalFormOptions(props),
        uiSchemaDefinitions: uiSchema[utils.UI_DEFINITIONS_KEY] ?? {}
      };
    }
    /** Provides a function that can be used to programmatically submit the `Form` */
    submit = () => {
      if (this.formElement.current) {
        const submitCustomEvent = new CustomEvent("submit", {
          cancelable: true
        });
        submitCustomEvent.preventDefault();
        this.formElement.current.dispatchEvent(submitCustomEvent);
        this.formElement.current.requestSubmit();
      }
    };
    /** Attempts to focus on the field associated with the `error`. Uses the `property` field to compute path of the error
     * field, then, using the `idPrefix` and `idSeparator` converts that path into an id. Then the input element with that
     * id is attempted to be found using the `formElement` ref. If it is located, then it is focused.
     *
     * @param error - The error on which to focus
     */
    focusOnError(error) {
      const { idPrefix = "root", idSeparator = "_" } = this.props;
      const { property } = error;
      const path = _toPath(property);
      if (path[0] === "") {
        path[0] = idPrefix;
      } else {
        path.unshift(idPrefix);
      }
      const elementId = path.join(idSeparator);
      let field = this.formElement.current.elements[elementId];
      if (!field) {
        field = this.formElement.current.querySelector(`input[id^="${elementId}"], button[id^="${elementId}"]`);
      }
      if (field && field.length) {
        field = field[0];
      }
      if (field) {
        field.focus();
      }
    }
    /** Validates the form using the given `formData`. For use on form submission or on programmatic validation.
     * If `onError` is provided, then it will be called with the list of errors.
     *
     * @param formData - The form data to validate
     * @returns - True if the form is valid, false otherwise.
     */
    validateFormWithFormData = (formData) => {
      const { extraErrors, extraErrorsBlockSubmit, focusOnFirstError, onError } = this.props;
      const { errors: prevErrors } = this.state;
      const schemaValidation = this.validate(formData);
      const { errors, errorSchema } = extraErrors ? _Form.mergeErrors(schemaValidation, extraErrors) : schemaValidation;
      const hasError = schemaValidation.errors.length > 0 || extraErrors && extraErrorsBlockSubmit;
      if (hasError) {
        if (focusOnFirstError) {
          if (typeof focusOnFirstError === "function") {
            focusOnFirstError(errors[0]);
          } else {
            this.focusOnError(errors[0]);
          }
        }
        this.setState(
          {
            errors,
            errorSchema,
            schemaValidationErrors: schemaValidation.errors,
            schemaValidationErrorSchema: schemaValidation.errorSchema
          },
          () => {
            if (onError) {
              onError(errors);
            } else {
              console.error("Form validation failed", errors);
            }
          }
        );
      } else if (errors.length > 0) {
        this.setState({
          errors,
          errorSchema,
          schemaValidationErrors: [],
          schemaValidationErrorSchema: {}
        });
      } else if (prevErrors.length > 0) {
        this.setState({
          errors: [],
          errorSchema: {},
          schemaValidationErrors: [],
          schemaValidationErrorSchema: {}
        });
      }
      return !hasError;
    };
    /** Programmatically validate the form.  If `omitExtraData` is true, the `formData` will first be filtered to remove
     * any extra data not in a form field. If `onError` is provided, then it will be called with the list of errors the
     * same way as would happen on form submission.
     *
     * @returns - True if the form is valid, false otherwise.
     */
    validateForm() {
      const { omitExtraData } = this.props;
      let { formData: newFormData } = this.state;
      if (omitExtraData === true) {
        newFormData = this.omitExtraData(newFormData);
      }
      return this.validateFormWithFormData(newFormData);
    }
    /** Renders the `Form` fields inside the <form> | `tagName` or `_internalFormWrapper`, rendering any errors if
     * needed along with the submit button or any children of the form.
     */
    render() {
      const {
        children,
        id,
        className = "",
        tagName,
        name,
        method,
        target,
        action,
        autoComplete,
        enctype,
        acceptCharset,
        noHtml5Validate = false,
        disabled,
        readonly,
        showErrorList = "top",
        _internalFormWrapper
      } = this.props;
      const { schema, uiSchema, formData, errorSchema, fieldPathId, registry } = this.state;
      const { SchemaField: SchemaFieldComponent } = registry.fields;
      const { SubmitButton: SubmitButton2 } = registry.templates.ButtonTemplates;
      const as = _internalFormWrapper ? tagName : void 0;
      const FormTag = _internalFormWrapper || tagName || "form";
      let { [utils.SUBMIT_BTN_OPTIONS_KEY]: submitOptions = {} } = utils.getUiOptions(uiSchema);
      if (disabled) {
        submitOptions = { ...submitOptions, props: { ...submitOptions.props, disabled: true } };
      }
      const submitUiSchema = { [utils.UI_OPTIONS_KEY]: { [utils.SUBMIT_BTN_OPTIONS_KEY]: submitOptions } };
      return /* @__PURE__ */ jsxRuntime.jsxs(
        FormTag,
        {
          className: className || "rjsf",
          id,
          name,
          method,
          target,
          action,
          autoComplete,
          encType: enctype,
          acceptCharset,
          noValidate: noHtml5Validate,
          onSubmit: this.onSubmit,
          as,
          ref: this.formElement,
          children: [
            showErrorList === "top" && this.renderErrors(registry),
            /* @__PURE__ */ jsxRuntime.jsx(
              SchemaFieldComponent,
              {
                name: "",
                schema,
                uiSchema,
                errorSchema,
                fieldPathId,
                formData,
                onChange: this.onChange,
                onBlur: this.onBlur,
                onFocus: this.onFocus,
                registry,
                disabled,
                readonly
              }
            ),
            children || /* @__PURE__ */ jsxRuntime.jsx(SubmitButton2, { uiSchema: submitUiSchema, registry }),
            showErrorList === "bottom" && this.renderErrors(registry)
          ]
        }
      );
    }
  };
  function getTestRegistry(rootSchema, fields2 = {}, templates2 = {}, widgets2 = {}, formContext = {}, globalFormOptions = {
    idPrefix: utils.DEFAULT_ID_PREFIX,
    idSeparator: utils.DEFAULT_ID_SEPARATOR,
    useFallbackUiForUnsupportedType: false
  }) {
    const defaults = getDefaultRegistry();
    const schemaUtils = utils.createSchemaUtils(validator, rootSchema);
    return {
      fields: { ...defaults.fields, ...fields2 },
      templates: { ...defaults.templates, ...templates2 },
      widgets: { ...defaults.widgets, ...widgets2 },
      formContext,
      rootSchema,
      schemaUtils,
      translateString: utils.englishStringTranslator,
      globalFormOptions
    };
  }
  function withTheme(themeProps) {
    return react.forwardRef(
      ({ fields: propFields, widgets: propWidgets, templates: propTemplates, ...directProps }, ref) => {
        const fields2 = { ...themeProps?.fields, ...propFields };
        const widgets2 = { ...themeProps?.widgets, ...propWidgets };
        const templates2 = {
          ...themeProps?.templates,
          ...propTemplates,
          ButtonTemplates: {
            ...themeProps?.templates?.ButtonTemplates,
            ...propTemplates?.ButtonTemplates
          }
        };
        return /* @__PURE__ */ jsxRuntime.jsx(
          Form,
          {
            ...themeProps,
            ...directProps,
            fields: fields2,
            widgets: widgets2,
            templates: templates2,
            ref
          }
        );
      }
    );
  }

  // src/index.ts
  var index_default = Form;

  exports.RichDescription = RichDescription;
  exports.RichHelp = RichHelp;
  exports.SchemaExamples = SchemaExamples;
  exports.default = index_default;
  exports.getDefaultRegistry = getDefaultRegistry;
  exports.getTestRegistry = getTestRegistry;
  exports.withTheme = withTheme;

  Object.defineProperty(exports, '__esModule', { value: true });

}));
